<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang='en' xml:lang='en' xmlns='http://www.w3.org/1999/xhtml'>
  <head>
    <title>lib/bio/db/nbrf.rb</title>
    <link href="screen.css" media="all" rel="stylesheet" type="text/css" />
    <link href="print.css" media="print" rel="stylesheet" type="text/css" />
    
    <meta http-equiv="Content-type" content="text/html;charset=UTF-8" />
    <script type="text/javascript" src="rcov.js"></script>
  </head>
  <body>
    <h1>Bioruby C0 Coverage Information - RCov</h1>
    <h2>lib/bio/db/nbrf.rb</h2>

    

    <div class="report_table_wrapper">
      <table class='report' id='report_table'>
        <thead>
          <tr>
            <th class="left_align">Name</th>
            <th class="right_align">Total Lines</th>
            <th class="right_align">Lines of Code</th>
            <th class="left_align">Total Coverage</th>
            <th class="left_align">Code Coverage</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td class="left_align"><a href="lib-bio-db-nbrf_rb.html">lib/bio/db/nbrf.rb</a></td>
            <td class='right_align'><tt>191</tt></td>
            <td class='right_align'><tt>96</tt></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>84.82%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:85px"></div>
            <div class="uncovered" style="width:15px"></div>
          </div></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>70.83%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:71px"></div>
            <div class="uncovered" style="width:29px"></div>
          </div></td>
          </tr>
        </tbody>
      </table>
    </div>
    
    <h3>Key</h3>
    
    <div class="key"><pre><span class='marked'>Code reported as executed by Ruby looks like this...</span><span class='marked1'>and this: this line is also marked as covered.</span><span class='inferred'>Lines considered as run by rcov, but not reported by Ruby, look like this,</span><span class='inferred1'>and this: these lines were inferred by rcov (using simple heuristics).</span><span class='uncovered'>Finally, here's a line marked as not executed.</span></pre></div>

    <h3>Coverage Details</h3>

    <table class="details">
      <tbody>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1">1</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line2">2</a> # = bio/db/nbrf.rb - NBRF/PIR format sequence data class</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line3">3</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line4">4</a> # Copyright:: Copyright (C) 2001-2003,2006 Naohisa Goto &lt;ng@bioruby.org&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line5">5</a> #             Copyright (C) 2001-2002 Toshiaki Katayama &lt;k@bioruby.org&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line6">6</a> # License::   The Ruby License</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line7">7</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line8">8</a> #  $Id: nbrf.rb,v 1.10 2007/04/05 23:35:40 trevor Exp $</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line9">9</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line10">10</a> # Sequence data class for NBRF/PIR flatfile format.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line11">11</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line12">12</a> # = References</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line13">13</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line14">14</a> # * http://pir.georgetown.edu/pirwww/otherinfo/doc/techbulletin.html</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line15">15</a> # * http://www.sander.embl-ebi.ac.uk/Services/webin/help/webin-align/align_format_help.html#pir</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line16">16</a> # * http://www.cmbi.kun.nl/bioinf/tools/crab_pir.html</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line17">17</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line18">18</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line19">19</a> require 'bio/db'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line20">20</a> require 'bio/sequence'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line21">21</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line22">22</a> module Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line23">23</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line24">24</a>   # Sequence data class for NBRF/PIR flatfile format.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line25">25</a>   class NBRF &lt; DB</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line26">26</a>     #--</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line27">27</a>     # based on Bio::FastaFormat class</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line28">28</a>     #++</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line29">29</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line30">30</a>     # Delimiter of each entry. Bio::FlatFile uses it.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line31">31</a>     DELIMITER	= RS = &quot;\n&gt;&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line32">32</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line33">33</a>     # (Integer) excess read size included in DELIMITER.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line34">34</a>     DELIMITER_OVERRUN = 1 # '&gt;'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line35">35</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line36">36</a>     #--</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line37">37</a>     # Note: DELIMITER is changed due to the change of Bio::FlatFile.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line38">38</a>     # DELIMITER	= RS = &quot;*\n&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line39">39</a>     #++</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line40">40</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line41">41</a>     # Creates a new NBRF object. It stores the comment and sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line42">42</a>     # information from one entry of the NBRF/PIR format string.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line43">43</a>     # If the argument contains more than one</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line44">44</a>     # entry, only the first entry is used.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line45">45</a>     def initialize(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line46">46</a>       str = str.sub(/\A[\r\n]+/, '') # remove first void lines</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line47">47</a>       line1, line2, rest = str.split(/^/, 3)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line48">48</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line49">49</a>       rest = rest.to_s</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line50">50</a>       rest.sub!(/^&gt;.*/m, '') # remove trailing entries for sure</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line51">51</a>       @entry_overrun = $&amp;</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line52">52</a>       rest.sub!(/\*\s*\z/, '') # remove last '*' and &quot;\n&quot;</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line53">53</a>       @data = rest</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line54">54</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line55">55</a>       @definition = line2.to_s.chomp</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line56">56</a>       if /^&gt;?([A-Za-z0-9]{2})\;(.*)/ =~ line1.to_s then</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line57">57</a>         @seq_type = $1</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line58">58</a>         @entry_id = $2</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line59">59</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line60">60</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line61">61</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line62">62</a>     # Returns sequence type described in the entry.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line63">63</a>     #  P1 (protein), F1 (protein fragment)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line64">64</a>     #  DL (DNA linear), DC (DNA circular)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line65">65</a>     #  RL (DNA linear), RC (DNA circular)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line66">66</a>     #  N3 (tRNA), N1 (other functional RNA)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line67">67</a>     attr_accessor :seq_type</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line68">68</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line69">69</a>     # Returns ID described in the entry.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line70">70</a>     attr_accessor :entry_id</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line71">71</a>     alias accession entry_id</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line72">72</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line73">73</a>     # Returns the description line of the NBRF/PIR formatted data.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line74">74</a>     attr_accessor :definition</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line75">75</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line76">76</a>     # sequence data of the entry (???)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line77">77</a>     attr_accessor :data</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line78">78</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line79">79</a>     # piece of next entry. Bio::FlatFile uses it.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line80">80</a>     attr_reader :entry_overrun</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line81">81</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line82">82</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line83">83</a>     # Returns the stored one entry as a NBRF/PIR format. (same as to_s)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line84">84</a>     def entry</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line85">85</a>       @entry = &quot;&gt;#{@seq_type or 'XX'};#{@entry_id}\n#{definition}\n#{@data}*\n&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line86">86</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line87">87</a>     alias to_s entry</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line88">88</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line89">89</a>     # Returns Bio::Sequence::AA, Bio::Sequence::NA, or Bio::Sequence,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line90">90</a>     # depending on sequence type.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line91">91</a>     def seq_class</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line92">92</a>       case @seq_type</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line93">93</a>       when /[PF]1/</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line94">94</a>         # protein</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line95">95</a>         Sequence::AA</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line96">96</a>       when /[DR][LC]/, /N[13]/</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line97">97</a>         # nucleic</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line98">98</a>         Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line99">99</a>       else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line100">100</a>         Sequence</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line101">101</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line102">102</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line103">103</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line104">104</a>     # Returns sequence data.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line105">105</a>     # Returns Bio::Sequence::NA, Bio::Sequence::AA or Bio::Sequence,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line106">106</a>     # according to the sequence type.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line107">107</a>     def seq</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line108">108</a>       unless defined?(@seq)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line109">109</a>         @seq = seq_class.new(@data.tr(&quot; \t\r\n0-9&quot;, '')) # lazy clean up</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line110">110</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line111">111</a>       @seq</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line112">112</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line113">113</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line114">114</a>     # Returns sequence length.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line115">115</a>     def length</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line116">116</a>       seq.length</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line117">117</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line118">118</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line119">119</a>     # Returens the nucleic acid sequence.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line120">120</a>     # If you call naseq for protein sequence, RuntimeError will be occurred.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line121">121</a>     # Use the method if you know whether the sequence is NA or AA.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line122">122</a>     def naseq</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line123">123</a>       if seq.is_a?(Bio::Sequence::AA) then</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line124">124</a>         raise 'not nucleic but protein sequence'</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line125">125</a>       elsif seq.is_a?(Bio::Sequence::NA) then</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line126">126</a>         seq</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line127">127</a>       else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line128">128</a>         Bio::Sequence::NA.new(seq)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line129">129</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line130">130</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line131">131</a>       </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line132">132</a>     # Returens the length of sequence.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line133">133</a>     # If you call nalen for protein sequence, RuntimeError will be occurred.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line134">134</a>     # Use the method if you know whether the sequence is NA or AA.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line135">135</a>     def nalen</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line136">136</a>       naseq.length</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line137">137</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line138">138</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line139">139</a>     # Returens the protein (amino acids) sequence.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line140">140</a>     # If you call aaseq for nucleic acids sequence,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line141">141</a>     # RuntimeError will be occurred.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line142">142</a>     # Use the method if you know whether the sequence is NA or AA.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line143">143</a>     def aaseq</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line144">144</a>       if seq.is_a?(Bio::Sequence::NA) then</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line145">145</a>         raise 'not nucleic but protein sequence'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line146">146</a>       elsif seq.is_a?(Bio::Sequence::AA) then</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line147">147</a>         seq</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line148">148</a>       else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line149">149</a>         Bio::Sequence::AA.new(seq)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line150">150</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line151">151</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line152">152</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line153">153</a>     # Returens the length of protein (amino acids) sequence.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line154">154</a>     # If you call aaseq for nucleic acids sequence,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line155">155</a>     # RuntimeError will be occurred.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line156">156</a>     # Use the method if you know whether the sequence is NA or AA.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line157">157</a>     def aalen</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line158">158</a>       aaseq.length</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line159">159</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line160">160</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line161">161</a>     #--</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line162">162</a>     #class method</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line163">163</a>     #++</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line164">164</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line165">165</a>     # Creates a NBRF/PIR formatted text.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line166">166</a>     # Parameters can be omitted.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line167">167</a>     def self.to_nbrf(hash)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line168">168</a>       seq_type = hash[:seq_type]</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line169">169</a>       seq = hash[:seq]</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line170">170</a>       unless seq_type</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line171">171</a>         if seq.is_a?(Bio::Sequence::AA) then</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line172">172</a>           seq_type = 'P1'</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line173">173</a>         elsif seq.is_a?(Bio::Sequence::NA) then</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line174">174</a>           seq_type = /u/i =~ seq ? 'RL' : 'DL'</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line175">175</a>         else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line176">176</a>           seq_type = 'XX'</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line177">177</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line178">178</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line179">179</a>       width = hash.has_key?(:width) ? hash[:width] : 70</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line180">180</a>       if width then</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line181">181</a>         seq = seq.to_s + &quot;*&quot;</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line182">182</a>         seq.gsub!(Regexp.new(&quot;.{1,#{width}}&quot;), &quot;\\0\n&quot;)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line183">183</a>       else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line184">184</a>         seq = seq.to_s + &quot;*\n&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line185">185</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line186">186</a>       &quot;&gt;#{seq_type};#{hash[:entry_id]}\n#{hash[:definition]}\n#{seq}&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line187">187</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line188">188</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line189">189</a>   end #class NBRF</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line190">190</a> end #module Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line191">191</a> </pre></td>
          </tr>
        
      </tbody>
    </table>

    <p>Generated on Fri Jun 25 17:37:23 +0900 2010 with <a href="http://github.com/relevance/rcov">rcov 0.9.8</a></p>

  </body>
</html>
