<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang='en' xml:lang='en' xmlns='http://www.w3.org/1999/xhtml'>
  <head>
    <title>lib/bio/db/kegg/genome.rb</title>
    <link href="screen.css" media="all" rel="stylesheet" type="text/css" />
    <link href="print.css" media="print" rel="stylesheet" type="text/css" />
    
    <meta http-equiv="Content-type" content="text/html;charset=UTF-8" />
    <script type="text/javascript" src="rcov.js"></script>
  </head>
  <body>
    <h1>Bioruby C0 Coverage Information - RCov</h1>
    <h2>lib/bio/db/kegg/genome.rb</h2>

    

    <div class="report_table_wrapper">
      <table class='report' id='report_table'>
        <thead>
          <tr>
            <th class="left_align">Name</th>
            <th class="right_align">Total Lines</th>
            <th class="right_align">Lines of Code</th>
            <th class="left_align">Total Coverage</th>
            <th class="left_align">Code Coverage</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td class="left_align"><a href="lib-bio-db-kegg-genome_rb.html">lib/bio/db/kegg/genome.rb</a></td>
            <td class='right_align'><tt>208</tt></td>
            <td class='right_align'><tt>132</tt></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>100.00%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:100px"></div>
            <div class="uncovered" style="width:0px"></div>
          </div></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>100.00%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:100px"></div>
            <div class="uncovered" style="width:0px"></div>
          </div></td>
          </tr>
        </tbody>
      </table>
    </div>
    
    <h3>Key</h3>
    
    <div class="key"><pre><span class='marked'>Code reported as executed by Ruby looks like this...</span><span class='marked1'>and this: this line is also marked as covered.</span><span class='inferred'>Lines considered as run by rcov, but not reported by Ruby, look like this,</span><span class='inferred1'>and this: these lines were inferred by rcov (using simple heuristics).</span><span class='uncovered'>Finally, here's a line marked as not executed.</span></pre></div>

    <h3>Coverage Details</h3>

    <table class="details">
      <tbody>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1">1</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line2">2</a> # = bio/db/kegg/genome.rb - KEGG/GENOME database class</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line3">3</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line4">4</a> # Copyright::  Copyright (C) 2001, 2002, 2007 Toshiaki Katayama &lt;k@bioruby.org&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line5">5</a> # License::    The Ruby License</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line6">6</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line7">7</a> # $Id:$</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line8">8</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line9">9</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line10">10</a> require 'bio/db'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line11">11</a> require 'bio/reference'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line12">12</a> require 'bio/db/kegg/common'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line13">13</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line14">14</a> module Bio</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line15">15</a> class KEGG</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line16">16</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line17">17</a> # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line18">18</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line19">19</a> # Parser for the KEGG GENOME database</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line20">20</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line21">21</a> # == References</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line22">22</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line23">23</a> # * ftp://ftp.genome.jp/pub/kegg/genomes/genome</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line24">24</a> # * http://www.genome.jp/dbget-bin/www_bfind?genome</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line25">25</a> # * http://www.genome.jp/kegg/catalog/org_list.html</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line26">26</a> #</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line27">27</a> class GENOME &lt; KEGGDB</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line28">28</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line29">29</a>   DELIMITER	= RS = &quot;\n///\n&quot;</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line30">30</a>   TAGSIZE	= 12</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line31">31</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line32">32</a>   include Common::References</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line33">33</a>   # REFERENCE -- Returns contents of the REFERENCE records as an Array of</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line34">34</a>   # Bio::Reference objects.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line35">35</a>   def references; super; end if false #dummy for RDoc</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line36">36</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line37">37</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line38">38</a>   def initialize(entry)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line39">39</a>     super(entry, TAGSIZE)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line40">40</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line41">41</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line42">42</a>   # (private) Returns a tag name of the field as a String.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line43">43</a>   # Needed to redefine because of the PLASMID field.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line44">44</a>   def tag_get(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line45">45</a>     if /\APLASMID\s+/ =~ str.to_s then</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line46">46</a>       'PLASMID'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line47">47</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line48">48</a>       super(str)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line49">49</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line50">50</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line51">51</a>   private :tag_get</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line52">52</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line53">53</a>   # (private) Returns a String of the field without a tag name.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line54">54</a>   # Needed to redefine because of the PLASMID field.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line55">55</a>   def tag_cut(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line56">56</a>     if /\APLASMID\s+/ =~ str.to_s then</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line57">57</a>       $'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line58">58</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line59">59</a>       super(str)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line60">60</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line61">61</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line62">62</a>   private :tag_cut</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line63">63</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line64">64</a>   # ENTRY -- Returns contents of the ENTRY record as a String.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line65">65</a>   def entry_id</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line66">66</a>     field_fetch('ENTRY')[/\S+/]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line67">67</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line68">68</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line69">69</a>   # NAME -- Returns contents of the NAME record as a String.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line70">70</a>   def name</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line71">71</a>     field_fetch('NAME')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line72">72</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line73">73</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line74">74</a>   # DEFINITION -- Returns contents of the DEFINITION record as a String.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line75">75</a>   def definition</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line76">76</a>     field_fetch('DEFINITION')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line77">77</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line78">78</a>   alias organism definition</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line79">79</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line80">80</a>   # TAXONOMY -- Returns contents of the TAXONOMY record as a Hash.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line81">81</a>   def taxonomy</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line82">82</a>     unless @data['TAXONOMY']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line83">83</a>       taxid, lineage = subtag2array(get('TAXONOMY'))</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line84">84</a>       taxid   = taxid   ? truncate(tag_cut(taxid))   : ''</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line85">85</a>       lineage = lineage ? truncate(tag_cut(lineage)) : ''</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line86">86</a>       @data['TAXONOMY'] = {</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line87">87</a>         'taxid'	=&gt; taxid,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line88">88</a>         'lineage'	=&gt; lineage,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line89">89</a>       }</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line90">90</a>       @data['TAXONOMY'].default = ''</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line91">91</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line92">92</a>     @data['TAXONOMY']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line93">93</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line94">94</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line95">95</a>   # Returns NCBI taxonomy ID from the TAXONOMY record as a String.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line96">96</a>   def taxid</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line97">97</a>     taxonomy['taxid']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line98">98</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line99">99</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line100">100</a>   # Returns contents of the TAXONOMY/LINEAGE record as a String.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line101">101</a>   def lineage</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line102">102</a>     taxonomy['lineage']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line103">103</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line104">104</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line105">105</a>   # DATA_SOURCE -- Returns contents of the DATA_SOURCE record as a String.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line106">106</a>   def data_source</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line107">107</a>     field_fetch('DATA_SOURCE')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line108">108</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line109">109</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line110">110</a>   # ORIGINAL_DB -- Returns contents of the ORIGINAL_DB record as a String.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line111">111</a>   def original_db</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line112">112</a>     #field_fetch('ORIGINAL_DB')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line113">113</a>     unless defined?(@original_db)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line114">114</a>       @original_db = fetch('ORIGINAL_DB')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line115">115</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line116">116</a>     @original_db</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line117">117</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line118">118</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line119">119</a>   # Returns ORIGINAL_DB record as an Array containing String objects.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line120">120</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line121">121</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line122">122</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line123">123</a>   # *Returns*:: Array containing String objects</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line124">124</a>   def original_databases</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line125">125</a>     lines_fetch('ORIGINAL_DB')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line126">126</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line127">127</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line128">128</a>   # DISEASE -- Returns contents of the COMMENT record as a String.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line129">129</a>   def disease</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line130">130</a>     field_fetch('DISEASE')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line131">131</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line132">132</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line133">133</a>   # COMMENT -- Returns contents of the COMMENT record as a String.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line134">134</a>   def comment</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line135">135</a>     field_fetch('COMMENT')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line136">136</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line137">137</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line138">138</a>   # CHROMOSOME -- Returns contents of the CHROMOSOME records as an Array</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line139">139</a>   # of Hash.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line140">140</a>   def chromosomes</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line141">141</a>     unless @data['CHROMOSOME']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line142">142</a>       @data['CHROMOSOME'] = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line143">143</a>       toptag2array(get('CHROMOSOME')).each do |chr|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line144">144</a>         hash = Hash.new('')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line145">145</a>         subtag2array(chr).each do |field|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line146">146</a>           hash[tag_get(field)] = truncate(tag_cut(field))</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line147">147</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line148">148</a>         @data['CHROMOSOME'].push(hash)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line149">149</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line150">150</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line151">151</a>     @data['CHROMOSOME']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line152">152</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line153">153</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line154">154</a>   # PLASMID -- Returns contents of the PLASMID records as an Array of Hash.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line155">155</a>   def plasmids</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line156">156</a>     unless @data['PLASMID']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line157">157</a>       @data['PLASMID'] = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line158">158</a>       toptag2array(get('PLASMID')).each do |chr|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line159">159</a>         hash = Hash.new('')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line160">160</a>         subtag2array(chr).each do |field|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line161">161</a>           hash[tag_get(field)] = truncate(tag_cut(field))</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line162">162</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line163">163</a>         @data['PLASMID'].push(hash)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line164">164</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line165">165</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line166">166</a>     @data['PLASMID']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line167">167</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line168">168</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line169">169</a>   # STATISTICS -- Returns contents of the STATISTICS record as a Hash.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line170">170</a>   def statistics</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line171">171</a>     unless @data['STATISTICS']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line172">172</a>       hash = Hash.new(0.0)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line173">173</a>       get('STATISTICS').each_line do |line|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line174">174</a>         case line</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line175">175</a>         when /nucleotides:\s+(\d+)/</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line176">176</a>           hash['num_nuc'] = $1.to_i</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line177">177</a>         when /protein genes:\s+(\d+)/</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line178">178</a>           hash['num_gene'] = $1.to_i</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line179">179</a>         when /RNA genes:\s+(\d+)/</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line180">180</a>           hash['num_rna'] = $1.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line181">181</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line182">182</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line183">183</a>       @data['STATISTICS'] = hash</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line184">184</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line185">185</a>     @data['STATISTICS']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line186">186</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line187">187</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line188">188</a>   # Returns number of nucleotides from the STATISTICS record as a Fixnum.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line189">189</a>   def nalen</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line190">190</a>     statistics['num_nuc']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line191">191</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line192">192</a>   alias length nalen</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line193">193</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line194">194</a>   # Returns number of protein genes from the STATISTICS record as a Fixnum.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line195">195</a>   def num_gene</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line196">196</a>     statistics['num_gene']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line197">197</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line198">198</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line199">199</a>   # Returns number of rna from the STATISTICS record as a Fixnum.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line200">200</a>   def num_rna</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line201">201</a>     statistics['num_rna']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line202">202</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line203">203</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line204">204</a> end # GENOME</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line205">205</a>     </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line206">206</a> end # KEGG</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line207">207</a> end # Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line208">208</a> </pre></td>
          </tr>
        
      </tbody>
    </table>

    <p>Generated on Fri Jun 25 17:37:23 +0900 2010 with <a href="http://github.com/relevance/rcov">rcov 0.9.8</a></p>

  </body>
</html>
