<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang='en' xml:lang='en' xmlns='http://www.w3.org/1999/xhtml'>
  <head>
    <title>lib/bio/db/embl/embl.rb</title>
    <link href="screen.css" media="all" rel="stylesheet" type="text/css" />
    <link href="print.css" media="print" rel="stylesheet" type="text/css" />
    
    <meta http-equiv="Content-type" content="text/html;charset=UTF-8" />
    <script type="text/javascript" src="rcov.js"></script>
  </head>
  <body>
    <h1>Bioruby C0 Coverage Information - RCov</h1>
    <h2>lib/bio/db/embl/embl.rb</h2>

    

    <div class="report_table_wrapper">
      <table class='report' id='report_table'>
        <thead>
          <tr>
            <th class="left_align">Name</th>
            <th class="right_align">Total Lines</th>
            <th class="right_align">Lines of Code</th>
            <th class="left_align">Total Coverage</th>
            <th class="left_align">Code Coverage</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td class="left_align"><a href="lib-bio-db-embl-embl_rb.html">lib/bio/db/embl/embl.rb</a></td>
            <td class='right_align'><tt>497</tt></td>
            <td class='right_align'><tt>232</tt></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>96.18%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:96px"></div>
            <div class="uncovered" style="width:4px"></div>
          </div></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>91.81%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:92px"></div>
            <div class="uncovered" style="width:8px"></div>
          </div></td>
          </tr>
        </tbody>
      </table>
    </div>
    
    <h3>Key</h3>
    
    <div class="key"><pre><span class='marked'>Code reported as executed by Ruby looks like this...</span><span class='marked1'>and this: this line is also marked as covered.</span><span class='inferred'>Lines considered as run by rcov, but not reported by Ruby, look like this,</span><span class='inferred1'>and this: these lines were inferred by rcov (using simple heuristics).</span><span class='uncovered'>Finally, here's a line marked as not executed.</span></pre></div>

    <h3>Coverage Details</h3>

    <table class="details">
      <tbody>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1">1</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line2">2</a> # = bio/db/embl/embl.rb - EMBL database class</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line3">3</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line4">4</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line5">5</a> # Copyright::   Copyright (C) 2001-2007</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line6">6</a> #               Mitsuteru C. Nakao &lt;n@bioruby.org&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line7">7</a> #               Jan Aerts &lt;jan.aerts@bbsrc.ac.uk&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line8">8</a> # License::     The Ruby License</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line9">9</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line10">10</a> # $Id: embl.rb,v 1.29.2.7 2008/06/17 16:04:36 ngoto Exp $</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line11">11</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line12">12</a> # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line13">13</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line14">14</a> # Parser class for EMBL database entry.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line15">15</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line16">16</a> # == Examples</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line17">17</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line18">18</a> #   emb = Bio::EMBL.new($&lt;.read)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line19">19</a> #   emb.entry_id</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line20">20</a> #   emb.each_cds do |cds|</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line21">21</a> #     cds # A CDS in feature table.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line22">22</a> #   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line23">23</a> #   emb.seq #=&gt; &quot;ACGT...&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line24">24</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line25">25</a> # == References</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line26">26</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line27">27</a> # * The EMBL Nucleotide Sequence Database</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line28">28</a> #   http://www.ebi.ac.uk/embl/</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line29">29</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line30">30</a> # * The EMBL Nucleotide Sequence Database: Users Manual</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line31">31</a> #   http://www.ebi.ac.uk/embl/Documentation/User_manual/usrman.html</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line32">32</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line33">33</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line34">34</a> require 'date'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line35">35</a> require 'bio/db'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line36">36</a> require 'bio/db/embl/common'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line37">37</a> require 'bio/compat/features'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line38">38</a> require 'bio/compat/references'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line39">39</a> require 'bio/sequence'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line40">40</a> require 'bio/sequence/dblink'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line41">41</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line42">42</a> module Bio</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line43">43</a> class EMBL &lt; EMBLDB</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line44">44</a>   include Bio::EMBLDB::Common</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line45">45</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line46">46</a>   # returns contents in the ID line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line47">47</a>   # * Bio::EMBL#id_line -&gt; &lt;ID Hash&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line48">48</a>   # where &lt;ID Hash&gt; is:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line49">49</a>   #  {'ENTRY_NAME' =&gt; String, 'MOLECULE_TYPE' =&gt; String, 'DIVISION' =&gt; String,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line50">50</a>   #   'SEQUENCE_LENGTH' =&gt; Int, 'SEQUENCE_VERSION' =&gt; Int}</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line51">51</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line52">52</a>   # ID Line</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line53">53</a>   #  &quot;ID  ENTRY_NAME DATA_CLASS; MOLECULE_TYPE; DIVISION; SEQUENCE_LENGTH BP.&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line54">54</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line55">55</a>   # DATA_CLASS = ['standard']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line56">56</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line57">57</a>   # MOLECULE_TYPE: DNA RNA XXX</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line58">58</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line59">59</a>   # Code ( DIVISION )</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line60">60</a>   #  EST (ESTs)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line61">61</a>   #  PHG (Bacteriophage)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line62">62</a>   #  FUN (Fungi)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line63">63</a>   #  GSS (Genome survey)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line64">64</a>   #  HTC (High Throughput cDNAs) </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line65">65</a>   #  HTG (HTGs)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line66">66</a>   #  HUM (Human)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line67">67</a>   #  INV (Invertebrates)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line68">68</a>   #  ORG (Organelles)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line69">69</a>   #  MAM (Other Mammals)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line70">70</a>   #  VRT (Other Vertebrates)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line71">71</a>   #  PLN (Plants)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line72">72</a>   #  PRO (Prokaryotes)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line73">73</a>   #  ROD (Rodents)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line74">74</a>   #  SYN (Synthetic)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line75">75</a>   #  STS (STSs)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line76">76</a>   #  UNC (Unclassified)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line77">77</a>   #  VRL (Viruses)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line78">78</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line79">79</a>   # Rel 89-</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line80">80</a>   # ID   CD789012; SV 4; linear; genomic DNA; HTG; MAM; 500 BP.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line81">81</a>   # ID &lt;1&gt;; SV &lt;2&gt;; &lt;3&gt;; &lt;4&gt;; &lt;5&gt;; &lt;6&gt;; &lt;7&gt; BP.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line82">82</a>   # 1. Primary accession number</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line83">83</a>   # 2. Sequence version number</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line84">84</a>   # 3. Topology: 'circular' or 'linear'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line85">85</a>   # 4. Molecule type (see note 1 below)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line86">86</a>   # 5. Data class (see section 3.1)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line87">87</a>   # 6. Taxonomic division (see section 3.2)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line88">88</a>   # 7. Sequence length (see note 2 below)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line89">89</a>   def id_line(key=nil)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line90">90</a>     unless @data['ID']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line91">91</a>       tmp = Hash.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line92">92</a>       idline = fetch('ID').split(/; +/)         </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line93">93</a>       tmp['ENTRY_NAME'], tmp['DATA_CLASS'] = idline.shift.split(/ +/)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line94">94</a>       if idline.first =~ /^SV/</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line95">95</a>         tmp['SEQUENCE_VERSION'] = idline.shift.split(' ').last</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line96">96</a>         tmp['TOPOLOGY'] = idline.shift</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line97">97</a>         tmp['MOLECULE_TYPE'] = idline.shift</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line98">98</a>         tmp['DATA_CLASS'] = idline.shift</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line99">99</a>       else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line100">100</a>         tmp['MOLECULE_TYPE'] = idline.shift</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line101">101</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line102">102</a>       tmp['DIVISION'] = idline.shift</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line103">103</a>       tmp['SEQUENCE_LENGTH'] = idline.shift.strip.split(' ').first.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line104">104</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line105">105</a>       @data['ID'] = tmp</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line106">106</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line107">107</a>     </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line108">108</a>     if key</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line109">109</a>       @data['ID'][key]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line110">110</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line111">111</a>       @data['ID']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line112">112</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line113">113</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line114">114</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line115">115</a>   # returns ENTRY_NAME in the ID line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line116">116</a>   # * Bio::EMBL#entry -&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line117">117</a>   def entry</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line118">118</a>     id_line('ENTRY_NAME')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line119">119</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line120">120</a>   alias entry_name entry</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line121">121</a>   alias entry_id entry</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line122">122</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line123">123</a>   # returns MOLECULE_TYPE in the ID line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line124">124</a>   # * Bio::EMBL#molecule -&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line125">125</a>   def molecule</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line126">126</a>     id_line('MOLECULE_TYPE')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line127">127</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line128">128</a>   alias molecule_type molecule</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line129">129</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line130">130</a>   def data_class</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line131">131</a>     id_line('DATA_CLASS')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line132">132</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line133">133</a>   </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line134">134</a>   def topology</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line135">135</a>     id_line('TOPOLOGY')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line136">136</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line137">137</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line138">138</a>   # returns DIVISION in the ID line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line139">139</a>   # * Bio::EMBL#division -&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line140">140</a>   def division</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line141">141</a>     id_line('DIVISION')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line142">142</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line143">143</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line144">144</a>   # returns SEQUENCE_LENGTH in the ID line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line145">145</a>   # * Bio::EMBL#sequencelength -&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line146">146</a>   def sequence_length</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line147">147</a>     id_line('SEQUENCE_LENGTH')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line148">148</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line149">149</a>   alias seqlen sequence_length</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line150">150</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line151">151</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line152">152</a>   # AC Line</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line153">153</a>   # &quot;AC   A12345; B23456;&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line154">154</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line155">155</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line156">156</a>   # returns the version information in the sequence version (SV) line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line157">157</a>   # * Bio::EMBL#sv -&gt; Accession.Version in String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line158">158</a>   # * Bio::EMBL#version -&gt; accession in Int</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line159">159</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line160">160</a>   # SV Line; sequence version (1/entry)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line161">161</a>   #  SV    Accession.Version</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line162">162</a>   def sv</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line163">163</a>     if (v = field_fetch('SV').sub(/;/,'')) == &quot;&quot;</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line164">164</a>       [id_line['ENTRY_NAME'], id_line['SEQUENCE_VERSION']].join('.') </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line165">165</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line166">166</a>       v</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line167">167</a>     end  </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line168">168</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line169">169</a>   def version</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line170">170</a>     (sv.split(&quot;.&quot;)[1] || id_line['SEQUENCE_VERSION']).to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line171">171</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line172">172</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line173">173</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line174">174</a>   # returns contents in the date (DT) line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line175">175</a>   # * Bio::EMBL#dt  -&gt; &lt;DT Hash&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line176">176</a>   # where &lt;DT Hash&gt; is:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line177">177</a>   #  {}</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line178">178</a>   # * Bio::EMBL#dt(key)  -&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line179">179</a>   # keys: 'created' and 'updated'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line180">180</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line181">181</a>   # DT Line; date (2/entry)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line182">182</a>   def dt(key=nil)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line183">183</a>     unless @data['DT']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line184">184</a>       tmp = Hash.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line185">185</a>       dt_line = self.get('DT').split(/\n/)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line186">186</a>       tmp['created'] = dt_line[0].sub(/\w{2}   /,'').strip</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line187">187</a>       tmp['updated'] = dt_line[1].sub(/\w{2}   /,'').strip</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line188">188</a>       @data['DT'] = tmp</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line189">189</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line190">190</a>     if key</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line191">191</a>       @data['DT'][key]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line192">192</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line193">193</a>       @data['DT']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line194">194</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line195">195</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line196">196</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line197">197</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line198">198</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line199">199</a>   ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line200">200</a>   # DE Line; description (&gt;=1)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line201">201</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line202">202</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line203">203</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line204">204</a>   ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line205">205</a>   # KW Line; keyword (&gt;=1)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line206">206</a>   # KW   [Keyword;]+</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line207">207</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line208">208</a>   # Bio::EMBLDB#kw  -&gt; Array</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line209">209</a>   #            #keywords  -&gt; Array</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line210">210</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line211">211</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line212">212</a>   ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line213">213</a>   # OS Line; organism species (&gt;=1)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line214">214</a>   # OS   Genus species (name)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line215">215</a>   # &quot;OS   Trifolium repens (white clover)&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line216">216</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line217">217</a>   # Bio::EMBLDB#os  -&gt; Array</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line218">218</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line219">219</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line220">220</a>   ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line221">221</a>   # OC Line; organism classification (&gt;=1)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line222">222</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line223">223</a>   # Bio::EMBLDB#oc  -&gt; Array</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line224">224</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line225">225</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line226">226</a>   ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line227">227</a>   # OG Line; organella (0 or 1/entry)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line228">228</a>   # [&quot;Mitochondrion&quot;, &quot;Chloroplast&quot;,&quot;Kinetoplast&quot;, &quot;Cyanelle&quot;, &quot;Plastid&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line229">229</a>   #  or a plasmid name (e.g. &quot;Plasmid pBR322&quot;).  </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line230">230</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line231">231</a>   # Bio::EMBLDB#og  -&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line232">232</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line233">233</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line234">234</a>   ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line235">235</a>   # R Lines</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line236">236</a>   # RN RC RP RX RA RT RL</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line237">237</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line238">238</a>   # Bio::EMBLDB#ref</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line239">239</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line240">240</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line241">241</a>   ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line242">242</a>   # DR Line; defabases cross-regerence (&gt;=0)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line243">243</a>   # &quot;DR  database_identifier; primary_identifier; secondary_identifier.&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line244">244</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line245">245</a>   # Bio::EMBLDB#dr</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line246">246</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line247">247</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line248">248</a>   # returns feature table header (String) in the feature header (FH) line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line249">249</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line250">250</a>   # FH Line; feature table header (0 or 2)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line251">251</a>   def fh</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line252">252</a>     fetch('FH')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line253">253</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line254">254</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line255">255</a>   # returns contents in the feature table (FT) lines.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line256">256</a>   # * Bio::EMBL#ft -&gt; Bio::Features</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line257">257</a>   # * Bio::EMBL#ft {} -&gt; {|Bio::Feature| }</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line258">258</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line259">259</a>   # same as features method in bio/db/genbank.rb </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line260">260</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line261">261</a>   # FT Line; feature table data (&gt;=0)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line262">262</a>   def ft</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line263">263</a>     unless @data['FT']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line264">264</a>       ary = Array.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line265">265</a>       in_quote = false</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line266">266</a>       @orig['FT'].each_line do |line|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line267">267</a>         next if line =~ /^FEATURES/</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line268">268</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line269">269</a>         head = line[0,20].strip  # feature key (source, CDS, ...)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line270">270</a>         body = line[20,60].chomp # feature value (position, /qualifier=)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line271">271</a>         if line =~ /^FT {3}(\S+)/</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line272">272</a>           ary.push([ $1, body ]) # [ feature, position, /q=&quot;data&quot;, ... ]</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line273">273</a>         elsif body =~ /^ \// and not in_quote</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line274">274</a>           ary.last.push(body)    # /q=&quot;data..., /q=data, /q</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line275">275</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line276">276</a>           if body =~ /=&quot; / and body !~ /&quot;$/</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line277">277</a>             in_quote = true</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line278">278</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line279">279</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line280">280</a>         else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line281">281</a>           ary.last.last &lt;&lt; body # ...data..., ...data...&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line282">282</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line283">283</a>           if body =~ /&quot;$/</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line284">284</a>             in_quote = false</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line285">285</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line286">286</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line287">287</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line288">288</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line289">289</a>       ary.map! do |subary|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line290">290</a>         parse_qualifiers(subary)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line291">291</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line292">292</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line293">293</a>       @data['FT'] = ary.extend(Bio::Features::BackwardCompatibility)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line294">294</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line295">295</a>     if block_given?</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line296">296</a>       @data['FT'].each do |feature|</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line297">297</a>         yield feature</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line298">298</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line299">299</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line300">300</a>       @data['FT']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line301">301</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line302">302</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line303">303</a>   alias features ft</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line304">304</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line305">305</a>   # iterates on CDS features in the FT lines.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line306">306</a>   def each_cds</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line307">307</a>     ft.each do |cds_feature|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line308">308</a>       if cds_feature.feature == 'CDS'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line309">309</a>         yield cds_feature</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line310">310</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line311">311</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line312">312</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line313">313</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line314">314</a>   # iterates on gene features in the FT lines.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line315">315</a>   def each_gene</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line316">316</a>     ft.each do |gene_feature|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line317">317</a>       if gene_feature.feature == 'gene'</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line318">318</a>         yield gene_feature</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line319">319</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line320">320</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line321">321</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line322">322</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line323">323</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line324">324</a>   # returns comment text in the comments (CC) line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line325">325</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line326">326</a>   # CC Line; comments of notes (&gt;=0)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line327">327</a>   def cc</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line328">328</a>     get('CC').to_s.gsub(/^CC   /, '')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line329">329</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line330">330</a>   alias comment cc</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line331">331</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line332">332</a>   ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line333">333</a>   # XX Line; spacer line (many)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line334">334</a>   #  def nxx</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line335">335</a>   #  end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line336">336</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line337">337</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line338">338</a>   # returns sequence header information in the sequence header (SQ) line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line339">339</a>   # * Bio::EMBL#sq  -&gt; &lt;SQ Hash&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line340">340</a>   # where &lt;SQ Hash&gt; is:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line341">341</a>   #     {'ntlen' =&gt; Int, 'other' =&gt; Int,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line342">342</a>   #      'a' =&gt; Int, 'c' =&gt; Int, 'g' =&gt; Int, 't' =&gt; Int}</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line343">343</a>   # * Bio::EMBL#sq(base)  -&gt; &lt;base content in Int&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line344">344</a>   # * Bio::EMBL#sq[base]  -&gt; &lt;base content in Int&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line345">345</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line346">346</a>   # SQ Line; sequence header (1/entry)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line347">347</a>   #  SQ   Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line348">348</a>   def sq(base = nil)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line349">349</a>     unless @data['SQ']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line350">350</a>       fetch('SQ') =~ \</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line351">351</a>              /(\d+) BP\; (\d+) A; (\d+) C; (\d+) G; (\d+) T; (\d+) other;/</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line352">352</a>       @data['SQ'] = {'ntlen' =&gt; $1.to_i, 'other' =&gt; $6.to_i,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line353">353</a>                      'a' =&gt; $2.to_i, 'c' =&gt; $3.to_i , 'g' =&gt; $4.to_i, 't' =&gt; $5.to_i}</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line354">354</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line355">355</a>       @data['SQ']</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line356">356</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line357">357</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line358">358</a>     if base</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line359">359</a>       @data['SQ'][base.downcase]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line360">360</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line361">361</a>       @data['SQ']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line362">362</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line363">363</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line364">364</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line365">365</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line366">366</a>   # returns the nucleotie sequence in this entry.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line367">367</a>   # * Bio::EMBL#seq  -&gt; Bio::Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line368">368</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line369">369</a>   # @orig[''] as sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line370">370</a>   # bb Line; (blanks) sequence data (&gt;=1)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line371">371</a>   def seq</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line372">372</a>     Bio::Sequence::NA.new( fetch('').gsub(/ /,'').gsub(/\d+/,'') )</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line373">373</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line374">374</a>   alias naseq seq</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line375">375</a>   alias ntseq seq</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line376">376</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line377">377</a>   #--</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line378">378</a>   # // Line; termination line (end; 1/entry)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line379">379</a>   #++</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line380">380</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line381">381</a>   # modified date. Returns Date object, String or nil.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line382">382</a>   def date_modified</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line383">383</a>     parse_date(self.dt['updated'])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line384">384</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line385">385</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line386">386</a>   # created date. Returns Date object, String or nil.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line387">387</a>   def date_created</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line388">388</a>     parse_date(self.dt['created'])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line389">389</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line390">390</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line391">391</a>   # release number when last updated</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line392">392</a>   def release_modified</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line393">393</a>     parse_release_version(self.dt['updated'])[0]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line394">394</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line395">395</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line396">396</a>   # release number when created</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line397">397</a>   def release_created</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line398">398</a>     parse_release_version(self.dt['created'])[0]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line399">399</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line400">400</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line401">401</a>   # entry version number numbered by EMBL</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line402">402</a>   def entry_version</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line403">403</a>     parse_release_version(self.dt['updated'])[1]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line404">404</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line405">405</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line406">406</a>   # parse date string. Returns Date object.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line407">407</a>   def parse_date(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line408">408</a>     begin</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line409">409</a>       Date.parse(str)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line410">410</a>     rescue ArgumentError, TypeError, NoMethodError, NameError</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line411">411</a>       str</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line412">412</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line413">413</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line414">414</a>   private :parse_date</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line415">415</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line416">416</a>   # extracts release and version numbers from DT line</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line417">417</a>   def parse_release_version(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line418">418</a>     return [ nil, nil ] unless str</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line419">419</a>     a = str.split(/[\(\,\)]/)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line420">420</a>     dstr = a.shift</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line421">421</a>     rel = nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line422">422</a>     ver = nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line423">423</a>     a.each do |x|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line424">424</a>       case x</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line425">425</a>       when /Rel\.\s*(.+)/</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line426">426</a>         rel = $1.strip</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line427">427</a>       when /Version\s*(.+)/</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line428">428</a>         ver = $1.strip</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line429">429</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line430">430</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line431">431</a>     [ rel, ver ]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line432">432</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line433">433</a>   private :parse_release_version</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line434">434</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line435">435</a>   # database references (DR).</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line436">436</a>   # Returns an array of Bio::Sequence::DBLink objects.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line437">437</a>   def dblinks</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line438">438</a>     get('DR').split(/\n/).collect { |x|</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line439">439</a>       Bio::Sequence::DBLink.parse_embl_DR_line(x)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line440">440</a>     }</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line441">441</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line442">442</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line443">443</a>   # species</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line444">444</a>   def species</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line445">445</a>     self.fetch('OS')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line446">446</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line447">447</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line448">448</a>   # taxonomy classfication</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line449">449</a>   alias classification oc</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line450">450</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line451">451</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line452">452</a>   # converts the entry to Bio::Sequence object</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line453">453</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line454">454</a>   # *Arguments*::</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line455">455</a>   # *Returns*:: Bio::Sequence object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line456">456</a>   def to_biosequence</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line457">457</a>     Bio::Sequence.adapter(self, Bio::Sequence::Adapter::EMBL)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line458">458</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line459">459</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line460">460</a>   ### private methods</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line461">461</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line462">462</a>   private</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line463">463</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line464">464</a>   ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line465">465</a>   # same as Bio::GenBank#parse_qualifiers(feature)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line466">466</a>   def parse_qualifiers(ary)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line467">467</a>     feature = Feature.new</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line468">468</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line469">469</a>     feature.feature = ary.shift</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line470">470</a>     feature.position = ary.shift.gsub(/\s/, '')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line471">471</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line472">472</a>     ary.each do |f|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line473">473</a>       if f =~ %r{/([^=]+)=?&quot;?([^&quot;]*)&quot;?}</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line474">474</a>         qualifier, value = $1, $2</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line475">475</a>         </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line476">476</a>         if value.empty?</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line477">477</a>           value = true</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line478">478</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line479">479</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line480">480</a>         case qualifier</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line481">481</a>         when 'translation'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line482">482</a>           value = Sequence::AA.new(value.gsub(/\s/, ''))</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line483">483</a>         when 'codon_start'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line484">484</a>           value = value.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line485">485</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line486">486</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line487">487</a>         feature.append(Feature::Qualifier.new(qualifier, value))</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line488">488</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line489">489</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line490">490</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line491">491</a>     return feature</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line492">492</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line493">493</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line494">494</a> end # class EMBL</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line495">495</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line496">496</a> end # module Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line497">497</a> </pre></td>
          </tr>
        
      </tbody>
    </table>

    <p>Generated on Fri Jun 25 17:37:22 +0900 2010 with <a href="http://github.com/relevance/rcov">rcov 0.9.8</a></p>

  </body>
</html>
