<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang='en' xml:lang='en' xmlns='http://www.w3.org/1999/xhtml'>
  <head>
    <title>lib/bio/db/kegg/genes.rb</title>
    <link href="screen.css" media="all" rel="stylesheet" type="text/css" />
    <link href="print.css" media="print" rel="stylesheet" type="text/css" />
    
    <meta http-equiv="Content-type" content="text/html;charset=UTF-8" />
    <script type="text/javascript" src="rcov.js"></script>
  </head>
  <body>
    <h1>Bioruby C0 Coverage Information - RCov</h1>
    <h2>lib/bio/db/kegg/genes.rb</h2>

    

    <div class="report_table_wrapper">
      <table class='report' id='report_table'>
        <thead>
          <tr>
            <th class="left_align">Name</th>
            <th class="right_align">Total Lines</th>
            <th class="right_align">Lines of Code</th>
            <th class="left_align">Total Coverage</th>
            <th class="left_align">Code Coverage</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td class="left_align"><a href="lib-bio-db-kegg-genes_rb.html">lib/bio/db/kegg/genes.rb</a></td>
            <td class='right_align'><tt>381</tt></td>
            <td class='right_align'><tt>173</tt></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>97.64%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:98px"></div>
            <div class="uncovered" style="width:2px"></div>
          </div></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>94.80%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:95px"></div>
            <div class="uncovered" style="width:5px"></div>
          </div></td>
          </tr>
        </tbody>
      </table>
    </div>
    
    <h3>Key</h3>
    
    <div class="key"><pre><span class='marked'>Code reported as executed by Ruby looks like this...</span><span class='marked1'>and this: this line is also marked as covered.</span><span class='inferred'>Lines considered as run by rcov, but not reported by Ruby, look like this,</span><span class='inferred1'>and this: these lines were inferred by rcov (using simple heuristics).</span><span class='uncovered'>Finally, here's a line marked as not executed.</span></pre></div>

    <h3>Coverage Details</h3>

    <table class="details">
      <tbody>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1">1</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line2">2</a> # = bio/db/kegg/genes.rb - KEGG/GENES database class</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line3">3</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line4">4</a> # Copyright::   Copyright (C) 2001, 2002, 2006</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line5">5</a> #               Toshiaki Katayama &lt;k@bioruby.org&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line6">6</a> # License::     The Ruby License</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line7">7</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line8">8</a> # $Id:$</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line9">9</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line10">10</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line11">11</a> # == KEGG GENES parser</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line12">12</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line13">13</a> # See http://www.genome.jp/kegg/genes.html</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line14">14</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line15">15</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line16">16</a> # === Examples</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line17">17</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line18">18</a> #  require 'bio/io/fetch'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line19">19</a> #  entry_string = Bio::Fetch.query('genes', 'b0002')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line20">20</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line21">21</a> #  entry = Bio::KEGG::GENES.new(entry_string)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line22">22</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line23">23</a> #  # ENTRY</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line24">24</a> #  p entry.entry       # =&gt; Hash</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line25">25</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line26">26</a> #  p entry.entry_id    # =&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line27">27</a> #  p entry.division    # =&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line28">28</a> #  p entry.organism    # =&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line29">29</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line30">30</a> #  # NAME</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line31">31</a> #  p entry.name        # =&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line32">32</a> #  p entry.genes       # =&gt; Array</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line33">33</a> #  p entry.gene        # =&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line34">34</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line35">35</a> #  # DEFINITION</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line36">36</a> #  p entry.definition  # =&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line37">37</a> #  p entry.eclinks     # =&gt; Array</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line38">38</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line39">39</a> #  # PATHWAY</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line40">40</a> #  p entry.pathway     # =&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line41">41</a> #  p entry.pathways    # =&gt; Hash</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line42">42</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line43">43</a> #  # POSITION</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line44">44</a> #  p entry.position    # =&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line45">45</a> #  p entry.chromosome  # =&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line46">46</a> #  p entry.gbposition  # =&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line47">47</a> #  p entry.locations   # =&gt; Bio::Locations</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line48">48</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line49">49</a> #  # MOTIF</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line50">50</a> #  p entry.motif       # =&gt; Hash of Array</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line51">51</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line52">52</a> #  # DBLINKS</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line53">53</a> #  p entry.dblinks     # =&gt; Hash of Array</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line54">54</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line55">55</a> #  # STRUCTURE</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line56">56</a> #  p entry.structure   # =&gt; Array</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line57">57</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line58">58</a> #  # CODON_USAGE</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line59">59</a> #  p entry.codon_usage # =&gt; Hash</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line60">60</a> #  p entry.cu_list     # =&gt; Array</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line61">61</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line62">62</a> #  # AASEQ</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line63">63</a> #  p entry.aaseq       # =&gt; Bio::Sequence::AA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line64">64</a> #  p entry.aalen       # =&gt; Fixnum</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line65">65</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line66">66</a> #  # NTSEQ</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line67">67</a> #  p entry.ntseq       # =&gt; Bio::Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line68">68</a> #  p entry.naseq       # =&gt; Bio::Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line69">69</a> #  p entry.ntlen       # =&gt; Fixnum</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line70">70</a> #  p entry.nalen       # =&gt; Fixnum</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line71">71</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line72">72</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line73">73</a> module Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line74">74</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line75">75</a>   autoload :KEGGDB,    'bio/db'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line76">76</a>   autoload :Locations, 'bio/location'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line77">77</a>   autoload :Sequence,  'bio/sequence'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line78">78</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line79">79</a>   require 'bio/db/kegg/common'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line80">80</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line81">81</a> class KEGG</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line82">82</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line83">83</a> # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line84">84</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line85">85</a> # KEGG GENES entry parser.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line86">86</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line87">87</a> # == References</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line88">88</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line89">89</a> # * http://www.genome.jp/kegg/genes.html</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line90">90</a> #</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line91">91</a> class GENES &lt; KEGGDB</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line92">92</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line93">93</a>   DELIMITER	= RS = &quot;\n///\n&quot;</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line94">94</a>   TAGSIZE	= 12</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line95">95</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line96">96</a>   include Common::DblinksAsHash</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line97">97</a>   # Returns a Hash of the DB name and an Array of entry IDs in DBLINKS field.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line98">98</a>   def dblinks_as_hash; super; end if false #dummy for RDoc</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line99">99</a>   alias dblinks dblinks_as_hash</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line100">100</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line101">101</a>   include Common::PathwaysAsHash</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line102">102</a>   # Returns a Hash of the pathway ID and name in PATHWAY field.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line103">103</a>   def pathways_as_hash; super; end if false #dummy for RDoc</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line104">104</a>   alias pathways pathways_as_hash</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line105">105</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line106">106</a>   include Common::OrthologsAsHash</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line107">107</a>   # Returns a Hash of the orthology ID and definition in ORTHOLOGY field.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line108">108</a>   def orthologs_as_hash; super; end if false #dummy for RDoc</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line109">109</a>   alias orthologs orthologs_as_hash</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line110">110</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line111">111</a>   # Creates a new Bio::KEGG::GENES object.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line112">112</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line113">113</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line114">114</a>   # * (required) _entry_: (String) single entry as a string</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line115">115</a>   # *Returns*:: Bio::KEGG::GENES object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line116">116</a>   def initialize(entry)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line117">117</a>     super(entry, TAGSIZE)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line118">118</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line119">119</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line120">120</a>   # Returns the &quot;ENTRY&quot; line content as a Hash.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line121">121</a>   # For example, </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line122">122</a>   #   {&quot;organism&quot;=&gt;&quot;E.coli&quot;, &quot;division&quot;=&gt;&quot;CDS&quot;, &quot;id&quot;=&gt;&quot;b0356&quot;}</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line123">123</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line124">124</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line125">125</a>   # *Returns*:: Hash</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line126">126</a>   def entry</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line127">127</a>     unless @data['ENTRY']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line128">128</a>       hash = Hash.new('')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line129">129</a>       if get('ENTRY').length &gt; 30</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line130">130</a>         e = get('ENTRY')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line131">131</a>         hash['id']       = e[12..29].strip</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line132">132</a>         hash['division'] = e[30..39].strip</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line133">133</a>         hash['organism'] = e[40..80].strip</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line134">134</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line135">135</a>       @data['ENTRY'] = hash</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line136">136</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line137">137</a>     @data['ENTRY']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line138">138</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line139">139</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line140">140</a>   # ID of the entry, described in the ENTRY line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line141">141</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line142">142</a>   # *Returns*:: String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line143">143</a>   def entry_id</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line144">144</a>     entry['id']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line145">145</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line146">146</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line147">147</a>   # Division of the entry, described in the ENTRY line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line148">148</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line149">149</a>   # *Returns*:: String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line150">150</a>   def division</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line151">151</a>     entry['division']			# CDS, tRNA etc.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line152">152</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line153">153</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line154">154</a>   # Organism name of the entry, described in the ENTRY line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line155">155</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line156">156</a>   # *Returns*:: String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line157">157</a>   def organism</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line158">158</a>     entry['organism']			# H.sapiens etc.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line159">159</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line160">160</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line161">161</a>   # Returns the NAME line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line162">162</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line163">163</a>   # *Returns*:: String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line164">164</a>   def name</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line165">165</a>     field_fetch('NAME')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line166">166</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line167">167</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line168">168</a>   # Names of the entry as an Array, described in the NAME line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line169">169</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line170">170</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line171">171</a>   # *Returns*:: Array containing String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line172">172</a>   def genes</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line173">173</a>     name.split(', ')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line174">174</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line175">175</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line176">176</a>   # Returns the first gene name described in the NAME line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line177">177</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line178">178</a>   # *Returns*:: String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line179">179</a>   def gene</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line180">180</a>     genes.first</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line181">181</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line182">182</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line183">183</a>   # Definition of the entry, described in the DEFINITION line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line184">184</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line185">185</a>   # *Returns*:: String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line186">186</a>   def definition</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line187">187</a>     field_fetch('DEFINITION')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line188">188</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line189">189</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line190">190</a>   # Enzyme's EC numbers shown in the DEFINITION line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line191">191</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line192">192</a>   # *Returns*:: Array containing String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line193">193</a>   def eclinks</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line194">194</a>     unless defined? @eclinks</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line195">195</a>       ec_list = </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line196">196</a>         definition.slice(/\[EC\:([^\]]+)\]/, 1) ||</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line197">197</a>         definition.slice(/\(EC\:([^\)]+)\)/, 1)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line198">198</a>       ary = ec_list ? ec_list.strip.split(/\s+/) : []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line199">199</a>       @eclinks = ary</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line200">200</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line201">201</a>     @eclinks</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line202">202</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line203">203</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line204">204</a>   # Orthologs described in the ORTHOLOGY lines.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line205">205</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line206">206</a>   # *Returns*:: Array containing String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line207">207</a>   def orthologs_as_strings</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line208">208</a>     lines_fetch('ORTHOLOGY')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line209">209</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line210">210</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line211">211</a>   # Returns the PATHWAY lines as a String.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line212">212</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line213">213</a>   # *Returns*:: String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line214">214</a>   def pathway</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line215">215</a>     unless defined? @pathway</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line216">216</a>       @pathway = fetch('PATHWAY')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line217">217</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line218">218</a>     @pathway</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line219">219</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line220">220</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line221">221</a>   # Pathways described in the PATHWAY lines.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line222">222</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line223">223</a>   # *Returns*:: Array containing String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line224">224</a>   def pathways_as_strings</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line225">225</a>     lines_fetch('PATHWAY')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line226">226</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line227">227</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line228">228</a>   # The position in the genome described in the POSITION line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line229">229</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line230">230</a>   # *Returns*:: String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line231">231</a>   def position</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line232">232</a>     unless @data['POSITION']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line233">233</a>       @data['POSITION'] = fetch('POSITION').gsub(/\s/, '')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line234">234</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line235">235</a>     @data['POSITION']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line236">236</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line237">237</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line238">238</a>   # Chromosome described in the POSITION line.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line239">239</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line240">240</a>   # *Returns*:: String or nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line241">241</a>   def chromosome</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line242">242</a>     if position[/:/]</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line243">243</a>       position.sub(/:.*/, '')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line244">244</a>     elsif ! position[/\.\./]</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line245">245</a>       position</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line246">246</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line247">247</a>       nil</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line248">248</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line249">249</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line250">250</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line251">251</a>   # The position in the genome described in the POSITION line</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line252">252</a>   # as GenBank feature table location formatted string.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line253">253</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line254">254</a>   # *Returns*:: String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line255">255</a>   def gbposition</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line256">256</a>     position.sub(/.*?:/, '')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line257">257</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line258">258</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line259">259</a>   # The position in the genome described in the POSITION line</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line260">260</a>   # as Bio::Locations object.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line261">261</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line262">262</a>   # *Returns*:: Bio::Locations object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line263">263</a>   def locations</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line264">264</a>     Bio::Locations.new(gbposition)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line265">265</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line266">266</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line267">267</a>   # Motif information described in the MOTIF lines.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line268">268</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line269">269</a>   # *Returns*:: Hash</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line270">270</a>   def motif</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line271">271</a>     unless @data['MOTIF']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line272">272</a>       hash = {}</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line273">273</a>       db = nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line274">274</a>       lines_fetch('MOTIF').each do |line|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line275">275</a>         if line[/^\S+:/]</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line276">276</a>           db, str = line.split(/:/)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line277">277</a>         else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line278">278</a>           str = line</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line279">279</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line280">280</a>         hash[db] ||= []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line281">281</a>         hash[db] += str.strip.split(/\s+/)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line282">282</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line283">283</a>       @data['MOTIF'] = hash</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line284">284</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line285">285</a>     @data['MOTIF']		# Hash of Array of IDs in MOTIF</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line286">286</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line287">287</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line288">288</a>   # Links to other databases described in the DBLINKS lines.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line289">289</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line290">290</a>   # *Returns*:: Array containing String objects</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line291">291</a>   def dblinks_as_strings</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line292">292</a>     lines_fetch('DBLINKS')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line293">293</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line294">294</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line295">295</a>   # Returns structure ID information described in the STRUCTURE lines.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line296">296</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line297">297</a>   # *Returns*:: Array containing String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line298">298</a>   def structure</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line299">299</a>     unless @data['STRUCTURE']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line300">300</a>       @data['STRUCTURE'] = fetch('STRUCTURE').sub(/(PDB: )*/,'').split(/\s+/)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line301">301</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line302">302</a>     @data['STRUCTURE'] # ['PDB:1A9X', ...]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line303">303</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line304">304</a>   alias structures structure</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line305">305</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line306">306</a>   # Codon usage data described in the CODON_USAGE lines.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line307">307</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line308">308</a>   # *Returns*:: Hash</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line309">309</a>   def codon_usage(codon = nil)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line310">310</a>     unless @data['CODON_USAGE']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line311">311</a>       hash = Hash.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line312">312</a>       list = cu_list</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line313">313</a>       base = %w(t c a g)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line314">314</a>       base.each_with_index do |x, i|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line315">315</a>         base.each_with_index do |y, j|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line316">316</a>           base.each_with_index do |z, k|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line317">317</a>             hash[&quot;#{x}#{y}#{z}&quot;] = list[i*16 + j*4 + k]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line318">318</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line319">319</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line320">320</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line321">321</a>       @data['CODON_USAGE'] = hash</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line322">322</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line323">323</a>     @data['CODON_USAGE']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line324">324</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line325">325</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line326">326</a>   # Codon usage data described in the CODON_USAGE lines as an array.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line327">327</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line328">328</a>   # *Returns*:: Array</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line329">329</a>   def cu_list</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line330">330</a>     ary = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line331">331</a>     get('CODON_USAGE').sub(/.*/,'').each_line do |line|	# cut 1st line</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line332">332</a>       line.chomp.sub(/^.{11}/, '').scan(/..../) do |cu|</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line333">333</a>         ary.push(cu.to_i)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line334">334</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line335">335</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line336">336</a>     return ary</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line337">337</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line338">338</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line339">339</a>   # Returns amino acid sequence described in the AASEQ lines.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line340">340</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line341">341</a>   # *Returns*:: Bio::Sequence::AA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line342">342</a>   def aaseq</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line343">343</a>     unless @data['AASEQ']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line344">344</a>       @data['AASEQ'] = Bio::Sequence::AA.new(fetch('AASEQ').gsub(/\d+/, ''))</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line345">345</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line346">346</a>     @data['AASEQ']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line347">347</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line348">348</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line349">349</a>   # Returns length of the amino acid sequence described in the AASEQ lines.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line350">350</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line351">351</a>   # *Returns*:: Integer</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line352">352</a>   def aalen</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line353">353</a>     fetch('AASEQ')[/\d+/].to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line354">354</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line355">355</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line356">356</a>   # Returns nucleic acid sequence described in the NTSEQ lines.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line357">357</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line358">358</a>   # *Returns*:: Bio::Sequence::NA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line359">359</a>   def ntseq</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line360">360</a>     unless @data['NTSEQ']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line361">361</a>       @data['NTSEQ'] = Bio::Sequence::NA.new(fetch('NTSEQ').gsub(/\d+/, ''))</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line362">362</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line363">363</a>     @data['NTSEQ']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line364">364</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line365">365</a>   alias naseq ntseq</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line366">366</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line367">367</a>   # Returns nucleic acid sequence length.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line368">368</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line369">369</a>   # *Returns*:: Integer</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line370">370</a>   def ntlen</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line371">371</a>     fetch('NTSEQ')[/\d+/].to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line372">372</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line373">373</a>   alias nalen ntlen</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line374">374</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line375">375</a> end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line376">376</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line377">377</a> end # KEGG</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line378">378</a> end # Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line379">379</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line380">380</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line381">381</a> </pre></td>
          </tr>
        
      </tbody>
    </table>

    <p>Generated on Fri Jun 25 17:37:22 +0900 2010 with <a href="http://github.com/relevance/rcov">rcov 0.9.8</a></p>

  </body>
</html>
