<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang='en' xml:lang='en' xmlns='http://www.w3.org/1999/xhtml'>
  <head>
    <title>lib/bio/sequence/common.rb</title>
    <link href="screen.css" media="all" rel="stylesheet" type="text/css" />
    <link href="print.css" media="print" rel="stylesheet" type="text/css" />
    
    <meta http-equiv="Content-type" content="text/html;charset=UTF-8" />
    <script type="text/javascript" src="rcov.js"></script>
  </head>
  <body>
    <h1>Bioruby C0 Coverage Information - RCov</h1>
    <h2>lib/bio/sequence/common.rb</h2>

    

    <div class="report_table_wrapper">
      <table class='report' id='report_table'>
        <thead>
          <tr>
            <th class="left_align">Name</th>
            <th class="right_align">Total Lines</th>
            <th class="right_align">Lines of Code</th>
            <th class="left_align">Total Coverage</th>
            <th class="left_align">Code Coverage</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td class="left_align"><a href="lib-bio-sequence-common_rb.html">lib/bio/sequence/common.rb</a></td>
            <td class='right_align'><tt>310</tt></td>
            <td class='right_align'><tt>104</tt></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>77.42%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:77px"></div>
            <div class="uncovered" style="width:23px"></div>
          </div></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>33.65%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:34px"></div>
            <div class="uncovered" style="width:66px"></div>
          </div></td>
          </tr>
        </tbody>
      </table>
    </div>
    
    <h3>Key</h3>
    
    <div class="key"><pre><span class='marked'>Code reported as executed by Ruby looks like this...</span><span class='marked1'>and this: this line is also marked as covered.</span><span class='inferred'>Lines considered as run by rcov, but not reported by Ruby, look like this,</span><span class='inferred1'>and this: these lines were inferred by rcov (using simple heuristics).</span><span class='uncovered'>Finally, here's a line marked as not executed.</span></pre></div>

    <h3>Coverage Details</h3>

    <table class="details">
      <tbody>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1">1</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line2">2</a> # = bio/sequence/common.rb - common methods for biological sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line3">3</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line4">4</a> # Copyright::   Copyright (C) 2006</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line5">5</a> #               Toshiaki Katayama &lt;k@bioruby.org&gt;,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line6">6</a> #               Ryan Raaum &lt;ryan@raaum.org&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line7">7</a> # License::     The Ruby License</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line8">8</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line9">9</a> # $Id:$</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line10">10</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line11">11</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line12">12</a> module Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line13">13</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line14">14</a>   autoload :Locations, 'bio/location'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line15">15</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line16">16</a> class Sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line17">17</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line18">18</a> # = DESCRIPTION</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line19">19</a> # Bio::Sequence::Common is a </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line20">20</a> # Mixin[http://www.rubycentral.com/book/tut_modules.html]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line21">21</a> # implementing methods common to</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line22">22</a> # Bio::Sequence::AA and Bio::Sequence::NA.  All of these methods</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line23">23</a> # are available to either Amino Acid or Nucleic Acid sequences, and</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line24">24</a> # by encapsulation are also available to Bio::Sequence objects.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line25">25</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line26">26</a> # = USAGE</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line27">27</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line28">28</a> #   # Create a sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line29">29</a> #   dna = Bio::Sequence.auto('atgcatgcatgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line30">30</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line31">31</a> #   # Splice out a subsequence using a Genbank-style location string</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line32">32</a> #   puts dna.splice('complement(1..4)')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line33">33</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line34">34</a> #   # What is the base composition?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line35">35</a> #   puts dna.composition</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line36">36</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line37">37</a> #   # Create a random sequence with the composition of a current sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line38">38</a> #   puts dna.randomize</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line39">39</a> module Common</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line40">40</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line41">41</a>   # Return sequence as </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line42">42</a>   # String[http://corelib.rubyonrails.org/classes/String.html].</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line43">43</a>   # The original sequence is unchanged.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line44">44</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line45">45</a>   #   seq = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line46">46</a>   #   puts s.to_s                             #=&gt; 'atgc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line47">47</a>   #   puts s.to_s.class                       #=&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line48">48</a>   #   puts s                                  #=&gt; 'atgc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line49">49</a>   #   puts s.class                            #=&gt; Bio::Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line50">50</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line51">51</a>   # *Returns*:: String object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line52">52</a>   def to_s</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line53">53</a>     String.new(self)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line54">54</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line55">55</a>   alias to_str to_s</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line56">56</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line57">57</a>   # Create a new sequence based on the current sequence.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line58">58</a>   # The original sequence is unchanged.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line59">59</a>   # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line60">60</a>   #   s = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line61">61</a>   #   s2 = s.seq</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line62">62</a>   #   puts s2                                 #=&gt; 'atgc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line63">63</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line64">64</a>   # *Returns*:: new Bio::Sequence::NA/AA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line65">65</a>   def seq</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line66">66</a>     self.class.new(self)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line67">67</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line68">68</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line69">69</a>   # Normalize the current sequence, removing all whitespace and </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line70">70</a>   # transforming all positions to uppercase if the sequence is AA or</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line71">71</a>   # transforming all positions to lowercase if the sequence is NA.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line72">72</a>   # The original sequence is modified.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line73">73</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line74">74</a>   #   s = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line75">75</a>   #   s.normalize!</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line76">76</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line77">77</a>   # *Returns*:: current Bio::Sequence::NA/AA object (modified)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line78">78</a>   def normalize!</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line79">79</a>     initialize(self)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line80">80</a>     self</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line81">81</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line82">82</a>   alias seq! normalize!</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line83">83</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line84">84</a>   # Add new data to the end of the current sequence.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line85">85</a>   # The original sequence is modified.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line86">86</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line87">87</a>   #   s = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line88">88</a>   #   s &lt;&lt; 'atgc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line89">89</a>   #   puts s                                  #=&gt; &quot;atgcatgc&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line90">90</a>   #   s &lt;&lt; s</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line91">91</a>   #   puts s                                  #=&gt; &quot;atgcatgcatgcatgc&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line92">92</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line93">93</a>   # *Returns*:: current Bio::Sequence::NA/AA object (modified)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line94">94</a>   def concat(*arg)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line95">95</a>     super(self.class.new(*arg))</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line96">96</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line97">97</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line98">98</a>   def &lt;&lt;(*arg)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line99">99</a>     concat(*arg)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line100">100</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line101">101</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line102">102</a>   # Create a new sequence by adding to an existing sequence.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line103">103</a>   # The existing sequence is not modified.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line104">104</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line105">105</a>   #   s = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line106">106</a>   #   s2 = s + 'atgc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line107">107</a>   #   puts s2                                 #=&gt; &quot;atgcatgc&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line108">108</a>   #   puts s                                  #=&gt; &quot;atgc&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line109">109</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line110">110</a>   # The new sequence is of the same class as the existing sequence if </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line111">111</a>   # the new data was added to an existing sequence,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line112">112</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line113">113</a>   #   puts s2.class == s.class                #=&gt; true</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line114">114</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line115">115</a>   # but if an existing sequence is added to a String, the result is a String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line116">116</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line117">117</a>   #   s3 = 'atgc' + s</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line118">118</a>   #   puts s3.class                           #=&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line119">119</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line120">120</a>   # *Returns*:: new Bio::Sequence::NA/AA *or* String object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line121">121</a>   def +(*arg)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line122">122</a>     self.class.new(super(*arg))</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line123">123</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line124">124</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line125">125</a>   # Returns a new sequence containing the subsequence identified by the </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line126">126</a>   # start and end numbers given as parameters.  *Important:* Biological </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line127">127</a>   # sequence numbering conventions (one-based) rather than ruby's </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line128">128</a>   # (zero-based) numbering conventions are used.  </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line129">129</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line130">130</a>   #   s = Bio::Sequence::NA.new('atggaatga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line131">131</a>   #   puts s.subseq(1,3)                      #=&gt; &quot;atg&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line132">132</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line133">133</a>   # Start defaults to 1 and end defaults to the entire existing string, so</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line134">134</a>   # subseq called without any parameters simply returns a new sequence </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line135">135</a>   # identical to the existing sequence.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line136">136</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line137">137</a>   #   puts s.subseq                           #=&gt; &quot;atggaatga&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line138">138</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line139">139</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line140">140</a>   # * (optional) _s_(start): Integer (default 1)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line141">141</a>   # * (optional) _e_(end): Integer (default current sequence length)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line142">142</a>   # *Returns*:: new Bio::Sequence::NA/AA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line143">143</a>   def subseq(s = 1, e = self.length)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line144">144</a>     raise &quot;Error: start/end position must be a positive integer&quot; unless s &gt; 0 and e &gt; 0</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line145">145</a>     s -= 1</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line146">146</a>     e -= 1</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line147">147</a>     self[s..e]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line148">148</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line149">149</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line150">150</a>   # This method steps through a sequences in steps of 'step_size' by </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line151">151</a>   # subsequences of 'window_size'. Typically used with a block.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line152">152</a>   # Any remaining sequence at the terminal end will be returned.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line153">153</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line154">154</a>   # Prints average GC% on each 100bp</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line155">155</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line156">156</a>   #   s.window_search(100) do |subseq|</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line157">157</a>   #     puts subseq.gc</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line158">158</a>   #   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line159">159</a>   #   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line160">160</a>   # Prints every translated peptide (length 5aa) in the same frame</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line161">161</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line162">162</a>   #   s.window_search(15, 3) do |subseq|</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line163">163</a>   #     puts subseq.translate</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line164">164</a>   #   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line165">165</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line166">166</a>   # Split genome sequence by 10000bp with 1000bp overlap in fasta format</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line167">167</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line168">168</a>   #   i = 1</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line169">169</a>   #   remainder = s.window_search(10000, 9000) do |subseq|</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line170">170</a>   #     puts subseq.to_fasta(&quot;segment #{i}&quot;, 60)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line171">171</a>   #     i += 1</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line172">172</a>   #   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line173">173</a>   #   puts remainder.to_fasta(&quot;segment #{i}&quot;, 60)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line174">174</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line175">175</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line176">176</a>   # * (required) _window_size_: Fixnum</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line177">177</a>   # * (optional) _step_size_: Fixnum (default 1)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line178">178</a>   # *Returns*:: new Bio::Sequence::NA/AA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line179">179</a>   def window_search(window_size, step_size = 1)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line180">180</a>     last_step = 0</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line181">181</a>     0.step(self.length - window_size, step_size) do |i| </pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line182">182</a>       yield self[i, window_size]                        </pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line183">183</a>       last_step = i</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line184">184</a>     end                          </pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line185">185</a>     return self[last_step + window_size .. -1] </pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line186">186</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line187">187</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line188">188</a>   # Returns a float total value for the sequence given a hash of</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line189">189</a>   # base or residue values,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line190">190</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line191">191</a>   #   values = {'a' =&gt; 0.1, 't' =&gt; 0.2, 'g' =&gt; 0.3, 'c' =&gt; 0.4}</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line192">192</a>   #   s = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line193">193</a>   #   puts s.total(values)                    #=&gt; 1.0</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line194">194</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line195">195</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line196">196</a>   # * (required) _hash_: Hash object</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line197">197</a>   # *Returns*:: Float object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line198">198</a>   def total(hash)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line199">199</a>     hash.default = 0.0 unless hash.default</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line200">200</a>     sum = 0.0</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line201">201</a>     self.each_byte do |x|</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line202">202</a>       begin</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line203">203</a>         sum += hash[x.chr]</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line204">204</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line205">205</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line206">206</a>     return sum</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line207">207</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line208">208</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line209">209</a>   # Returns a hash of the occurrence counts for each residue or base.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line210">210</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line211">211</a>   #   s = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line212">212</a>   #   puts s.composition              #=&gt; {&quot;a&quot;=&gt;1, &quot;c&quot;=&gt;1, &quot;g&quot;=&gt;1, &quot;t&quot;=&gt;1}</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line213">213</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line214">214</a>   # *Returns*:: Hash object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line215">215</a>   def composition</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line216">216</a>     count = Hash.new(0)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line217">217</a>     self.scan(/./) do |x|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line218">218</a>       count[x] += 1</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line219">219</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line220">220</a>     return count</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line221">221</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line222">222</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line223">223</a>   # Returns a randomized sequence. The default is to retain the same </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line224">224</a>   # base/residue composition as the original.  If a hash of base/residue </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line225">225</a>   # counts is given, the new sequence will be based on that hash </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line226">226</a>   # composition.  If a block is given, each new randomly selected </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line227">227</a>   # position will be passed into the block.  In all cases, the</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line228">228</a>   # original sequence is not modified.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line229">229</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line230">230</a>   #   s = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line231">231</a>   #   puts s.randomize                        #=&gt; &quot;tcag&quot;  (for example)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line232">232</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line233">233</a>   #   new_composition = {'a' =&gt; 2, 't' =&gt; 2}</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line234">234</a>   #   puts s.randomize(new_composition)       #=&gt; &quot;ttaa&quot;  (for example)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line235">235</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line236">236</a>   #   count = 0</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line237">237</a>   #   s.randomize { |x| count += 1 }</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line238">238</a>   #   puts count                              #=&gt; 4</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line239">239</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line240">240</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line241">241</a>   # * (optional) _hash_: Hash object</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line242">242</a>   # *Returns*:: new Bio::Sequence::NA/AA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line243">243</a>   def randomize(hash = nil)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line244">244</a>     if hash</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line245">245</a>       tmp = ''</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line246">246</a>       hash.each {|k, v|</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line247">247</a>         tmp += k * v.to_i</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line248">248</a>       }</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line249">249</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line250">250</a>       tmp = self</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line251">251</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line252">252</a>     seq = self.class.new(tmp)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line253">253</a>     # Reference: http://en.wikipedia.org/wiki/Fisher-Yates_shuffle</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line254">254</a>     seq.length.downto(2) do |n|</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line255">255</a>       k = rand(n)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line256">256</a>       c = seq[n - 1]</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line257">257</a>       seq[n - 1] = seq[k]</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line258">258</a>       seq[k] = c</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line259">259</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line260">260</a>     if block_given? then</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line261">261</a>       (0...seq.length).each do |i|</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line262">262</a>         yield seq[i, 1]</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line263">263</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line264">264</a>       return self.class.new('')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line265">265</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line266">266</a>       return seq</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line267">267</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line268">268</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line269">269</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line270">270</a>   # Return a new sequence extracted from the original using a GenBank style </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line271">271</a>   # position string.  See also documentation for the Bio::Location class.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line272">272</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line273">273</a>   #   s = Bio::Sequence::NA.new('atgcatgcatgcatgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line274">274</a>   #   puts s.splice('1..3')                           #=&gt; &quot;atg&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line275">275</a>   #   puts s.splice('join(1..3,8..10)')               #=&gt; &quot;atgcat&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line276">276</a>   #   puts s.splice('complement(1..3)')               #=&gt; &quot;cat&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line277">277</a>   #   puts s.splice('complement(join(1..3,8..10))')   #=&gt; &quot;atgcat&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line278">278</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line279">279</a>   # Note that 'complement'ed Genbank position strings will have no </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line280">280</a>   # effect on Bio::Sequence::AA objects.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line281">281</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line282">282</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line283">283</a>   # * (required) _position_: String *or* Bio::Location object</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line284">284</a>   # *Returns*:: Bio::Sequence::NA/AA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line285">285</a>   def splice(position)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line286">286</a>     unless position.is_a?(Locations) then</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line287">287</a>       position = Locations.new(position)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line288">288</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line289">289</a>     s = ''</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line290">290</a>     position.each do |location|</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line291">291</a>       if location.sequence</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line292">292</a>         s &lt;&lt; location.sequence</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line293">293</a>       else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line294">294</a>         exon = self.subseq(location.from, location.to)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line295">295</a>         begin</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line296">296</a>           exon.complement! if location.strand &lt; 0</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line297">297</a>         rescue NameError</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line298">298</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line299">299</a>         s &lt;&lt; exon</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line300">300</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line301">301</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line302">302</a>     return self.class.new(s)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line303">303</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line304">304</a>   alias splicing splice</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line305">305</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line306">306</a> end # Common</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line307">307</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line308">308</a> end # Sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line309">309</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line310">310</a> end # Bio</pre></td>
          </tr>
        
      </tbody>
    </table>

    <p>Generated on Fri Jun 25 17:37:23 +0900 2010 with <a href="http://github.com/relevance/rcov">rcov 0.9.8</a></p>

  </body>
</html>
