<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang='en' xml:lang='en' xmlns='http://www.w3.org/1999/xhtml'>
  <head>
    <title>lib/bio/sequence/na.rb</title>
    <link href="screen.css" media="all" rel="stylesheet" type="text/css" />
    <link href="print.css" media="print" rel="stylesheet" type="text/css" />
    
    <meta http-equiv="Content-type" content="text/html;charset=UTF-8" />
    <script type="text/javascript" src="rcov.js"></script>
  </head>
  <body>
    <h1>Bioruby C0 Coverage Information - RCov</h1>
    <h2>lib/bio/sequence/na.rb</h2>

    

    <div class="report_table_wrapper">
      <table class='report' id='report_table'>
        <thead>
          <tr>
            <th class="left_align">Name</th>
            <th class="right_align">Total Lines</th>
            <th class="right_align">Lines of Code</th>
            <th class="left_align">Total Coverage</th>
            <th class="left_align">Code Coverage</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td class="left_align"><a href="lib-bio-sequence-na_rb.html">lib/bio/sequence/na.rb</a></td>
            <td class='right_align'><tt>491</tt></td>
            <td class='right_align'><tt>159</tt></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>75.97%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:76px"></div>
            <div class="uncovered" style="width:24px"></div>
          </div></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>25.79%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:26px"></div>
            <div class="uncovered" style="width:74px"></div>
          </div></td>
          </tr>
        </tbody>
      </table>
    </div>
    
    <h3>Key</h3>
    
    <div class="key"><pre><span class='marked'>Code reported as executed by Ruby looks like this...</span><span class='marked1'>and this: this line is also marked as covered.</span><span class='inferred'>Lines considered as run by rcov, but not reported by Ruby, look like this,</span><span class='inferred1'>and this: these lines were inferred by rcov (using simple heuristics).</span><span class='uncovered'>Finally, here's a line marked as not executed.</span></pre></div>

    <h3>Coverage Details</h3>

    <table class="details">
      <tbody>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1">1</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line2">2</a> # = bio/sequence/na.rb - nucleic acid sequence class</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line3">3</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line4">4</a> # Copyright::   Copyright (C) 2006</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line5">5</a> #               Toshiaki Katayama &lt;k@bioruby.org&gt;,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line6">6</a> #               Ryan Raaum &lt;ryan@raaum.org&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line7">7</a> # License::     The Ruby License</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line8">8</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line9">9</a> # $Id: na.rb,v 1.7 2007/04/23 16:43:51 trevor Exp $</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line10">10</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line11">11</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line12">12</a> require 'bio/sequence/common'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line13">13</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line14">14</a> module Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line15">15</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line16">16</a>   autoload :NucleicAcid, 'bio/data/na'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line17">17</a>   autoload :CodonTable,  'bio/data/codontable'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line18">18</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line19">19</a> class Sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line20">20</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line21">21</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line22">22</a> # = DESCRIPTION</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line23">23</a> # Bio::Sequence::NA represents a bare Nucleic Acid sequence in bioruby.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line24">24</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line25">25</a> # = USAGE</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line26">26</a> #   # Create a Nucleic Acid sequence.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line27">27</a> #   dna = Bio::Sequence.auto('atgcatgcATGCATGCAAAA')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line28">28</a> #   rna = Bio::Sequence.auto('augcaugcaugcaugcaaaa')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line29">29</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line30">30</a> #   # What are the names of all the bases?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line31">31</a> #   puts dna.names</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line32">32</a> #   puts rna.names</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line33">33</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line34">34</a> #   # What is the GC percentage?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line35">35</a> #   puts dna.gc_percent</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line36">36</a> #   puts rna.gc_percent</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line37">37</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line38">38</a> #   # What is the molecular weight?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line39">39</a> #   puts dna.molecular_weight</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line40">40</a> #   puts rna.molecular_weight</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line41">41</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line42">42</a> #   # What is the reverse complement?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line43">43</a> #   puts dna.reverse_complement</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line44">44</a> #   puts dna.complement</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line45">45</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line46">46</a> #   # Is this sequence DNA or RNA?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line47">47</a> #   puts dna.rna?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line48">48</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line49">49</a> #   # Translate my sequence (see method docs for many options)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line50">50</a> #   puts dna.translate</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line51">51</a> #   puts rna.translate</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line52">52</a> class NA &lt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line53">53</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line54">54</a>   include Bio::Sequence::Common</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line55">55</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line56">56</a>   # Generate an nucleic acid sequence object from a string.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line57">57</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line58">58</a>   #   s = Bio::Sequence::NA.new(&quot;aagcttggaccgttgaagt&quot;)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line59">59</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line60">60</a>   # or maybe (if you have an nucleic acid sequence in a file)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line61">61</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line62">62</a>   #   s = Bio::Sequence:NA.new(File.open('dna.txt').read)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line63">63</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line64">64</a>   # Nucleic Acid sequences are *always* all lowercase in bioruby</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line65">65</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line66">66</a>   #   s = Bio::Sequence::NA.new(&quot;AAGcTtGG&quot;)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line67">67</a>   #   puts s                                  #=&gt; &quot;aagcttgg&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line68">68</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line69">69</a>   # Whitespace is stripped from the sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line70">70</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line71">71</a>   #   seq = Bio::Sequence::NA.new(&quot;atg\nggg\ttt\r  gc&quot;)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line72">72</a>   #   puts s                                  #=&gt; &quot;atggggttgc&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line73">73</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line74">74</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line75">75</a>   # * (required) _str_: String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line76">76</a>   # *Returns*:: Bio::Sequence::NA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line77">77</a>   def initialize(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line78">78</a>     super</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line79">79</a>     self.downcase!</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line80">80</a>     self.tr!(&quot; \t\n\r&quot;,'')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line81">81</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line82">82</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line83">83</a>   # Alias of Bio::Sequence::Common splice method, documented there.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line84">84</a>   def splicing(position) #:nodoc:</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line85">85</a>     mRNA = super</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line86">86</a>     if mRNA.rna?</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line87">87</a>       mRNA.tr!('t', 'u')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line88">88</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line89">89</a>       mRNA.tr!('u', 't')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line90">90</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line91">91</a>     mRNA</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line92">92</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line93">93</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line94">94</a>   # Returns a new complementary sequence object (without reversing).</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line95">95</a>   # The original sequence object is not modified.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line96">96</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line97">97</a>   #   s = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line98">98</a>   #   puts s.forward_complement               #=&gt; 'tacg'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line99">99</a>   #   puts s                                  #=&gt; 'atgc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line100">100</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line101">101</a>   # *Returns*:: new Bio::Sequence::NA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line102">102</a>   def forward_complement</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line103">103</a>     s = self.class.new(self)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line104">104</a>     s.forward_complement!</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line105">105</a>     s</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line106">106</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line107">107</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line108">108</a>   # Converts the current sequence into its complement (without reversing).</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line109">109</a>   # The original sequence object is modified.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line110">110</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line111">111</a>   #   seq = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line112">112</a>   #   puts s.forward_complement!              #=&gt; 'tacg'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line113">113</a>   #   puts s                                  #=&gt; 'tacg'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line114">114</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line115">115</a>   # *Returns*:: current Bio::Sequence::NA object (modified)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line116">116</a>   def forward_complement!</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line117">117</a>     if self.rna?</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line118">118</a>       self.tr!('augcrymkdhvbswn', 'uacgyrkmhdbvswn')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line119">119</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line120">120</a>       self.tr!('atgcrymkdhvbswn', 'tacgyrkmhdbvswn')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line121">121</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line122">122</a>     self</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line123">123</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line124">124</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line125">125</a>   # Returns a new sequence object with the reverse complement </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line126">126</a>   # sequence to the original.  The original sequence is not modified.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line127">127</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line128">128</a>   #   s = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line129">129</a>   #   puts s.reverse_complement               #=&gt; 'gcat'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line130">130</a>   #   puts s                                  #=&gt; 'atgc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line131">131</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line132">132</a>   # *Returns*:: new Bio::Sequence::NA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line133">133</a>   def reverse_complement</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line134">134</a>     s = self.class.new(self)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line135">135</a>     s.reverse_complement!</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line136">136</a>     s</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line137">137</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line138">138</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line139">139</a>   # Converts the original sequence into its reverse complement.  </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line140">140</a>   # The original sequence is modified.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line141">141</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line142">142</a>   #   s = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line143">143</a>   #   puts s.reverse_complement               #=&gt; 'gcat'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line144">144</a>   #   puts s                                  #=&gt; 'gcat'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line145">145</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line146">146</a>   # *Returns*:: current Bio::Sequence::NA object (modified)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line147">147</a>   def reverse_complement!</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line148">148</a>     self.reverse!</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line149">149</a>     self.forward_complement!</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line150">150</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line151">151</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line152">152</a>   # Alias for Bio::Sequence::NA#reverse_complement</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line153">153</a>   alias complement reverse_complement</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line154">154</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line155">155</a>   # Alias for Bio::Sequence::NA#reverse_complement!</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line156">156</a>   alias complement! reverse_complement!</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line157">157</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line158">158</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line159">159</a>   # Translate into an amino acid sequence.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line160">160</a>   #   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line161">161</a>   #   s = Bio::Sequence::NA.new('atggcgtga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line162">162</a>   #   puts s.translate                        #=&gt; &quot;MA*&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line163">163</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line164">164</a>   # By default, translate starts in reading frame position 1, but you</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line165">165</a>   # can start in either 2 or 3 as well,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line166">166</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line167">167</a>   #   puts s.translate(2)                     #=&gt; &quot;WR&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line168">168</a>   #   puts s.translate(3)                     #=&gt; &quot;GV&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line169">169</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line170">170</a>   # You may also translate the reverse complement in one step by using frame</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line171">171</a>   # values of -1, -2, and -3 (or 4, 5, and 6)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line172">172</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line173">173</a>   #   puts s.translate(-1)                    #=&gt; &quot;SRH&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line174">174</a>   #   puts s.translate(4)                     #=&gt; &quot;SRH&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line175">175</a>   #   puts s.reverse_complement.translate(1)  #=&gt; &quot;SRH&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line176">176</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line177">177</a>   # The default codon table in the translate function is the Standard</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line178">178</a>   # Eukaryotic codon table.  The translate function takes either a </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line179">179</a>   # number or a Bio::CodonTable object for its table argument. </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line180">180</a>   # The available tables are </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line181">181</a>   # (NCBI[http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=t]):</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line182">182</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line183">183</a>   #   1. &quot;Standard (Eukaryote)&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line184">184</a>   #   2. &quot;Vertebrate Mitochondrial&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line185">185</a>   #   3. &quot;Yeast Mitochondorial&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line186">186</a>   #   4. &quot;Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line187">187</a>   #   5. &quot;Invertebrate Mitochondrial&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line188">188</a>   #   6. &quot;Ciliate Macronuclear and Dasycladacean&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line189">189</a>   #   9. &quot;Echinoderm Mitochondrial&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line190">190</a>   #   10. &quot;Euplotid Nuclear&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line191">191</a>   #   11. &quot;Bacteria&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line192">192</a>   #   12. &quot;Alternative Yeast Nuclear&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line193">193</a>   #   13. &quot;Ascidian Mitochondrial&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line194">194</a>   #   14. &quot;Flatworm Mitochondrial&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line195">195</a>   #   15. &quot;Blepharisma Macronuclear&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line196">196</a>   #   16. &quot;Chlorophycean Mitochondrial&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line197">197</a>   #   21. &quot;Trematode Mitochondrial&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line198">198</a>   #   22. &quot;Scenedesmus obliquus mitochondrial&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line199">199</a>   #   23. &quot;Thraustochytrium Mitochondrial&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line200">200</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line201">201</a>   # If you are using anything other than the default table, you must specify </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line202">202</a>   # frame in the translate method call,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line203">203</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line204">204</a>   #   puts s.translate                #=&gt; &quot;MA*&quot;  (using defaults)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line205">205</a>   #   puts s.translate(1,1)           #=&gt; &quot;MA*&quot;  (same as above, but explicit)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line206">206</a>   #   puts s.translate(1,2)           #=&gt; &quot;MAW&quot;  (different codon table)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line207">207</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line208">208</a>   # and using a Bio::CodonTable instance in the translate method call,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line209">209</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line210">210</a>   #   mt_table = Bio::CodonTable[2]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line211">211</a>   #   puts s.translate(1, mt_table)           #=&gt; &quot;MAW&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line212">212</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line213">213</a>   # By default, any invalid or unknown codons (as could happen if the </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line214">214</a>   # sequence contains ambiguities) will be represented by 'X' in the </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line215">215</a>   # translated sequence. </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line216">216</a>   # You may change this to any character of your choice.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line217">217</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line218">218</a>   #   s = Bio::Sequence::NA.new('atgcNNtga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line219">219</a>   #   puts s.translate                        #=&gt; &quot;MX*&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line220">220</a>   #   puts s.translate(1,1,'9')               #=&gt; &quot;M9*&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line221">221</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line222">222</a>   # The translate method considers gaps to be unknown characters and treats </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line223">223</a>   # them as such (i.e. does not collapse sequences prior to translation), so</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line224">224</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line225">225</a>   #   s = Bio::Sequence::NA.new('atgc--tga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line226">226</a>   #   puts s.translate                        #=&gt; &quot;MX*&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line227">227</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line228">228</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line229">229</a>   # * (optional) _frame_:  one of 1,2,3,4,5,6,-1,-2,-3 (default 1)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line230">230</a>   # * (optional) _table_: Fixnum in range 1,23 or Bio::CodonTable object</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line231">231</a>   #   (default 1)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line232">232</a>   # * (optional) _unknown_: Character (default 'X')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line233">233</a>   # *Returns*:: Bio::Sequence::AA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line234">234</a>   def translate(frame = 1, table = 1, unknown = 'X')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line235">235</a>     if table.is_a?(Bio::CodonTable)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line236">236</a>       ct = table</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line237">237</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line238">238</a>       ct = Bio::CodonTable[table]</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line239">239</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line240">240</a>     naseq = self.dna</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line241">241</a>     case frame</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line242">242</a>     when 1, 2, 3</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line243">243</a>       from = frame - 1</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line244">244</a>     when 4, 5, 6</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line245">245</a>       from = frame - 4</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line246">246</a>       naseq.complement!</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line247">247</a>     when -1, -2, -3</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line248">248</a>       from = -1 - frame</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line249">249</a>       naseq.complement!</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line250">250</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line251">251</a>       from = 0</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line252">252</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line253">253</a>     nalen = naseq.length - from</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line254">254</a>     nalen -= nalen % 3</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line255">255</a>     aaseq = naseq[from, nalen].gsub(/.{3}/) {|codon| ct[codon] or unknown}</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line256">256</a>     return Bio::Sequence::AA.new(aaseq)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line257">257</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line258">258</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line259">259</a>   # Returns counts of each codon in the sequence in a hash.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line260">260</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line261">261</a>   #   s = Bio::Sequence::NA.new('atggcgtga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line262">262</a>   #   puts s.codon_usage                #=&gt; {&quot;gcg&quot;=&gt;1, &quot;tga&quot;=&gt;1, &quot;atg&quot;=&gt;1}</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line263">263</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line264">264</a>   # This method does not validate codons!  Any three letter group is a 'codon'. So,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line265">265</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line266">266</a>   #   s = Bio::Sequence::NA.new('atggNNtga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line267">267</a>   #   puts s.codon_usage                #=&gt; {&quot;tga&quot;=&gt;1, &quot;gnn&quot;=&gt;1, &quot;atg&quot;=&gt;1}</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line268">268</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line269">269</a>   #   seq = Bio::Sequence::NA.new('atgg--tga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line270">270</a>   #   puts s.codon_usage                #=&gt; {&quot;tga&quot;=&gt;1, &quot;g--&quot;=&gt;1, &quot;atg&quot;=&gt;1}</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line271">271</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line272">272</a>   # Also, there is no option to work in any frame other than the first.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line273">273</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line274">274</a>   # *Returns*:: Hash object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line275">275</a>   def codon_usage</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line276">276</a>     hash = Hash.new(0)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line277">277</a>     self.window_search(3, 3) do |codon|</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line278">278</a>       hash[codon] += 1</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line279">279</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line280">280</a>     return hash</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line281">281</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line282">282</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line283">283</a>   # Calculate the ratio of GC / ATGC bases as a percentage rounded to </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line284">284</a>   # the nearest whole number. U is regarded as T.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line285">285</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line286">286</a>   #   s = Bio::Sequence::NA.new('atggcgtga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line287">287</a>   #   puts s.gc_percent                       #=&gt; 55</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line288">288</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line289">289</a>   # *Returns*:: Fixnum</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line290">290</a>   def gc_percent</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line291">291</a>     count = self.composition</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line292">292</a>     at = count['a'] + count['t'] + count['u']</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line293">293</a>     gc = count['g'] + count['c']</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line294">294</a>     return 0 if at + gc == 0</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line295">295</a>     gc = 100 * gc / (at + gc)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line296">296</a>     return gc</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line297">297</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line298">298</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line299">299</a>   # Calculate the ratio of GC / ATGC bases. U is regarded as T.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line300">300</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line301">301</a>   #   s = Bio::Sequence::NA.new('atggcgtga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line302">302</a>   #   puts s.gc_content                       #=&gt; 0.555555555555556</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line303">303</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line304">304</a>   # *Returns*:: Float</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line305">305</a>   def gc_content</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line306">306</a>     count = self.composition</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line307">307</a>     at = count['a'] + count['t'] + count['u']</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line308">308</a>     gc = count['g'] + count['c']</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line309">309</a>     return 0.0 if at + gc == 0</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line310">310</a>     return gc.quo(at + gc)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line311">311</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line312">312</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line313">313</a>   # Calculate the ratio of AT / ATGC bases. U is regarded as T.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line314">314</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line315">315</a>   #   s = Bio::Sequence::NA.new('atggcgtga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line316">316</a>   #   puts s.at_content                       #=&gt; 0.444444444444444</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line317">317</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line318">318</a>   # *Returns*:: Float</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line319">319</a>   def at_content</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line320">320</a>     count = self.composition</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line321">321</a>     at = count['a'] + count['t'] + count['u']</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line322">322</a>     gc = count['g'] + count['c']</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line323">323</a>     return 0.0 if at + gc == 0</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line324">324</a>     return at.quo(at + gc)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line325">325</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line326">326</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line327">327</a>   # Calculate the ratio of (G - C) / (G + C) bases.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line328">328</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line329">329</a>   #   s = Bio::Sequence::NA.new('atggcgtga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line330">330</a>   #   puts s.gc_skew                          #=&gt; 0.6</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line331">331</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line332">332</a>   # *Returns*:: Float</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line333">333</a>   def gc_skew</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line334">334</a>     count = self.composition</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line335">335</a>     g = count['g']</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line336">336</a>     c = count['c']</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line337">337</a>     return 0.0 if g + c == 0</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line338">338</a>     return (g - c).quo(g + c)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line339">339</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line340">340</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line341">341</a>   # Calculate the ratio of (A - T) / (A + T) bases. U is regarded as T.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line342">342</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line343">343</a>   #   s = Bio::Sequence::NA.new('atgttgttgttc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line344">344</a>   #   puts s.at_skew                          #=&gt; -0.75</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line345">345</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line346">346</a>   # *Returns*:: Float</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line347">347</a>   def at_skew</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line348">348</a>     count = self.composition</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line349">349</a>     a = count['a']</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line350">350</a>     t = count['t'] + count['u']</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line351">351</a>     return 0.0 if a + t == 0</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line352">352</a>     return (a - t).quo(a + t)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line353">353</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line354">354</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line355">355</a>   # Returns an alphabetically sorted array of any non-standard bases </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line356">356</a>   # (other than 'atgcu').</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line357">357</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line358">358</a>   #   s = Bio::Sequence::NA.new('atgStgQccR')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line359">359</a>   #   puts s.illegal_bases                    #=&gt; [&quot;q&quot;, &quot;r&quot;, &quot;s&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line360">360</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line361">361</a>   # *Returns*:: Array object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line362">362</a>   def illegal_bases</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line363">363</a>     self.scan(/[^atgcu]/).sort.uniq</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line364">364</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line365">365</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line366">366</a>   # Estimate molecular weight (using the values from BioPerl's </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line367">367</a>   # SeqStats.pm[http://doc.bioperl.org/releases/bioperl-1.0.1/Bio/Tools/SeqStats.html] module).</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line368">368</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line369">369</a>   #   s = Bio::Sequence::NA.new('atggcgtga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line370">370</a>   #   puts s.molecular_weight                 #=&gt; 2841.00708</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line371">371</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line372">372</a>   # RNA and DNA do not have the same molecular weights,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line373">373</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line374">374</a>   #   s = Bio::Sequence::NA.new('auggcguga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line375">375</a>   #   puts s.molecular_weight                 #=&gt; 2956.94708</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line376">376</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line377">377</a>   # *Returns*:: Float object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line378">378</a>   def molecular_weight</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line379">379</a>     if self.rna?</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line380">380</a>       Bio::NucleicAcid.weight(self, true)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line381">381</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line382">382</a>       Bio::NucleicAcid.weight(self)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line383">383</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line384">384</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line385">385</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line386">386</a>   # Create a ruby regular expression instance </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line387">387</a>   # (Regexp)[http://corelib.rubyonrails.org/classes/Regexp.html]  </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line388">388</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line389">389</a>   #   s = Bio::Sequence::NA.new('atggcgtga')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line390">390</a>   #   puts s.to_re                            #=&gt; /atggcgtga/</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line391">391</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line392">392</a>   # *Returns*:: Regexp object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line393">393</a>   def to_re</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line394">394</a>     if self.rna?</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line395">395</a>       Bio::NucleicAcid.to_re(self.dna, true)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line396">396</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line397">397</a>       Bio::NucleicAcid.to_re(self)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line398">398</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line399">399</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line400">400</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line401">401</a>   # Generate the list of the names of each nucleotide along with the</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line402">402</a>   # sequence (full name).  Names used in bioruby are found in the</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line403">403</a>   # Bio::AminoAcid::NAMES hash.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line404">404</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line405">405</a>   #   s = Bio::Sequence::NA.new('atg')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line406">406</a>   #   puts s.names                    #=&gt; [&quot;Adenine&quot;, &quot;Thymine&quot;, &quot;Guanine&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line407">407</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line408">408</a>   # *Returns*:: Array object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line409">409</a>   def names</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line410">410</a>     array = []</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line411">411</a>     self.each_byte do |x|</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line412">412</a>       array.push(Bio::NucleicAcid.names[x.chr.upcase])</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line413">413</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line414">414</a>     return array</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line415">415</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line416">416</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line417">417</a>   # Returns a new sequence object with any 'u' bases changed to 't'.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line418">418</a>   # The original sequence is not modified.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line419">419</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line420">420</a>   #   s = Bio::Sequence::NA.new('augc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line421">421</a>   #   puts s.dna                              #=&gt; 'atgc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line422">422</a>   #   puts s                                  #=&gt; 'augc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line423">423</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line424">424</a>   # *Returns*:: new Bio::Sequence::NA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line425">425</a>   def dna</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line426">426</a>     self.tr('u', 't')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line427">427</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line428">428</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line429">429</a>   # Changes any 'u' bases in the original sequence to 't'.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line430">430</a>   # The original sequence is modified.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line431">431</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line432">432</a>   #   s = Bio::Sequence::NA.new('augc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line433">433</a>   #   puts s.dna!                             #=&gt; 'atgc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line434">434</a>   #   puts s                                  #=&gt; 'atgc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line435">435</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line436">436</a>   # *Returns*:: current Bio::Sequence::NA object (modified)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line437">437</a>   def dna!</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line438">438</a>     self.tr!('u', 't')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line439">439</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line440">440</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line441">441</a>   # Returns a new sequence object with any 't' bases changed to 'u'.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line442">442</a>   # The original sequence is not modified.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line443">443</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line444">444</a>   #   s = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line445">445</a>   #   puts s.dna                              #=&gt; 'augc'  </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line446">446</a>   #   puts s                                  #=&gt; 'atgc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line447">447</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line448">448</a>   # *Returns*:: new Bio::Sequence::NA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line449">449</a>   def rna</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line450">450</a>     self.tr('t', 'u')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line451">451</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line452">452</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line453">453</a>   # Changes any 't' bases in the original sequence to 'u'.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line454">454</a>   # The original sequence is modified.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line455">455</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line456">456</a>   #   s = Bio::Sequence::NA.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line457">457</a>   #   puts s.dna!                             #=&gt; 'augc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line458">458</a>   #   puts s                                  #=&gt; 'augc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line459">459</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line460">460</a>   # *Returns*:: current Bio::Sequence::NA object (modified)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line461">461</a>   def rna!</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line462">462</a>     self.tr!('t', 'u')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line463">463</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line464">464</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line465">465</a>   def rna?</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line466">466</a>     self.index('u')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line467">467</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line468">468</a>   protected :rna?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line469">469</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line470">470</a>   # Example:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line471">471</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line472">472</a>   #   seq = Bio::Sequence::NA.new('gaattc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line473">473</a>   #   cuts = seq.cut_with_enzyme('EcoRI')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line474">474</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line475">475</a>   # _or_</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line476">476</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line477">477</a>   #   seq = Bio::Sequence::NA.new('gaattc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line478">478</a>   #   cuts = seq.cut_with_enzyme('g^aattc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line479">479</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line480">480</a>   # See Bio::RestrictionEnzyme::Analysis.cut</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line481">481</a>   def cut_with_enzyme(*args)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line482">482</a>     Bio::RestrictionEnzyme::Analysis.cut(self, *args)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line483">483</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line484">484</a>   alias cut_with_enzymes cut_with_enzyme</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line485">485</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line486">486</a> end # NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line487">487</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line488">488</a> end # Sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line489">489</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line490">490</a> end # Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line491">491</a> </pre></td>
          </tr>
        
      </tbody>
    </table>

    <p>Generated on Fri Jun 25 17:37:23 +0900 2010 with <a href="http://github.com/relevance/rcov">rcov 0.9.8</a></p>

  </body>
</html>
