<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang='en' xml:lang='en' xmlns='http://www.w3.org/1999/xhtml'>
  <head>
    <title>lib/bio/db/phyloxml/phyloxml_elements.rb</title>
    <link href="screen.css" media="all" rel="stylesheet" type="text/css" />
    <link href="print.css" media="print" rel="stylesheet" type="text/css" />
    
    <meta http-equiv="Content-type" content="text/html;charset=UTF-8" />
    <script type="text/javascript" src="rcov.js"></script>
  </head>
  <body>
    <h1>Bioruby C0 Coverage Information - RCov</h1>
    <h2>lib/bio/db/phyloxml/phyloxml_elements.rb</h2>

    

    <div class="report_table_wrapper">
      <table class='report' id='report_table'>
        <thead>
          <tr>
            <th class="left_align">Name</th>
            <th class="right_align">Total Lines</th>
            <th class="right_align">Lines of Code</th>
            <th class="left_align">Total Coverage</th>
            <th class="left_align">Code Coverage</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td class="left_align"><a href="lib-bio-db-phyloxml-phyloxml_elements_rb.html">lib/bio/db/phyloxml/phyloxml_elements.rb</a></td>
            <td class='right_align'><tt>1174</tt></td>
            <td class='right_align'><tt>671</tt></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>93.61%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:94px"></div>
            <div class="uncovered" style="width:6px"></div>
          </div></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>89.42%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:89px"></div>
            <div class="uncovered" style="width:11px"></div>
          </div></td>
          </tr>
        </tbody>
      </table>
    </div>
    
    <h3>Key</h3>
    
    <div class="key"><pre><span class='marked'>Code reported as executed by Ruby looks like this...</span><span class='marked1'>and this: this line is also marked as covered.</span><span class='inferred'>Lines considered as run by rcov, but not reported by Ruby, look like this,</span><span class='inferred1'>and this: these lines were inferred by rcov (using simple heuristics).</span><span class='uncovered'>Finally, here's a line marked as not executed.</span></pre></div>

    <h3>Coverage Details</h3>

    <table class="details">
      <tbody>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1">1</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line2">2</a> # = bio/db/phyloxml_elements.rb - PhyloXML Element classes</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line3">3</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line4">4</a> # Copyright::   Copyright (C) 2009</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line5">5</a> #               Diana Jaunzeikare &lt;latvianlinuxgirl@gmail.com&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line6">6</a> # License::     The Ruby License</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line7">7</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line8">8</a> # $Id:$</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line9">9</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line10">10</a> # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line11">11</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line12">12</a> # This file containts the classes to represent PhyloXML elements.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line13">13</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line14">14</a> # == References</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line15">15</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line16">16</a> # * http://www.phyloxml.org</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line17">17</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line18">18</a> # * https://www.nescent.org/wg_phyloinformatics/PhyloSoC:PhyloXML_support_in_BioRuby</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line19">19</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line20">20</a> require 'bio/tree'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line21">21</a> require 'bio/sequence'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line22">22</a> require 'bio/reference'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line23">23</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line24">24</a> # Autoload definition</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line25">25</a> module Bio</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line26">26</a>   module PhyloXML</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line27">27</a>     autoload :Parser, 'bio/db/phyloxml/phyloxml_parser'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line28">28</a>     autoload :Writer, 'bio/db/phyloxml/phyloxml_writer'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line29">29</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line30">30</a> end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line31">31</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line32">32</a> require 'libxml'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line33">33</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line34">34</a> module Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line35">35</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line36">36</a>    # This is general Taxonomy class.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line37">37</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line38">38</a>   class Taxonomy</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line39">39</a>     #pattern = [a-zA-Z0-9_]{2,10} Can refer to any code/abbreviation/mnemonic, such as Bsu for Bacillus subtilis.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line40">40</a>     attr_accessor :code</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line41">41</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line42">42</a>     # String.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line43">43</a>     attr_accessor :scientific_name</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line44">44</a>     # An array of strings</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line45">45</a>     attr_accessor :common_names</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line46">46</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line47">47</a>     # value comes from list: domain kingdom, subkingdom, branch, infrakingdom,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line48">48</a>     # superphylum, phylum, subphylum, infraphylum, microphylum, superdivision,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line49">49</a>     # division, subdivision, infradivision, superclass, class, subclass,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line50">50</a>     # infraclass, superlegion, legion, sublegion, infralegion, supercohort,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line51">51</a>     # cohort, subcohort, infracohort, superorder, order, suborder,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line52">52</a>     # superfamily, family, subfamily, supertribe, tribe, subtribe, infratribe,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line53">53</a>     # genus, subgenus, superspecies, species, subspecies, variety, subvariety,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line54">54</a>     # form, subform, cultivar, unknown, other</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line55">55</a>     attr_accessor :rank</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line56">56</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line57">57</a>     # is used to keep the authority, such as 'J. G. Cooper, 1863', associated with the 'scientific_name'.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line58">58</a>     attr_accessor :authority</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line59">59</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line60">60</a>     # An array of strings. Holds synonyms for scientific names or common names.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line61">61</a>     attr_accessor :synonyms</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line62">62</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line63">63</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line64">64</a>     def initialize</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line65">65</a>       @common_names = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line66">66</a>       @synonyms = []</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line67">67</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line68">68</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line69">69</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line70">70</a> module PhyloXML</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line71">71</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line72">72</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line73">73</a>   # Taxonomy class</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line74">74</a>   class Taxonomy &lt; Bio::Taxonomy</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line75">75</a>     # String. Unique identifier of a taxon.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line76">76</a>     attr_accessor :taxonomy_id</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line77">77</a>     #Used to link other elements to a taxonomy (on the xml-level)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line78">78</a>     attr_accessor :id_source</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line79">79</a>     # Uri object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line80">80</a>     attr_accessor :uri</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line81">81</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line82">82</a>     # Array of Other objects. Used to save additional information from other than</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line83">83</a>     # PhyloXML namspace.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line84">84</a>     attr_accessor :other</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line85">85</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line86">86</a>     def initialize</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line87">87</a>       super</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line88">88</a>       @other = []</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line89">89</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line90">90</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line91">91</a>     # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line92">92</a>     def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line93">93</a>       taxonomy = LibXML::XML::Node.new('taxonomy')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line94">94</a>       taxonomy[&quot;type&quot;] = @type if @type != nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line95">95</a>       taxonomy[&quot;id_source&quot;] = @id_source if @id_source != nil</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line96">96</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line97">97</a>       PhyloXML::Writer.generate_xml(taxonomy, self, [[:complex, 'id', @taxonomy_id],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line98">98</a>         [:pattern, 'code', @code, Regexp.new(&quot;^[a-zA-Z0-9_]{2,10}$&quot;)],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line99">99</a>         [:simple, 'scientific_name', @scientific_name],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line100">100</a>         [:simple, 'authority', @authority],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line101">101</a>         [:simplearr, 'common_name', @common_names],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line102">102</a>         [:simplearr, 'synonym', @synonyms],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line103">103</a>         [:simple, 'rank', @rank],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line104">104</a>         [:complex, 'uri',@uri]])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line105">105</a>         #@todo anything else</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line106">106</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line107">107</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line108">108</a>       return taxonomy</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line109">109</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line110">110</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line111">111</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line112">112</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line113">113</a>   # Object to hold one phylogeny element (and its subelements.) Extended version of Bio::Tree.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line114">114</a>   class Tree &lt; Bio::Tree</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line115">115</a>     # String. Name of tree (name subelement of phylogeny element).</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line116">116</a>     attr_accessor :name</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line117">117</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line118">118</a>     # Id object.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line119">119</a>     attr_accessor :phylogeny_id</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line120">120</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line121">121</a>     # String. Description of tree.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line122">122</a>     attr_accessor :description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line123">123</a>    </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line124">124</a>     # Boolean. Can be used to indicate that the phylogeny is not allowed to be rooted differently (i.e. because it is associated with root dependent data, such as gene duplications).</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line125">125</a>     attr_accessor :rerootable</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line126">126</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line127">127</a>     # Boolean. Required element.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line128">128</a>     attr_accessor  :rooted</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line129">129</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line130">130</a>     # Array of Property object. Allows for typed and referenced properties from external resources to be attached.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line131">131</a>     attr_accessor :properties</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line132">132</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line133">133</a>     # CladeRelation object. This is used to express a typed relationship between two clades. For example it could be used to describe multiple parents of a clade.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line134">134</a>     attr_accessor :clade_relations</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line135">135</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line136">136</a>     # SequenceRelation object. This is used to express a typed relationship between two sequences. For example it could be used to describe an orthology.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line137">137</a>     attr_accessor :sequence_relations</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line138">138</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line139">139</a>     # Array of confidence object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line140">140</a>     attr_accessor :confidences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line141">141</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line142">142</a>     # String.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line143">143</a>     attr_accessor :branch_length_unit</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line144">144</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line145">145</a>     # String. Indicate the type of phylogeny (i.e. 'gene tree').</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line146">146</a>     attr_accessor :type</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line147">147</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line148">148</a>     # String. Date</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line149">149</a>     attr_accessor :date</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line150">150</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line151">151</a>     # Array of Other objects. Used to save additional information from other than</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line152">152</a>     # PhyloXML namspace.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line153">153</a>     attr_accessor :other</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line154">154</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line155">155</a>    def initialize</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line156">156</a>      super</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line157">157</a>      @sequence_relations = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line158">158</a>      @clade_relations = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line159">159</a>      @confidences = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line160">160</a>      @properties = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line161">161</a>      @other = []</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line162">162</a>    end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line163">163</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line164">164</a>  end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line165">165</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line166">166</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line167">167</a>   # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line168">168</a>   # Class to hold clade element of phyloXML.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line169">169</a>   class Node</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line170">170</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line171">171</a>     # Events at the root node of a clade (e.g. one gene duplication).</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line172">172</a>     attr_accessor :events</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line173">173</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line174">174</a>     # String. Used to link other elements to a clade (node) (on the xml-level).</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line175">175</a>     attr_accessor :id_source</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line176">176</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line177">177</a>     # String. Name of the node.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line178">178</a>     attr_accessor :name</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line179">179</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line180">180</a>     # Float. Branch width for this node (including parent branch). Applies for the whole clade unless overwritten in sub-clades.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line181">181</a>     attr_reader :width</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line182">182</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line183">183</a>     def width=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line184">184</a>       @width = str.to_f</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line185">185</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line186">186</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line187">187</a>     # Array of Taxonomy objects. Describes taxonomic information for a clade.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line188">188</a>     attr_accessor :taxonomies</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line189">189</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line190">190</a>     # Array of Confidence objects. Indicates the support for a clade/parent branch.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line191">191</a>     attr_accessor :confidences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line192">192</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line193">193</a>     # BranchColor object. Apply for the whole clade unless overwritten in sub-clade.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line194">194</a>     attr_accessor :color</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line195">195</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line196">196</a>     # Id object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line197">197</a>     attr_accessor :node_id</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line198">198</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line199">199</a>     # Array of Sequence objects. Represents a molecular sequence (Protein, DNA, RNA) associated with a node.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line200">200</a>     attr_accessor :sequences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line201">201</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line202">202</a>     # BinaryCharacters object. The names and/or counts of binary characters present, gained, and lost at the root of a clade.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line203">203</a>     attr_accessor :binary_characters</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line204">204</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line205">205</a>     # Array of Distribution objects. The geographic distribution of the items of a clade (species, sequences), intended for phylogeographic applications.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line206">206</a>     attr_accessor :distributions</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line207">207</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line208">208</a>     # Date object. A date associated with a clade/node.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line209">209</a>     attr_accessor :date</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line210">210</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line211">211</a>     #Array of Reference objects. A literature reference for a clade.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line212">212</a>     attr_accessor :references</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line213">213</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line214">214</a>     #An array of Property objects, for example depth for sea animals.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line215">215</a>     attr_accessor :properties</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line216">216</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line217">217</a>     # Array of Other objects. Used to save additional information from other than</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line218">218</a>     # PhyloXML namspace.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line219">219</a>     attr_accessor :other</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line220">220</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line221">221</a>     def initialize</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line222">222</a>       @confidences = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line223">223</a>       @sequences = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line224">224</a>       @taxonomies = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line225">225</a>       @distributions = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line226">226</a>       @references = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line227">227</a>       @properties = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line228">228</a>       @other = []</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line229">229</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line230">230</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line231">231</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line232">232</a>     # Converts to a Bio::Tree::Node object. If it contains several taxonomies </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line233">233</a>     # Bio::Tree::Node#scientific name will get the scientific name of the first </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line234">234</a>     # taxonomy.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line235">235</a>     # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line236">236</a>     # If there are several confidence values, the first with bootstrap type will </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line237">237</a>     # be returned as Bio::Tree::Node#bootstrap</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line238">238</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line239">239</a>     # tree = phyloxmlparser.next_tree</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line240">240</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line241">241</a>     # node = tree.get_node_by_name(&quot;A&quot;).to_biotreenode</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line242">242</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line243">243</a>     # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line244">244</a>     # *Returns*:: Bio::Tree::Node</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line245">245</a>     def to_biotreenode</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line246">246</a>       node = Bio::Tree::Node.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line247">247</a>       node.name = @name</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line248">248</a>       node.scientific_name = @taxonomies[0].scientific_name if not @taxonomies.empty?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line249">249</a>       #@todo what if there are more?</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line250">250</a>       node.taxonomy_id = @taxonomies[0].taxononmy_id if @taxonomies[0] != nil</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line251">251</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line252">252</a>       if not @confidences.empty?</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line253">253</a>         @confidences.each do |confidence|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line254">254</a>           if confidence.type == &quot;bootstrap&quot;</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line255">255</a>             node.bootstrap = confidence.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line256">256</a>             break</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line257">257</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line258">258</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line259">259</a>       end      </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line260">260</a>       return node</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line261">261</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line262">262</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line263">263</a>     # Extracts the relevant information from node (specifically taxonomy and</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line264">264</a>     # sequence) to create Bio::Sequence object. Node can have several sequences,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line265">265</a>     # so parameter to this method is to specify which sequence to extract. </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line266">266</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line267">267</a>     # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line268">268</a>     # *Returns*:: Bio::Sequence</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line269">269</a>     def extract_biosequence(seq_i=0)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line270">270</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line271">271</a>       seq = @sequences[seq_i].to_biosequence</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line272">272</a>       seq.classification = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line273">273</a>       @taxonomies.each do |t|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line274">274</a>         seq.classification &lt;&lt; t.scientific_name</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line275">275</a>         if t.rank == &quot;species&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line276">276</a>           seq.species = t.scientific_name</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line277">277</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line278">278</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line279">279</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line280">280</a>       #seq.division =&gt; .. http://www.ebi.ac.uk/embl/Documentation/User_manual/usrman.html#3_2</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line281">281</a>       # It doesn't seem there is anything in PhyloXML corresponding to this.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line282">282</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line283">283</a>       return seq</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line284">284</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line285">285</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line286">286</a>     # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line287">287</a>     def to_xml(branch_length,  write_branch_length_as_subelement)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line288">288</a>       clade = LibXML::XML::Node.new('clade')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line289">289</a>       </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line290">290</a>       PhyloXML::Writer.generate_xml(clade, self, [[:simple, 'name', @name]])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line291">291</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line292">292</a>       if branch_length != nil       </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line293">293</a>         if write_branch_length_as_subelement</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line294">294</a>           clade &lt;&lt; LibXML::XML::Node.new('branch_length', branch_length.to_s)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line295">295</a>         else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line296">296</a>           clade[&quot;branch_length&quot;] = branch_length.to_s</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line297">297</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line298">298</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line299">299</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line300">300</a>       #generate all elements, except clade</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line301">301</a>       PhyloXML::Writer.generate_xml(clade, self, [</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line302">302</a>           [:attr, &quot;id_source&quot;],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line303">303</a>           [:objarr, 'confidence', 'confidences'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line304">304</a>           [:simple, 'width', @width],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line305">305</a>           [:complex, 'branch_color', @branch_color],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line306">306</a>           [:simple, 'node_id', @node_id],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line307">307</a>           [:objarr, 'taxonomy', 'taxonomies'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line308">308</a>           [:objarr, 'sequence', 'sequences'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line309">309</a>           [:complex, 'events', @events],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line310">310</a>           [:complex, 'binary_characters', @binary_characters],          </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line311">311</a>           [:objarr, 'distribution', 'distributions'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line312">312</a>           [:complex, 'date', @date],          </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line313">313</a>           [:objarr, 'reference', 'references'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line314">314</a>           [:objarr, 'propery', 'properties']])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line315">315</a>      </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line316">316</a>       return clade</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line317">317</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line318">318</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line319">319</a>   end #Node</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line320">320</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line321">321</a>   # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line322">322</a>   # Events at the root node of a clade (e.g. one gene duplication).</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line323">323</a>   class Events</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line324">324</a>     #value comes from list: transfer, fusion, speciation_or_duplication, other, mixed, unassigned</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line325">325</a>     attr_accessor :type</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line326">326</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line327">327</a>     # Integer</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line328">328</a>     attr_reader :duplications, :speciations, :losses</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line329">329</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line330">330</a>     # Confidence object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line331">331</a>     attr_reader :confidence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line332">332</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line333">333</a>     def confidence=(type, value)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line334">334</a>       @confidence = Confidence.new(type, value)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line335">335</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line336">336</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line337">337</a>     def confidence=(conf)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line338">338</a>       @confidence = conf</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line339">339</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line340">340</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line341">341</a>     def duplications=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line342">342</a>       @duplications = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line343">343</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line344">344</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line345">345</a>     def losses=(str)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line346">346</a>       @losses = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line347">347</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line348">348</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line349">349</a>     def speciations=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line350">350</a>       @speciations=str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line351">351</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line352">352</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line353">353</a>     def type=(str)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line354">354</a>       @type = str</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line355">355</a>       #@todo add unit test for this</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line356">356</a>       if not ['transfer','fusion','speciation_or_duplication','other','mixed', 'unassigned'].include?(str)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line357">357</a>         raise &quot;Warning #{str} is not one of the allowed values&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line358">358</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line359">359</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line360">360</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line361">361</a>     # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line362">362</a>     def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line363">363</a>       #@todo add unit test</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line364">364</a>       events = LibXML::XML::Node.new('events')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line365">365</a>       PhyloXML::Writer.generate_xml(events, self, [</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line366">366</a>         [:simple, 'type', @type],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line367">367</a>         [:simple, 'duplications', @duplications],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line368">368</a>         [:simple, 'speciations', @speciations],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line369">369</a>         [:simple, 'losses', @losses],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line370">370</a>         [:complex, 'confidence', @confidence]])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line371">371</a>       return events</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line372">372</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line373">373</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line374">374</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line375">375</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line376">376</a>     # A general purpose confidence element. For example this can be used to express</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line377">377</a>     # the bootstrap support value of a clade (in which case the 'type' attribute</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line378">378</a>     # is 'bootstrap').</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line379">379</a>     class Confidence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line380">380</a>       # String. The type of confidence measure, for example, bootstrap.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line381">381</a>       attr_accessor :type</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line382">382</a>       # Float. The value of confidence measure.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line383">383</a>       attr_accessor :value</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line384">384</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line385">385</a>       def initialize(type, value)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line386">386</a>         @type = type</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line387">387</a>         @value = value.to_f</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line388">388</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line389">389</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line390">390</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line391">391</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line392">392</a>         if @type == nil</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line393">393</a>           raise &quot;Type is a required attribute for confidence.&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line394">394</a>         else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line395">395</a>           confidence = LibXML::XML::Node.new('confidence', @value.to_s)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line396">396</a>           confidence[&quot;type&quot;] = @type          </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line397">397</a>           return confidence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line398">398</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line399">399</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line400">400</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line401">401</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line402">402</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line403">403</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line404">404</a>     # The geographic distribution of the items of a clade (species, sequences),</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line405">405</a>     # intended for phylogeographic applications. </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line406">406</a>     class Distribution</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line407">407</a>       # String. Free text description of location.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line408">408</a>       attr_accessor :desc</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line409">409</a>       # Array of Point objects. Holds coordinates of the location.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line410">410</a>       attr_accessor :points</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line411">411</a>       # Array of Polygon objects.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line412">412</a>       attr_accessor :polygons</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line413">413</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line414">414</a>       def initialize</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line415">415</a>         @points = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line416">416</a>         @polygons = []</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line417">417</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line418">418</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line419">419</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line420">420</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line421">421</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line422">422</a>         distr = LibXML::XML::Node.new('distribution')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line423">423</a>         PhyloXML::Writer.generate_xml(distr, self, [</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line424">424</a>             [:simple, 'desc', @desc],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line425">425</a>             [:objarr, 'point', 'points'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line426">426</a>             [:objarr, 'polygon', 'polygons']])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line427">427</a>         return distr</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line428">428</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line429">429</a>       </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line430">430</a>     end #Distribution class</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line431">431</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line432">432</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line433">433</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line434">434</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line435">435</a>     # The coordinates of a point with an optional altitude. Required attribute</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line436">436</a>     # 'geodetic_datum' is used to indicate the geodetic datum (also called</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line437">437</a>     # 'map datum'), for example Google's KML uses 'WGS84'.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line438">438</a>     class Point</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line439">439</a>       # Float. Latitude</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line440">440</a>       attr_accessor :lat</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line441">441</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line442">442</a>       # Float. Longitute</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line443">443</a>       attr_accessor :long</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line444">444</a>       </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line445">445</a>       # Float. Altitude</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line446">446</a>       attr_accessor :alt</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line447">447</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line448">448</a>       # String. Altitude unit.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line449">449</a>       attr_accessor :alt_unit</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line450">450</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line451">451</a>       # Geodedic datum / map datum</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line452">452</a>       attr_accessor :geodetic_datum</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line453">453</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line454">454</a>       def lat=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line455">455</a>         @lat = str.to_f unless str.nil?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line456">456</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line457">457</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line458">458</a>       def long=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line459">459</a>         @long = str.to_f unless str.nil?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line460">460</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line461">461</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line462">462</a>       def alt=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line463">463</a>         @alt = str.to_f unless str.nil?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line464">464</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line465">465</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line466">466</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line467">467</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line468">468</a>         raise &quot;Geodedic datum is a required attribute of Point element.&quot; if @geodetic_datum.nil?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line469">469</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line470">470</a>         p = LibXML::XML::Node.new('point')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line471">471</a>         p[&quot;geodetic_datum&quot;] = @geodetic_datum</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line472">472</a>         p[&quot;alt_unit&quot;] = @alt_unit if @alt_unit != nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line473">473</a>         PhyloXML::Writer.generate_xml(p, self, [</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line474">474</a>             [:simple, 'lat', @lat],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line475">475</a>             [:simple, 'long', @long],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line476">476</a>             [:simple, 'alt', @alt]])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line477">477</a>         return p</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line478">478</a>         #@todo check if characters are correctly generated, like Zuric</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line479">479</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line480">480</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line481">481</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line482">482</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line483">483</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line484">484</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line485">485</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line486">486</a>     # A polygon defined by a list of Points objects.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line487">487</a>     class Polygon</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line488">488</a>       # Array of Point objects.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line489">489</a>       attr_accessor :points</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line490">490</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line491">491</a>       def initialize</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line492">492</a>         @points = []</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line493">493</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line494">494</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line495">495</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line496">496</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line497">497</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line498">498</a>         if @points.length &gt; 2          </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line499">499</a>           pol = LibXML::XML::Node.new('polygon')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line500">500</a>           @points.each do |p|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line501">501</a>             pol &lt;&lt; p.to_xml</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line502">502</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line503">503</a>           return pol</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line504">504</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line505">505</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line506">506</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line507">507</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line508">508</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line509">509</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line510">510</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line511">511</a>     # Element Sequence is used to represent a molecular sequence (Protein, DNA,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line512">512</a>     # RNA) associated with a node.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line513">513</a>     class Sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line514">514</a>       # Type of sequence (rna, dna, protein)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line515">515</a>       attr_accessor :type</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line516">516</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line517">517</a>       # Full name (e.g. muscle Actin )</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line518">518</a>       attr_accessor :name</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line519">519</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line520">520</a>       # String. Used to link with other elements.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line521">521</a>       attr_accessor :id_source</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line522">522</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line523">523</a>       # String. One intended use for 'id_ref' is to link a sequence to a taxonomy</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line524">524</a>       # (via the taxonomy's 'id_source') in the case of multiple sequences and taxonomies per node.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line525">525</a>       attr_accessor :id_ref</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line526">526</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line527">527</a>       # short (maximal ten characters) symbol of the sequence (e.g. 'ACTM')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line528">528</a>       attr_accessor :symbol</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line529">529</a>       # Accession object. Holds source and identifier for the sequence.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line530">530</a>       attr_accessor :accession</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line531">531</a>       # String. Location of a sequence on a genome/chromosome</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line532">532</a>       attr_accessor :location</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line533">533</a>       # String. The actual sequence is stored here.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line534">534</a>       attr_reader :mol_seq</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line535">535</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line536">536</a>       # Boolean. used to indicated that this molecular sequence is aligned with</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line537">537</a>       # all other sequences in the same phylogeny for which 'is aligned' is true</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line538">538</a>       # as well (which, in most cases, means that gaps were introduced, and that</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line539">539</a>       # all sequences for which 'is aligned' is true must have the same length)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line540">540</a>       attr_reader :is_aligned</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line541">541</a>       </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line542">542</a>       # Uri object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line543">543</a>       attr_accessor :uri</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line544">544</a>       # Array of Annotation objects. Annotations of molecular sequence.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line545">545</a>       attr_accessor :annotations</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line546">546</a>       # DomainArchitecture object. Describes domain architecture of a protein.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line547">547</a>       attr_accessor :domain_architecture</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line548">548</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line549">549</a>       # Array of Other objects. Used to save additional information from other than</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line550">550</a>       # PhyloXML namspace.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line551">551</a>       attr_accessor :other</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line552">552</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line553">553</a>       def initialize</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line554">554</a>         @annotations = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line555">555</a>         @other = []</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line556">556</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line557">557</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line558">558</a>       def is_aligned=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line559">559</a>         if str=='true'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line560">560</a>           @is_aligned=true</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line561">561</a>         elsif str=='false'</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line562">562</a>           @is_aligned = false</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line563">563</a>         else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line564">564</a>           @is_aligned = nil</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line565">565</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line566">566</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line567">567</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line568">568</a>       def is_aligned?</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line569">569</a>         @is_aligned</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line570">570</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line571">571</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line572">572</a>       def mol_seq=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line573">573</a>         if str =~ /^[a-zA-Z\.\-\?\*_]+$/</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line574">574</a>           @mol_seq = str</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line575">575</a>         else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line576">576</a>           raise &quot;mol_seq element of Sequence does not follow the pattern.&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line577">577</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line578">578</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line579">579</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line580">580</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line581">581</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line582">582</a>         </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line583">583</a>         seq = LibXML::XML::Node.new('sequence')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line584">584</a>         if @type != nil</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line585">585</a>           if [&quot;dna&quot;, &quot;rna&quot;, &quot;protein&quot;].include?(@type)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line586">586</a>             seq[&quot;type&quot;] = @type</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line587">587</a>           else </pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line588">588</a>             raise &quot;Type attribute of Sequence has to be one of dna, rna or a.&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line589">589</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line590">590</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line591">591</a>         </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line592">592</a>         PhyloXML::Writer.generate_xml(seq, self, [</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line593">593</a>             [:attr, 'id_source'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line594">594</a>             [:attr, 'id_ref'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line595">595</a>             [:pattern, 'symbol', @symbol, Regexp.new(&quot;^\\S{1,10}$&quot;)],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line596">596</a>             [:complex, 'accession', @accession],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line597">597</a>             [:simple, 'name', @name],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line598">598</a>             [:simple, 'location', @location]])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line599">599</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line600">600</a>         if @mol_seq != nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line601">601</a>           molseq = LibXML::XML::Node.new('mol_seq', @mol_seq)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line602">602</a>           molseq[&quot;is_aligned&quot;] = @is_aligned.to_s if @is_aligned != nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line603">603</a>           seq &lt;&lt; molseq</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line604">604</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line605">605</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line606">606</a>         PhyloXML::Writer.generate_xml(seq, self, [</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line607">607</a>             #[:pattern, 'mol_seq', @mol_seq, Regexp.new(&quot;^[a-zA-Z\.\-\?\*_]+$&quot;)],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line608">608</a>             [:complex, 'uri', @uri],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line609">609</a>             [:objarr, 'annotation', 'annotations'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line610">610</a>             [:complex, 'domain_architecture', @domain_architecture]])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line611">611</a>             #@todo test domain_architecture</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line612">612</a>         #any</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line613">613</a>         return seq</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line614">614</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line615">615</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line616">616</a>       # converts Bio::PhyloXML:Sequence to Bio::Sequence object.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line617">617</a>       # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line618">618</a>       # *Returns*:: Bio::Sequence</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line619">619</a>       def to_biosequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line620">620</a>         #type is not a required attribute in phyloxml (nor any other Sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line621">621</a>         #element) it might not hold any value, so we will not check what type it is.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line622">622</a>         seq = Bio::Sequence.auto(@mol_seq)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line623">623</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line624">624</a>         seq.id_namespace = @accession.source</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line625">625</a>         seq.entry_id = @accession.value</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line626">626</a>         # seq.primary_accession = @accession.value could be this</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line627">627</a>         seq.definition = @name</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line628">628</a>         #seq.comments = @name //this one?</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line629">629</a>         if @uri != nil</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line630">630</a>           h = {'url' =&gt; @uri.uri,</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line631">631</a>             'title' =&gt; @uri.desc }</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line632">632</a>           ref = Bio::Reference.new(h)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line633">633</a>           seq.references &lt;&lt; ref</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line634">634</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line635">635</a>         seq.molecule_type = 'RNA' if @type == 'rna'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line636">636</a>         seq.molecule_type = 'DNA' if @type == 'dna'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line637">637</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line638">638</a>         #@todo deal with the properties. There might be properties which look</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line639">639</a>         #like bio sequence attributes or features</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line640">640</a>         return seq</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line641">641</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line642">642</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line643">643</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line644">644</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line645">645</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line646">646</a>     # Element Accession is used to capture the local part in a sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line647">647</a>     # identifier.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line648">648</a>     class Accession</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line649">649</a>       #String. Source of the accession id. Example: &quot;UniProtKB&quot;</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line650">650</a>       attr_accessor :source</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line651">651</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line652">652</a>       #String. Value of the accession id. Example: &quot;P17304&quot;</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line653">653</a>       attr_accessor :value</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line654">654</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line655">655</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line656">656</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line657">657</a>         raise &quot;Source attribute is required for Accession object.&quot; if @source == nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line658">658</a>         accession = LibXML::XML::Node.new('accession', @value)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line659">659</a>         accession['source'] = @source</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line660">660</a>         return accession</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line661">661</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line662">662</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line663">663</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line664">664</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line665">665</a>     # A uniform resource identifier. In general, this is expected to be an URL</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line666">666</a>     # (for example, to link to an image on a website, in which case the 'type'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line667">667</a>     # attribute might be 'image' and 'desc' might be  'image of a California</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line668">668</a>     # sea hare')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line669">669</a>     class Uri</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line670">670</a>       # String. Description of the uri. For example, image of a California sea hare'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line671">671</a>       attr_accessor :desc</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line672">672</a>       # String. For example, image.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line673">673</a>       attr_accessor :type</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line674">674</a>       # String. URL of the resource.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line675">675</a>       attr_accessor :uri </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line676">676</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line677">677</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line678">678</a>       def to_xml        </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line679">679</a>         if @uri != nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line680">680</a>           xml_node = LibXML::XML::Node.new('uri', @uri)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line681">681</a>           Writer.generate_xml(xml_node, self, [</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line682">682</a>             [:attr, 'desc'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line683">683</a>             [:attr, 'type']])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line684">684</a>           return xml_node</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line685">685</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line686">686</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line687">687</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line688">688</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line689">689</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line690">690</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line691">691</a>     # The annotation of a molecular sequence. It is recommended to annotate by</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line692">692</a>     # using the optional 'ref' attribute (some examples of acceptable values</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line693">693</a>     # for the ref attribute: 'GO:0008270', 'KEGG:Tetrachloroethene degradation',</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line694">694</a>     #  'EC:1.1.1.1').</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line695">695</a>     class Annotation</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line696">696</a>       # String. For example, 'GO:0008270', 'KEGG:Tetrachloroethene degradation',</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line697">697</a>       # 'EC:1.1.1.1'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line698">698</a>       attr_accessor :ref</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line699">699</a>       # String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line700">700</a>       attr_accessor :source</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line701">701</a>       # String. evidence for a annotation as free text (e.g. 'experimental')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line702">702</a>       attr_accessor :evidence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line703">703</a>       # String. Type of the annotation.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line704">704</a>       attr_accessor :type</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line705">705</a>       # String. Free text description. </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line706">706</a>       attr_accessor :desc</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line707">707</a>       # Confidence object. Type and value of support for a annotation.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line708">708</a>       attr_accessor :confidence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line709">709</a>       # Array of Property objects. Allows for further, typed and referenced</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line710">710</a>       # annotations from external resources</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line711">711</a>       attr_accessor :properties</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line712">712</a>       # Uri object.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line713">713</a>       attr_accessor :uri</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line714">714</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line715">715</a>       def initialize</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line716">716</a>         #@todo add unit test for this, since didn't break anything when changed from property to properties</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line717">717</a>         @properties = []</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line718">718</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line719">719</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line720">720</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class. </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line721">721</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line722">722</a>         annot = LibXML::XML::Node.new('annotation')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line723">723</a>         annot[&quot;ref&quot;] = @ref if @ref != nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line724">724</a>         PhyloXML::Writer.generate_xml(annot, self, [[:simple, 'desc', @desc],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line725">725</a>           [:complex, 'confidence', @confidence],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line726">726</a>           [:objarr, 'property', 'properties'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line727">727</a>           [:complex, 'uri', @uri]])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line728">728</a>         return annot</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line729">729</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line730">730</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line731">731</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line732">732</a>     class Id</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line733">733</a>       # The provider of Id, for example, NCBI.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line734">734</a>       attr_accessor :provider</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line735">735</a>       # The value of Id. </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line736">736</a>       attr_accessor :value</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line737">737</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line738">738</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line739">739</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line740">740</a>         xml_node = LibXML::XML::Node.new('id', @value)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line741">741</a>         xml_node[&quot;provider&quot;] = @provider if @provider != nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line742">742</a>         return xml_node</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line743">743</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line744">744</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line745">745</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line746">746</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line747">747</a>     # This indicates the color of a node when rendered (the color applies</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line748">748</a>     # to the whole node and its children unless overwritten by the</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line749">749</a>     # color(s) of sub clades).</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line750">750</a>     class BranchColor</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line751">751</a>       #Integer</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line752">752</a>       attr_reader :red, :green, :blue</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line753">753</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line754">754</a>       def red=(str)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line755">755</a>         @red = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line756">756</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line757">757</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line758">758</a>       def green=(str)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line759">759</a>         @green = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line760">760</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line761">761</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line762">762</a>       def blue=(str)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line763">763</a>         @blue = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line764">764</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line765">765</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line766">766</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line767">767</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line768">768</a>         #@todo add unit test</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line769">769</a>         if @red == nil</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line770">770</a>           raise &quot;Subelement red of BranchColor element should not be nil&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line771">771</a>         elsif @green == nil</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line772">772</a>           raise &quot;Subelement green of BranchColor element should not be nil&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line773">773</a>         elsif @blue == nil</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line774">774</a>           raise &quot;Subelement blue of BranchColor element should not be nil&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line775">775</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line776">776</a> </pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line777">777</a>         c = LibXML::XML::Node.new('branch_color')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line778">778</a>         PhyloXML::Writer.generate_xml(c, self, [</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line779">779</a>             [:simple, 'red', @red],</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line780">780</a>             [:simple, 'green', @green],</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line781">781</a>             [:simple, 'blue', @blue]])</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line782">782</a>         return c</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line783">783</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line784">784</a> </pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line785">785</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line786">786</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line787">787</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line788">788</a>     # A date associated with a clade/node. Its value can be numerical by</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line789">789</a>     # using the 'value' element and/or free text with the 'desc' element'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line790">790</a>     # (e.g. 'Silurian'). If a numerical value is used, it is recommended to</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line791">791</a>     # employ the 'unit' attribute to indicate the type of the numerical</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line792">792</a>     # value (e.g. 'mya' for 'million years ago').</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line793">793</a>     class Date</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line794">794</a>       # String. Units in which value is stored.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line795">795</a>       attr_accessor :unit</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line796">796</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line797">797</a>       # Free text description of the date.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line798">798</a>       attr_accessor :desc</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line799">799</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line800">800</a>       # Integer. Minimum and maximum of the value.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line801">801</a>       attr_reader :minimum, :maximum</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line802">802</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line803">803</a>       # Integer. Value of the date.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line804">804</a>       attr_reader :value</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line805">805</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line806">806</a>       def minimum=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line807">807</a>         @minimum = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line808">808</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line809">809</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line810">810</a>       def maximum=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line811">811</a>         @maximum = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line812">812</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line813">813</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line814">814</a>       def value= (str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line815">815</a>         @value = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line816">816</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line817">817</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line818">818</a>       # Returns value + unit, for exampe &quot;7 mya&quot;</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line819">819</a>       def to_s</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line820">820</a>         return &quot;#{value} #{unit}&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line821">821</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line822">822</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line823">823</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line824">824</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line825">825</a>         date = LibXML::XML::Node.new('date')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line826">826</a>         PhyloXML::Writer.generate_xml(date, self, [</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line827">827</a>             [:attr, 'unit'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line828">828</a>             [:simple, 'desc', @desc],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line829">829</a>             [:simple, 'value', @value],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line830">830</a>             [:simple, 'minimum', @minimum],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line831">831</a>             [:simple, 'maximum', @maximum]])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line832">832</a>         return date</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line833">833</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line834">834</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line835">835</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line836">836</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line837">837</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line838">838</a>     # This is used describe the domain architecture of a protein. Attribute</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line839">839</a>     # 'length' is the total length of the protein</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line840">840</a>     class DomainArchitecture</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line841">841</a>       # Integer. Total length of the protein</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line842">842</a>       attr_accessor :length</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line843">843</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line844">844</a>       # Array of ProteinDomain objects.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line845">845</a>       attr_reader :domains</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line846">846</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line847">847</a>       def length=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line848">848</a>         @length = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line849">849</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line850">850</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line851">851</a>       def initialize</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line852">852</a>         @domains = []</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line853">853</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line854">854</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line855">855</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line856">856</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line857">857</a>         xml_node = LibXML::XML::Node.new('domain_architecture')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line858">858</a>         PhyloXML::Writer.generate_xml(xml_node, self,[</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line859">859</a>               [:attr, 'length'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line860">860</a>               [:objarr, 'domain', 'domains']])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line861">861</a>         return xml_node</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line862">862</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line863">863</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line864">864</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line865">865</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line866">866</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line867">867</a>     # To represent an individual domain in a domain architecture. The</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line868">868</a>     # name/unique identifier is described via the 'id' attribute.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line869">869</a>     class ProteinDomain</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line870">870</a>       #Float, for example to store E-values    4.7E-14</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line871">871</a>       attr_accessor :confidence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line872">872</a>       </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line873">873</a>       # String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line874">874</a>       attr_accessor :id, :value</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line875">875</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line876">876</a>       # Integer. Beginning of the domain.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line877">877</a>       attr_reader :from</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line878">878</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line879">879</a>       # Integer. End of the domain.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line880">880</a>       attr_reader :to</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line881">881</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line882">882</a>       def from=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line883">883</a>         @from = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line884">884</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line885">885</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line886">886</a>       def to=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line887">887</a>         @to = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line888">888</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line889">889</a>       </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line890">890</a>       def confidence=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line891">891</a>         @confidence = str.to_f</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line892">892</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line893">893</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line894">894</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line895">895</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line896">896</a>         if @from == nil</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line897">897</a>           raise &quot;from attribute of ProteinDomain class is required.&quot;</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line898">898</a>         elsif @to == nil</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line899">899</a>           raise &quot;to attribute of ProteinDomain class is required.&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line900">900</a>         else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line901">901</a>           xml_node = LibXML::XML::Node.new('domain', @value)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line902">902</a>           xml_node[&quot;from&quot;] = @from.to_s</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line903">903</a>           xml_node[&quot;to&quot;] = @to.to_s</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line904">904</a>           xml_node[&quot;id&quot;] = @id if @id != nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line905">905</a>           xml_node[&quot;confidence&quot;] = @confidence.to_s</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line906">906</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line907">907</a>           return xml_node</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line908">908</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line909">909</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line910">910</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line911">911</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line912">912</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line913">913</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line914">914</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line915">915</a>     #Property allows for typed and referenced properties from external resources</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line916">916</a>     #to be attached to 'Phylogeny', 'Clade', and 'Annotation'. The value of a</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line917">917</a>     #property is its mixed (free text) content. Attribute 'datatype' indicates</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line918">918</a>     #the type of a property and is limited to xsd-datatypes (e.g. 'xsd:string',</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line919">919</a>     #'xsd:boolean', 'xsd:integer', 'xsd:decimal', 'xsd:float', 'xsd:double',</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line920">920</a>     #'xsd:date', 'xsd:anyURI'). Attribute 'applies_to' indicates the item to</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line921">921</a>     #which a property applies to (e.g. 'node' for the parent node of a clade,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line922">922</a>     #'parent_branch' for the parent branch of a clade). Attribute 'id_ref' allows</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line923">923</a>     #to attached a property specifically to one element (on the xml-level).</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line924">924</a>     #Optional attribute 'unit' is used to indicate the unit of the property.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line925">925</a>     #An example: &lt;property datatype=&quot;xsd:integer&quot; ref=&quot;NOAA:depth&quot; applies_to=&quot;clade&quot; unit=&quot;METRIC:m&quot;&gt; 200 &lt;/property&gt;</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line926">926</a>     class Property</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line927">927</a>       # String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line928">928</a>       attr_accessor :ref, :unit, :id_ref, :value</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line929">929</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line930">930</a>       # String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line931">931</a>       attr_reader :datatype, :applies_to</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line932">932</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line933">933</a>       def datatype=(str)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line934">934</a>          #@todo add unit test or maybe remove, if assume that xml is valid.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line935">935</a>         unless ['xsd:string','xsd:boolean','xsd:decimal','xsd:float','xsd:double',</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line936">936</a>             'xsd:duration','xsd:dateTime','xsd:time','xsd:date','xsd:gYearMonth',</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line937">937</a>             'xsd:gYear','xsd:gMonthDay','xsd:gDay','xsd:gMonth','xsd:hexBinary',</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line938">938</a>             'xsd:base64Binary','xsd:anyURI','xsd:normalizedString','xsd:token',</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line939">939</a>             'xsd:integer','xsd:nonPositiveInteger','xsd:negativeInteger',</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line940">940</a>             'xsd:long','xsd:int','xsd:short','xsd:byte','xsd:nonNegativeInteger',</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line941">941</a>             'xsd:unsignedLong','xsd:unsignedInt','xsd:unsignedShort',</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line942">942</a>             'xsd:unsignedByte','xsd:positiveInteger'].include?(str)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line943">943</a>           raise &quot;Warning: #{str} is not in the list of allowed values.&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line944">944</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line945">945</a>         @datatype = str</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line946">946</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line947">947</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line948">948</a>       def applies_to=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line949">949</a>         unless ['phylogeny','clade','node','annotation','parent_branch','other'].include?(str)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line950">950</a>           puts &quot;Warning: #{str} is not in the list of allowed values.&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line951">951</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line952">952</a>         @applies_to = str</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line953">953</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line954">954</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line955">955</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line956">956</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line957">957</a>         #@todo  write unit test for this</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line958">958</a>         raise &quot;ref is an required element of property&quot;  if @ref.nil?</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line959">959</a>         raise &quot;datatype is an required element of property&quot; if @datatype.nil?</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line960">960</a>         raise &quot;applies_to is an required element of property&quot; if @applies_to.nil?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line961">961</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line962">962</a>         property = LibXML::XML::Node.new('property')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line963">963</a>         Writer.generate_xml(property, self, [</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line964">964</a>             [:attr, 'ref'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line965">965</a>             [:attr, 'unit'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line966">966</a>             [:attr, 'datatype'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line967">967</a>             [:attr, 'applies_to'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line968">968</a>             [:attr, 'id_ref']])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line969">969</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line970">970</a>         property &lt;&lt; @value if @value != nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line971">971</a>         return property</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line972">972</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line973">973</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line974">974</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line975">975</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line976">976</a>     # A literature reference for a clade. It is recommended to use the 'doi'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line977">977</a>     # attribute instead of the free text 'desc' element whenever possible.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line978">978</a>     class Reference</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line979">979</a>       # String. Digital Object Identifier.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line980">980</a>       attr_accessor :doi</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line981">981</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line982">982</a>       # String. Free text description.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line983">983</a>       attr_accessor :desc</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line984">984</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line985">985</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line986">986</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line987">987</a>         ref = LibXML::XML::Node.new('reference')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line988">988</a>         Writer.generate_xml(ref, self, [</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line989">989</a>               [:attr, 'doi'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line990">990</a>               [:simple, 'desc', @desc]])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line991">991</a>          return ref</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line992">992</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line993">993</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line994">994</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line995">995</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line996">996</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line997">997</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line998">998</a>     # This is used to express a typed relationship between two clades.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line999">999</a>     # For example it could be used to describe multiple parents of a clade.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1000">1000</a>     class CladeRelation</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1001">1001</a>       # Float</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1002">1002</a>       attr_accessor :distance</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1003">1003</a>       # String. Id of the referenced parents of a clade.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1004">1004</a>       attr_accessor :id_ref_0, :id_ref_1</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1005">1005</a>       # String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1006">1006</a>       attr_accessor :type</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1007">1007</a>       # Confidence object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1008">1008</a>       attr_accessor :confidence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1009">1009</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1010">1010</a>       def distance=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1011">1011</a>         @distance = str.to_f</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1012">1012</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1013">1013</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1014">1014</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1015">1015</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1016">1016</a>         if @id_ref_0 == nil or @id_ref_1 == nil or @type == nil</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line1017">1017</a>           raise &quot;Attributes id_ref_0, id_ref_1, type are required elements by SequenceRelation element.&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1018">1018</a>         else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1019">1019</a>           cr = LibXML::XML::Node.new('clade_relation')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1020">1020</a>           Writer.generate_xml(cr, self, [</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1021">1021</a>               [:attr, 'id_ref_0'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1022">1022</a>               [:attr, 'id_ref_1'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1023">1023</a>               [:attr, 'distance'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1024">1024</a>               [:attr, 'type'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1025">1025</a>               [:complex, 'confidence', @confidnece]])         </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1026">1026</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1027">1027</a>           return cr</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1028">1028</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1029">1029</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1030">1030</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1031">1031</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1032">1032</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1033">1033</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1034">1034</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1035">1035</a>     # The names and/or counts of binary characters present, gained, and</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1036">1036</a>     # lost at the root of a clade.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1037">1037</a>     class BinaryCharacters</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1038">1038</a>       attr_accessor :bc_type, :gained, :lost, :present, :absent</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1039">1039</a>       attr_reader :gained_count, :lost_count, :present_count, :absent_count</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1040">1040</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1041">1041</a>       def gained_count=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1042">1042</a>         @gained_count = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1043">1043</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1044">1044</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1045">1045</a>       def lost_count=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1046">1046</a>         @lost_count = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1047">1047</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1048">1048</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1049">1049</a>       def present_count=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1050">1050</a>         @present_count = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1051">1051</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1052">1052</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1053">1053</a>       def absent_count=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1054">1054</a>         @absent_count = str.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1055">1055</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1056">1056</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1057">1057</a>       def initialize</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1058">1058</a>         @gained = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1059">1059</a>         @lost = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1060">1060</a>         @present = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1061">1061</a>         @absent = []</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1062">1062</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1063">1063</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1064">1064</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1065">1065</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1066">1066</a>         bc = LibXML::XML::Node.new('binary_characters')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1067">1067</a>         bc['type'] = @bc_type</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1068">1068</a>         PhyloXML::Writer.generate_xml(bc, self, [</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1069">1069</a>             [:attr, 'gained_count'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1070">1070</a>             [:attr, 'lost_count'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1071">1071</a>             [:attr, 'present_count'],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1072">1072</a>             [:attr, 'absent_count']])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1073">1073</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1074">1074</a>         if not @gained.empty?</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1075">1075</a>           gained_xml = LibXML::XML::Node.new('gained')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1076">1076</a>           PhyloXML::Writer.generate_xml(gained_xml, self, [[:simplearr, 'bc', @gained]])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1077">1077</a>           bc &lt;&lt; gained_xml</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1078">1078</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1079">1079</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1080">1080</a>         if not @lost.empty?</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line1081">1081</a>           lost_xml = LibXML::XML::Node.new('lost')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line1082">1082</a>           PhyloXML::Writer.generate_xml(lost_xml, self, [[:simplearr, 'bc', @lost]])</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line1083">1083</a>           bc &lt;&lt; lost_xml</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line1084">1084</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1085">1085</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1086">1086</a>         if not @present.empty?</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1087">1087</a>           present_xml = LibXML::XML::Node.new('present')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1088">1088</a>           PhyloXML::Writer.generate_xml(present_xml, self, [[:simplearr, 'bc', @present]])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1089">1089</a>           bc &lt;&lt; present_xml</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1090">1090</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1091">1091</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1092">1092</a>         if not @absent.empty?</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line1093">1093</a>           absent_xml = LibXML::XML::Node.new('absent')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line1094">1094</a>           PhyloXML::Writer.generate_xml(absent_xml, self, [[:simplearr, 'bc', @absent]])</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line1095">1095</a>           bc &lt;&lt; absent_xml</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line1096">1096</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1097">1097</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1098">1098</a>         return bc</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1099">1099</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1100">1100</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1101">1101</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1102">1102</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1103">1103</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1104">1104</a>     # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1105">1105</a>     # This is used to express a typed relationship between two sequences.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1106">1106</a>     # For example it could be used to describe an orthology (in which case</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1107">1107</a>     # attribute 'type' is 'orthology').</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1108">1108</a>     class SequenceRelation</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1109">1109</a>       # String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1110">1110</a>       attr_accessor :id_ref_0, :id_ref_1, :type</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1111">1111</a>       # Float</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1112">1112</a>       attr_reader :distance</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1113">1113</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1114">1114</a>       #@todo it has Confidences objects.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1115">1115</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1116">1116</a>       def distance=(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1117">1117</a>         @distance = str.to_f if str != nil</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1118">1118</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1119">1119</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1120">1120</a>       def type=(str)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1121">1121</a>         #@todo do warning instead?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1122">1122</a>         #@todo do validation at actually writing xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1123">1123</a>         allowed_values = [&quot;orthology&quot;, &quot;one_to_one_orthology&quot;, &quot;super_orthology&quot;, &quot;paralogy&quot;,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1124">1124</a>             &quot;ultra_paralogy&quot;, &quot;xenology&quot;, &quot;unknown&quot;, &quot;other&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1125">1125</a>         if not allowed_values.include? str</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line1126">1126</a>           raise &quot;SequenceRelation#type has to be one one of #{allowed_values.join(&quot;; &quot;)}&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1127">1127</a>         else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1128">1128</a>           @type = str</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1129">1129</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1130">1130</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1131">1131</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1132">1132</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1133">1133</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1134">1134</a>         if @id_ref_0 == nil or @id_ref_1 == nil or @type == nil</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line1135">1135</a>           raise &quot;Attributes id_ref_0, id_ref_1, type are required elements by SequenceRelation element.&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1136">1136</a>         else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1137">1137</a>           sr = LibXML::XML::Node.new('sequence_relation')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1138">1138</a>           sr['id_ref_0'] = @id_ref_0</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1139">1139</a>           sr['id_ref_1'] = @id_ref_1</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1140">1140</a>           sr['distance'] = @distance.to_s if @distance != nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1141">1141</a>           sr['type'] = @type</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1142">1142</a>           return sr</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1143">1143</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1144">1144</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1145">1145</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1146">1146</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1147">1147</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1148">1148</a>    class Other</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1149">1149</a>       attr_accessor :element_name, :attributes, :children, :value</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1150">1150</a>       </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1151">1151</a>       def initialize</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1152">1152</a>         @children = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1153">1153</a>         @attributes = Hash.new</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1154">1154</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1155">1155</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1156">1156</a>       # Converts elements to xml representation. Called by PhyloXML::Writer class.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1157">1157</a>       def to_xml</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1158">1158</a>         o = LibXML::XML::Node.new(@element_name)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1159">1159</a>         @attributes.each do |key, value|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1160">1160</a>           o[key] = value</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1161">1161</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1162">1162</a>         o &lt;&lt; value if value != nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1163">1163</a>         children.each do |child_node|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1164">1164</a>           o &lt;&lt; child_node.to_xml</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1165">1165</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line1166">1166</a>         return o</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1167">1167</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1168">1168</a>       </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1169">1169</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1170">1170</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1171">1171</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1172">1172</a> end #module PhyloXML</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1173">1173</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1174">1174</a> end #end module Bio</pre></td>
          </tr>
        
      </tbody>
    </table>

    <p>Generated on Fri Jun 25 17:37:23 +0900 2010 with <a href="http://github.com/relevance/rcov">rcov 0.9.8</a></p>

  </body>
</html>
