<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang='en' xml:lang='en' xmlns='http://www.w3.org/1999/xhtml'>
  <head>
    <title>lib/bio/db/phyloxml/phyloxml_parser.rb</title>
    <link href="screen.css" media="all" rel="stylesheet" type="text/css" />
    <link href="print.css" media="print" rel="stylesheet" type="text/css" />
    
    <meta http-equiv="Content-type" content="text/html;charset=UTF-8" />
    <script type="text/javascript" src="rcov.js"></script>
  </head>
  <body>
    <h1>Bioruby C0 Coverage Information - RCov</h1>
    <h2>lib/bio/db/phyloxml/phyloxml_parser.rb</h2>

    

    <div class="report_table_wrapper">
      <table class='report' id='report_table'>
        <thead>
          <tr>
            <th class="left_align">Name</th>
            <th class="right_align">Total Lines</th>
            <th class="right_align">Lines of Code</th>
            <th class="left_align">Total Coverage</th>
            <th class="left_align">Code Coverage</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td class="left_align"><a href="lib-bio-db-phyloxml-phyloxml_parser_rb.html">lib/bio/db/phyloxml/phyloxml_parser.rb</a></td>
            <td class='right_align'><tt>954</tt></td>
            <td class='right_align'><tt>587</tt></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>89.20%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:89px"></div>
            <div class="uncovered" style="width:11px"></div>
          </div></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>85.52%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:86px"></div>
            <div class="uncovered" style="width:14px"></div>
          </div></td>
          </tr>
        </tbody>
      </table>
    </div>
    
    <h3>Key</h3>
    
    <div class="key"><pre><span class='marked'>Code reported as executed by Ruby looks like this...</span><span class='marked1'>and this: this line is also marked as covered.</span><span class='inferred'>Lines considered as run by rcov, but not reported by Ruby, look like this,</span><span class='inferred1'>and this: these lines were inferred by rcov (using simple heuristics).</span><span class='uncovered'>Finally, here's a line marked as not executed.</span></pre></div>

    <h3>Coverage Details</h3>

    <table class="details">
      <tbody>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1">1</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line2">2</a> # = bio/db/phyloxml_parser.rb - PhyloXML parser</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line3">3</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line4">4</a> # Copyright::   Copyright (C) 2009</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line5">5</a> #               Diana Jaunzeikare &lt;latvianlinuxgirl@gmail.com&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line6">6</a> # License::     The Ruby License</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line7">7</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line8">8</a> # $Id:$</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line9">9</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line10">10</a> # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line11">11</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line12">12</a> # This file containts parser for PhyloXML.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line13">13</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line14">14</a> # == Requirements</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line15">15</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line16">16</a> # Libxml2 XML parser is required. Install libxml-ruby bindings from</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line17">17</a> # http://libxml.rubyforge.org or</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line18">18</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line19">19</a> #   gem install -r libxml-ruby</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line20">20</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line21">21</a> # == References</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line22">22</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line23">23</a> # * http://www.phyloxml.org</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line24">24</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line25">25</a> # * https://www.nescent.org/wg_phyloinformatics/PhyloSoC:PhyloXML_support_in_BioRuby</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line26">26</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line27">27</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line28">28</a> require 'uri'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line29">29</a> require 'libxml'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line30">30</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line31">31</a> require 'bio/tree'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line32">32</a> require 'bio/db/phyloxml/phyloxml_elements'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line33">33</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line34">34</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line35">35</a> module Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line36">36</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line37">37</a> module PhyloXML</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line38">38</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line39">39</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line40">40</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line41">41</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line42">42</a>   # == Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line43">43</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line44">44</a>   # Bio::PhyloXML::Parser is for parsing phyloXML format files.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line45">45</a>   # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line46">46</a>   # == Requirements</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line47">47</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line48">48</a>   # Libxml2 XML parser is required. Install libxml-ruby bindings from</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line49">49</a>   # http://libxml.rubyforge.org or</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line50">50</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line51">51</a>   #   gem install -r libxml-ruby</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line52">52</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line53">53</a>   # == Usage</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line54">54</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line55">55</a>   #   require 'bio'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line56">56</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line57">57</a>   #  # Create new phyloxml parser</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line58">58</a>   #  phyloxml = Bio::PhyloXML::Parser.open('example.xml')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line59">59</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line60">60</a>   #  # Print the names of all trees in the file</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line61">61</a>   #  phyloxml.each do |tree|</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line62">62</a>   #    puts tree.name</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line63">63</a>   #  end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line64">64</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line65">65</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line66">66</a>   # == References</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line67">67</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line68">68</a>   # http://www.phyloxml.org/documentation/version_100/phyloxml.xsd.html</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line69">69</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line70">70</a>   class Parser</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line71">71</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line72">72</a>     include LibXML</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line73">73</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line74">74</a>     # After parsing all the trees, if there is anything else in other xml format,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line75">75</a>     # it is saved in this array of PhyloXML::Other objects</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line76">76</a>     attr_reader :other</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line77">77</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line78">78</a>     # Initializes LibXML::Reader and reads the file until it reaches the first</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line79">79</a>     # phylogeny element.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line80">80</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line81">81</a>     # Create a new Bio::PhyloXML::Parser object.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line82">82</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line83">83</a>     #   p = Bio::PhyloXML::Parser.open(&quot;./phyloxml_examples.xml&quot;)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line84">84</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line85">85</a>     # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line86">86</a>     # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line87">87</a>     # * (required) _filename_: Path to the file to parse.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line88">88</a>     # * (optional) _validate_: Whether to validate the file against schema or not. Default value is true.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line89">89</a>     # *Returns*:: Bio::PhyloXML::Parser object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line90">90</a>     def self.open(filename, validate=true)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line91">91</a>       obj = new(nil, validate)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line92">92</a>       obj.instance_eval {</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line93">93</a>         filename = _secure_filename(filename)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line94">94</a>         _validate(:file, filename) if validate</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line95">95</a>         # XML::Parser::Options::NONET for security reason</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line96">96</a>         @reader = XML::Reader.file(filename,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line97">97</a>                                    { :options =&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line98">98</a>                                      LibXML::XML::Parser::Options::NONET })</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line99">99</a>         _skip_leader</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line100">100</a>       }</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line101">101</a>       obj</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line102">102</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line103">103</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line104">104</a>     # Initializes LibXML::Reader and reads the file until it reaches the first</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line105">105</a>     # phylogeny element.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line106">106</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line107">107</a>     # Create a new Bio::PhyloXML::Parser object.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line108">108</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line109">109</a>     #   p = Bio::PhyloXML::Parser.open_uri(&quot;http://www.phyloxml.org/examples/apaf.xml&quot;)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line110">110</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line111">111</a>     # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line112">112</a>     # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line113">113</a>     # * (required) _uri_: (URI or String) URI to the data to parse</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line114">114</a>     # * (optional) _validate_: For URI reader, the &quot;validate&quot; option is ignored and no validation is executed.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line115">115</a>     # *Returns*:: Bio::PhyloXML::Parser object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line116">116</a>     def self.open_uri(uri, validate=true)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line117">117</a>       case uri</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line118">118</a>       when URI</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line119">119</a>         uri = uri.to_s</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line120">120</a>       else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line121">121</a>         # raises error if not a String</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line122">122</a>         uri = uri.to_str</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line123">123</a>         # raises error if invalid URI</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line124">124</a>         URI.parse(uri)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line125">125</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line126">126</a> </pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line127">127</a>       obj = new(nil, validate)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line128">128</a>       obj.instance_eval {</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line129">129</a>         @reader = XML::Reader.file(uri)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line130">130</a>         _skip_leader</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line131">131</a>       }</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line132">132</a>       obj</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line133">133</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line134">134</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line135">135</a>     # Special class for closed PhyloXML::Parser object.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line136">136</a>     # It raises error for any methods except essential methods.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line137">137</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line138">138</a>     # Bio::PhyloXML internal use only.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line139">139</a>     class ClosedPhyloXMLParser #:nodoc:</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line140">140</a>       def method_missing(*arg)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line141">141</a>         raise LibXML::XML::Error, 'closed PhyloXML::Parser object'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line142">142</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line143">143</a>     end #class ClosedPhyloXMLParser</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line144">144</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line145">145</a>     # Closes the LibXML::Reader inside the object.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line146">146</a>     # It also closes the opened file if it is created by using</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line147">147</a>     # Bio::PhyloXML::Parser.open method.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line148">148</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line149">149</a>     # When closed object is closed again, or closed object is used,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line150">150</a>     # it raises LibXML::XML::Error.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line151">151</a>     # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line152">152</a>     # *Returns*:: nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line153">153</a>     def close</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line154">154</a>       @reader.close</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line155">155</a>       @reader = ClosedPhyloXMLParser.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line156">156</a>       nil</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line157">157</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line158">158</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line159">159</a>     # Initializes LibXML::Reader and reads from the IO until it reaches</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line160">160</a>     # the first phylogeny element.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line161">161</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line162">162</a>     # Create a new Bio::PhyloXML::Parser object.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line163">163</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line164">164</a>     #   p = Bio::PhyloXML::Parser.for_io($stdin)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line165">165</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line166">166</a>     # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line167">167</a>     # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line168">168</a>     # * (required) _io_: IO object</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line169">169</a>     # * (optional) _validate_: For IO reader, the &quot;validate&quot; option is ignored and no validation is executed.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line170">170</a>     # *Returns*:: Bio::PhyloXML::Parser object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line171">171</a>     def self.for_io(io, validate=true)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line172">172</a>       obj = new(nil, validate)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line173">173</a>       obj.instance_eval {</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line174">174</a>         @reader = XML::Reader.io(io,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line175">175</a>                                  { :options =&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line176">176</a>                                    LibXML::XML::Parser::Options::NONET })</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line177">177</a>         _skip_leader</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line178">178</a>       }</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line179">179</a>       obj</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line180">180</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line181">181</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line182">182</a>     # (private) returns PhyloXML schema</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line183">183</a>     def _schema</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line184">184</a>       XML::Schema.document(XML::Document.file(File.join(File.dirname(__FILE__),'phyloxml.xsd')))</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line185">185</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line186">186</a>     private :_schema</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line187">187</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line188">188</a>     # (private) do validation</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line189">189</a>     # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line190">190</a>     # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line191">191</a>     # * (required) &lt;em&gt;data_type&lt;/em&gt;_: :file for filename, :string for string</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line192">192</a>     # * (required) _arg_: filename or string</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line193">193</a>     # *Returns*:: (undefined)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line194">194</a>     def _validate(data_type, arg)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line195">195</a>       options = { :options =&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line196">196</a>         (LibXML::XML::Parser::Options::NOERROR |   # no error messages</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line197">197</a>          LibXML::XML::Parser::Options::NOWARNING | # no warning messages</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line198">198</a>          LibXML::XML::Parser::Options::NONET)      # no network access</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line199">199</a>       }</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line200">200</a>       case data_type</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line201">201</a>       when :file</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line202">202</a>         # No validation when special file e.g. FIFO (named pipe)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line203">203</a>         return unless File.file?(arg)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line204">204</a>         xml_instance = XML::Document.file(arg, options)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line205">205</a>       when :string</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line206">206</a>         xml_instance = XML::Document.string(arg, options)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line207">207</a>       else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line208">208</a>         # no validation for unknown data type</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line209">209</a>         return</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line210">210</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line211">211</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line212">212</a>       schema = _schema</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line213">213</a>       begin</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line214">214</a>         flag = xml_instance.validate_schema(schema) do |msg, flag|</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line215">215</a>           # The document of libxml-ruby says that the block is called</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line216">216</a>           # when validation failed, but it seems it is never called</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line217">217</a>           # even when validation failed!</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line218">218</a>           raise &quot;Validation of the XML document against phyloxml.xsd schema failed. #{msg}&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line219">219</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line220">220</a>       rescue LibXML::XML::Error =&gt; evar</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line221">221</a>         raise &quot;Validation of the XML document against phyloxml.xsd schema failed, or XML error occurred. #{evar.message}&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line222">222</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line223">223</a>       unless flag then</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line224">224</a>         raise &quot;Validation of the XML document against phyloxml.xsd schema failed.&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line225">225</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line226">226</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line227">227</a>     private :_validate</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line228">228</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line229">229</a>     # (private) It seems that LibXML::XML::Reader reads from the network</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line230">230</a>     # even if LibXML::XML::Parser::Options::NONET is set.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line231">231</a>     # So, for URI-like filename, '://' is replaced with ':/'.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line232">232</a>     def _secure_filename(filename)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line233">233</a>       # for safety, URI-like filename is checked.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line234">234</a>       if /\A[a-zA-Z]+\:\/\// =~ filename then</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line235">235</a>         # for example, &quot;http://a/b&quot; is changed to &quot;http:/a/b&quot;.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line236">236</a>         filename = filename.sub(/\:\/\//, ':/')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line237">237</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line238">238</a>       filename</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line239">239</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line240">240</a>     private :_secure_filename</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line241">241</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line242">242</a>     # (private) loops through until reaches phylogeny stuff</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line243">243</a>     def _skip_leader</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line244">244</a>       #loops through until reaches phylogeny stuff</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line245">245</a>       # Have to leave this way, if accepting strings, instead of files</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line246">246</a>       @reader.read until is_element?('phylogeny')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line247">247</a>       nil</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line248">248</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line249">249</a>     private :_skip_leader</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line250">250</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line251">251</a>     # Initializes LibXML::Reader and reads the PhyloXML-formatted string</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line252">252</a>     # until it reaches the first phylogeny element.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line253">253</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line254">254</a>     # Create a new Bio::PhyloXML::Parser object.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line255">255</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line256">256</a>     #   str = File.read(&quot;./phyloxml_examples.xml&quot;)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line257">257</a>     #   p = Bio::PhyloXML::Parser.new(str)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line258">258</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line259">259</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line260">260</a>     # Deprecated usage: Reads data from a file. &lt;em&gt;str&lt;em&gt; is a filename.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line261">261</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line262">262</a>     #   p = Bio::PhyloXML::Parser.new(&quot;./phyloxml_examples.xml&quot;)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line263">263</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line264">264</a>     # Taking filename is deprecated. Use Bio::PhyloXML::Parser.open(filename).</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line265">265</a>     # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line266">266</a>     # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line267">267</a>     # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line268">268</a>     # * (required) _str_: PhyloXML-formatted string</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line269">269</a>     # * (optional) _validate_: Whether to validate the file against schema or not. Default value is true.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line270">270</a>     # *Returns*:: Bio::PhyloXML::Parser object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line271">271</a>     def initialize(str, validate=true)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line272">272</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line273">273</a>       @other = []</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line274">274</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line275">275</a>       return unless str</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line276">276</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line277">277</a>       # For compatibility, if filename-like string is given,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line278">278</a>       # treat it as a filename.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line279">279</a>       if /[\&lt;\&gt;\r\n]/ !~ str and File.exist?(str) then</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line280">280</a>         # assume that str is filename</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line281">281</a>         warn &quot;Bio::PhyloXML::Parser.new(filename) is deprecated. Use Bio::PhyloXML::Parser.open(filename).&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line282">282</a>         filename = _secure_filename(str)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line283">283</a>         _validate(:file, filename) if validate</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line284">284</a>         @reader = XML::Reader.file(filename)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line285">285</a>         _skip_leader</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line286">286</a>         return</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line287">287</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line288">288</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line289">289</a>       # initialize for string</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line290">290</a>       @reader = XML::Reader.string(str,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line291">291</a>                                    { :options =&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line292">292</a>                                      LibXML::XML::Parser::Options::NONET })</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line293">293</a>       _skip_leader</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line294">294</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line295">295</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line296">296</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line297">297</a>     # Iterate through all trees in the file.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line298">298</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line299">299</a>     #  phyloxml = Bio::PhyloXML::Parser.open('example.xml')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line300">300</a>     #  phyloxml.each do |tree|</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line301">301</a>     #    puts tree.name</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line302">302</a>     #  end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line303">303</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line304">304</a>     def each</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line305">305</a>       while tree = next_tree</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line306">306</a>         yield tree</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line307">307</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line308">308</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line309">309</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line310">310</a>     # Access the specified tree in the file. It parses trees until the specified</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line311">311</a>     # tree is reached.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line312">312</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line313">313</a>     #  # Get 3rd tree in the file (starts counting from 0).</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line314">314</a>     #  parser = PhyloXML::Parser.open('phyloxml_examples.xml')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line315">315</a>     #  tree = parser[2]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line316">316</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line317">317</a>     def [](i)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line318">318</a>       tree = nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line319">319</a>       (i+1).times do</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line320">320</a>        tree =  self.next_tree</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line321">321</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line322">322</a>       return tree</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line323">323</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line324">324</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line325">325</a>     # Parse and return the next phylogeny tree. If there are no more phylogeny</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line326">326</a>     # element, nil is returned. If there is something else besides phylogeny</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line327">327</a>     # elements, it is saved in the PhyloXML::Parser#other.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line328">328</a>     # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line329">329</a>     #  p = Bio::PhyloXML::Parser.open(&quot;./phyloxml_examples.xml&quot;)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line330">330</a>     #  tree = p.next_tree</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line331">331</a>     #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line332">332</a>     # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line333">333</a>     # *Returns*:: Bio::PhyloXML::Tree</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line334">334</a>     def next_tree()</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line335">335</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line336">336</a>       if not is_element?('phylogeny')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line337">337</a>         if @reader.node_type == XML::Reader::TYPE_END_ELEMENT</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line338">338</a>           if is_end_element?('phyloxml')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line339">339</a>             return nil</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line340">340</a>           else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line341">341</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line342">342</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line343">343</a>             if is_end_element?('phyloxml')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line344">344</a>               return nil</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line345">345</a>             end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line346">346</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line347">347</a>         end        </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line348">348</a>         # phyloxml can hold only phylogeny and &quot;other&quot; elements. If this is not</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line349">349</a>         # phylogeny element then it is other. Also, &quot;other&quot; always comes after</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line350">350</a>         # all phylogenies        </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line351">351</a>         @other &lt;&lt; parse_other        </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line352">352</a>         #return nil for tree, since this is not valid phyloxml tree.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line353">353</a>         return nil</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line354">354</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line355">355</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line356">356</a>       tree = Bio::PhyloXML::Tree.new</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line357">357</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line358">358</a>       # keep track of current node in clades array/stack. Current node is the</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line359">359</a>       # last element in the clades array</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line360">360</a>       clades = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line361">361</a>       clades.push tree</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line362">362</a>       </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line363">363</a>       #keep track of current edge to be able to parse branch_length tag</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line364">364</a>       current_edge = nil</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line365">365</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line366">366</a>       # we are going to parse clade iteratively by pointing (and changing) to</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line367">367</a>       # the current node in the tree. Since the property element is both in</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line368">368</a>       # clade and in the phylogeny, we need some boolean to know if we are</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line369">369</a>       # parsing the clade (there can be only max 1 clade in phylogeny) or</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line370">370</a>       # parsing phylogeny</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line371">371</a>       parsing_clade = false</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line372">372</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line373">373</a>       while not is_end_element?('phylogeny') do</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line374">374</a>         break if is_end_element?('phyloxml')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line375">375</a>         </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line376">376</a>         # parse phylogeny elements, except clade</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line377">377</a>         if not parsing_clade</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line378">378</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line379">379</a>           if is_element?('phylogeny')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line380">380</a>             @reader[&quot;rooted&quot;] == &quot;true&quot; ? tree.rooted = true : tree.rooted = false</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line381">381</a>             @reader[&quot;rerootable&quot;] == &quot;true&quot; ? tree.rerootable = true : tree.rerootable = false</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line382">382</a>             parse_attributes(tree, [&quot;branch_length_unit&quot;, 'type'])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line383">383</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line384">384</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line385">385</a>           parse_simple_elements(tree, [ &quot;name&quot;, 'description', &quot;date&quot;])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line386">386</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line387">387</a>           if is_element?('confidence')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line388">388</a>             tree.confidences &lt;&lt; parse_confidence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line389">389</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line390">390</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line391">391</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line392">392</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line393">393</a>         if @reader.node_type == XML::Reader::TYPE_ELEMENT</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line394">394</a>           case @reader.name</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line395">395</a>           when 'clade'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line396">396</a>             #parse clade element</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line397">397</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line398">398</a>             parsing_clade = true</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line399">399</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line400">400</a>             node= Bio::PhyloXML::Node.new</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line401">401</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line402">402</a>             branch_length = @reader['branch_length']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line403">403</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line404">404</a>             parse_attributes(node, [&quot;id_source&quot;])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line405">405</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line406">406</a>             #add new node to the tree</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line407">407</a>             tree.add_node(node)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line408">408</a>             # The first clade will always be root since by xsd schema phyloxml can</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line409">409</a>             # have 0 to 1 clades in it.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line410">410</a>             if tree.root == nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line411">411</a>               tree.root = node</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line412">412</a>             else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line413">413</a>               current_edge = tree.add_edge(clades[-1], node,</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line414">414</a>                                            Bio::Tree::Edge.new(branch_length))</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line415">415</a>             end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line416">416</a>             clades.push node</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line417">417</a>             #end if clade element</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line418">418</a>           else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line419">419</a>            parse_clade_elements(clades[-1], current_edge) if parsing_clade</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line420">420</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line421">421</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line422">422</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line423">423</a>         #end clade element, go one parent up</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line424">424</a>         if is_end_element?('clade')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line425">425</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line426">426</a>            #if we have reached the closing tag of the top-most clade, then our</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line427">427</a>           # curent node should point to the root, If thats the case, we are done</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line428">428</a>           # parsing the clade element</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line429">429</a>           if clades[-1] == tree.root</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line430">430</a>             parsing_clade = false</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line431">431</a>           else</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line432">432</a>             # set current node (clades[-1) to the previous clade in the array</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line433">433</a>             clades.pop</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line434">434</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line435">435</a>         end          </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line436">436</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line437">437</a>         #parsing phylogeny elements</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line438">438</a>         if not parsing_clade</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line439">439</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line440">440</a>           if @reader.node_type == XML::Reader::TYPE_ELEMENT</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line441">441</a>             case @reader.name</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line442">442</a>             when 'property'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line443">443</a>               tree.properties &lt;&lt; parse_property</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line444">444</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line445">445</a>             when 'clade_relation'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line446">446</a>               clade_relation = CladeRelation.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line447">447</a>               parse_attributes(clade_relation, [&quot;id_ref_0&quot;, &quot;id_ref_1&quot;, &quot;distance&quot;, &quot;type&quot;])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line448">448</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line449">449</a>               #@ add unit test for this</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line450">450</a>               if not @reader.empty_element?</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line451">451</a>                 @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line452">452</a>                 if is_element?('confidence')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line453">453</a>                   clade_relation.confidence = parse_confidence</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line454">454</a>                 end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line455">455</a>               end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line456">456</a>               tree.clade_relations &lt;&lt; clade_relation</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line457">457</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line458">458</a>             when 'sequence_relation'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line459">459</a>               sequence_relation = SequenceRelation.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line460">460</a>               parse_attributes(sequence_relation, [&quot;id_ref_0&quot;, &quot;id_ref_1&quot;, &quot;distance&quot;, &quot;type&quot;])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line461">461</a>               if not @reader.empty_element?</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line462">462</a>                 @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line463">463</a>                 if is_element?('confidence')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line464">464</a>                   sequence_relation.confidence = parse_confidence</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line465">465</a>                 end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line466">466</a>               end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line467">467</a>               tree.sequence_relations &lt;&lt; sequence_relation</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line468">468</a>             when 'phylogeny'</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line469">469</a>               #do nothing</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line470">470</a>             else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line471">471</a>               tree.other &lt;&lt; parse_other</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line472">472</a>               #puts &quot;Not recognized element. #{@reader.name}&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line473">473</a>             end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line474">474</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line475">475</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line476">476</a>         # go to next element        </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line477">477</a>         @reader.read    </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line478">478</a>       end #end while not &lt;/phylogeny&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line479">479</a>       #move on to the next tag after /phylogeny which is text, since phylogeny</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line480">480</a>       #end tag is empty element, which value is nil, therefore need to move to</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line481">481</a>       #the next meaningful element (therefore @reader.read twice)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line482">482</a>       @reader.read </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line483">483</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line484">484</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line485">485</a>       return tree</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line486">486</a>     end  </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line487">487</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line488">488</a>     # return tree of specified name.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line489">489</a>     # @todo Implement this method. </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line490">490</a>     # def get_tree_by_name(name)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line491">491</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line492">492</a> #      while not is_end_element?('phyloxml')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line493">493</a> #        if is_element?('phylogeny')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line494">494</a> #          @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line495">495</a> #          @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line496">496</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line497">497</a> #          if is_element?('name')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line498">498</a> #            @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line499">499</a> #            if @reader.value == name</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line500">500</a> #              puts &quot;equasl&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line501">501</a> #              tree = next_tree</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line502">502</a> #              puts tree</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line503">503</a> #            end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line504">504</a> #          end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line505">505</a> #        end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line506">506</a> #        @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line507">507</a> #      end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line508">508</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line509">509</a>   #  end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line510">510</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line511">511</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line512">512</a>     private</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line513">513</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line514">514</a>     ####</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line515">515</a>     # Utility methods</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line516">516</a>     ###</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line517">517</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line518">518</a>     def is_element?(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line519">519</a>       @reader.node_type == XML::Reader::TYPE_ELEMENT and @reader.name == str ? true : false</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line520">520</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line521">521</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line522">522</a>     def is_end_element?(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line523">523</a>       @reader.node_type==XML::Reader::TYPE_END_ELEMENT and @reader.name == str ? true : false</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line524">524</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line525">525</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line526">526</a>     def has_reached_end_element?(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line527">527</a>       if not(is_end_element?(str))        </pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line528">528</a>         raise &quot;Warning: Should have reached &lt;/#{str}&gt; element here&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line529">529</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line530">530</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line531">531</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line532">532</a>     # Parses a simple XML element. for example &lt;speciations&gt;1&lt;/speciations&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line533">533</a>     # It reads in the value and assigns it to object.speciation = 1</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line534">534</a>     # Also checks if have reached end tag (&lt;/speciations&gt; and gives warning</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line535">535</a>     # if not</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line536">536</a>     def parse_simple_element(object, name)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line537">537</a>       if is_element?(name)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line538">538</a>         @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line539">539</a>         object.send(&quot;#{name}=&quot;, @reader.value)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line540">540</a>         @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line541">541</a>         has_reached_end_element?(name)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line542">542</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line543">543</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line544">544</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line545">545</a>     def parse_simple_elements(object, elements)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line546">546</a>       elements.each do |elmt|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line547">547</a>           parse_simple_element(object, elmt)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line548">548</a>       end      </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line549">549</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line550">550</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line551">551</a>     #Parses list of attributes</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line552">552</a>     #use for the code like: clade_relation.type = @reader[&quot;type&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line553">553</a>     def parse_attributes(object, arr_of_attrs)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line554">554</a>       arr_of_attrs.each do |attr|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line555">555</a>         object.send(&quot;#{attr}=&quot;, @reader[attr])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line556">556</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line557">557</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line558">558</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line559">559</a>     def parse_clade_elements(current_node, current_edge)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line560">560</a>       #no loop inside, loop is already outside</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line561">561</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line562">562</a>       if @reader.node_type == XML::Reader::TYPE_ELEMENT</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line563">563</a>         case @reader.name</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line564">564</a>         when 'branch_length'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line565">565</a>           # @todo add unit test for this. current_edge is nil, if the root clade</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line566">566</a>           # has branch_length attribute. </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line567">567</a>           @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line568">568</a>           branch_length = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line569">569</a>           current_edge.distance = branch_length.to_f if current_edge != nil</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line570">570</a>           @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line571">571</a>         when 'width'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line572">572</a>           @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line573">573</a>           current_node.width = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line574">574</a>           @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line575">575</a>         when  'name'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line576">576</a>           @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line577">577</a>           current_node.name = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line578">578</a>           @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line579">579</a>         when 'events'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line580">580</a>           current_node.events = parse_events</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line581">581</a>         when 'confidence'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line582">582</a>           current_node.confidences &lt;&lt; parse_confidence</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line583">583</a>         when 'sequence'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line584">584</a>           current_node.sequences &lt;&lt; parse_sequence</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line585">585</a>         when 'property'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line586">586</a>           current_node.properties &lt;&lt; parse_property</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line587">587</a>         when 'taxonomy'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line588">588</a>           current_node.taxonomies &lt;&lt; parse_taxonomy</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line589">589</a>         when 'distribution'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line590">590</a>           current_node.distributions &lt;&lt; parse_distribution</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line591">591</a>         when 'node_id'</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line592">592</a>           id = Id.new</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line593">593</a>           id.type = @reader[&quot;type&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line594">594</a>           @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line595">595</a>           id.value = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line596">596</a>           @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line597">597</a>           #has_reached_end_element?('node_id')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line598">598</a>           #@todo write unit test for this. There is no example of this in the example files</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line599">599</a>           current_node.id = id</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line600">600</a>         when 'color'</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line601">601</a>           color = BranchColor.new</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line602">602</a>           parse_simple_element(color, 'red')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line603">603</a>           parse_simple_element(color, 'green')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line604">604</a>           parse_simple_element(color, 'blue')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line605">605</a>           current_node.color = color</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line606">606</a>           #@todo add unit test for this</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line607">607</a>         when 'date'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line608">608</a>           date = Date.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line609">609</a>           date.unit = @reader[&quot;unit&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line610">610</a>           #move to the next token, which is always empty, since date tag does not</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line611">611</a>           # have text associated with it</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line612">612</a>           @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line613">613</a>           @reader.read #now the token is the first tag under date tag</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line614">614</a>           while not(is_end_element?('date'))</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line615">615</a>             parse_simple_element(date, 'desc')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line616">616</a>             parse_simple_element(date, 'value')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line617">617</a>             parse_simple_element(date, 'minimum')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line618">618</a>             parse_simple_element(date, 'maximum')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line619">619</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line620">620</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line621">621</a>           current_node.date = date</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line622">622</a>         when 'reference'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line623">623</a>           reference = Reference.new()</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line624">624</a>           reference.doi = @reader['doi']</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line625">625</a>           if not @reader.empty_element?</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line626">626</a>             while not is_end_element?('reference')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line627">627</a>               parse_simple_element(reference, 'desc')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line628">628</a>               @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line629">629</a>             end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line630">630</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line631">631</a>           current_node.references &lt;&lt; reference</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line632">632</a>         when 'binary_characters'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line633">633</a>           current_node.binary_characters  = parse_binary_characters</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line634">634</a>         when 'clade'</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line635">635</a>           #do nothing</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line636">636</a>         else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line637">637</a>           current_node.other &lt;&lt; parse_other</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line638">638</a>           #puts &quot;No match found in parse_clade_elements.(#{@reader.name})&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line639">639</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line640">640</a> </pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line641">641</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line642">642</a> </pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line643">643</a>     end #parse_clade_elements</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line644">644</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line645">645</a>     def parse_events()</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line646">646</a>       events = PhyloXML::Events.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line647">647</a>       @reader.read #go to next element</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line648">648</a>       while not(is_end_element?('events')) do</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line649">649</a>         parse_simple_elements(events, ['type', 'duplications',</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line650">650</a>                                             'speciations', 'losses'])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line651">651</a>         if is_element?('confidence')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line652">652</a>           events.confidence = parse_confidence</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line653">653</a>           #@todo could add unit test for this (example file does not have this case)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line654">654</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line655">655</a>         @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line656">656</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line657">657</a>       return events</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line658">658</a>     end #parse_events</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line659">659</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line660">660</a>     def parse_taxonomy</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line661">661</a>       taxonomy = PhyloXML::Taxonomy.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line662">662</a>       parse_attributes(taxonomy, [&quot;id_source&quot;])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line663">663</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line664">664</a>       while not(is_end_element?('taxonomy')) do</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line665">665</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line666">666</a>         if @reader.node_type == XML::Reader::TYPE_ELEMENT</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line667">667</a>           case @reader.name</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line668">668</a>           when 'code'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line669">669</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line670">670</a>             taxonomy.code = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line671">671</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line672">672</a>           when 'scientific_name'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line673">673</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line674">674</a>             taxonomy.scientific_name = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line675">675</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line676">676</a>           when 'rank'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line677">677</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line678">678</a>             taxonomy.rank = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line679">679</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line680">680</a>           when 'authority'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line681">681</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line682">682</a>             taxonomy.authority = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line683">683</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line684">684</a>           when 'id'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line685">685</a>             taxonomy.taxonomy_id = parse_id('id')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line686">686</a>           when 'common_name'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line687">687</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line688">688</a>             taxonomy.common_names &lt;&lt; @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line689">689</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line690">690</a>             #has_reached_end_element?('common_name')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line691">691</a>           when 'synonym'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line692">692</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line693">693</a>             taxonomy.synonyms &lt;&lt; @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line694">694</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line695">695</a>             #has_reached_end_element?('synonym')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line696">696</a>           when 'uri'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line697">697</a>             taxonomy.uri = parse_uri</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line698">698</a>           else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line699">699</a>             taxonomy.other &lt;&lt; parse_other</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line700">700</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line701">701</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line702">702</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line703">703</a>         @reader.read  #move to next tag in the loop</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line704">704</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line705">705</a>       return taxonomy</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line706">706</a>     end #parse_taxonomy</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line707">707</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line708">708</a>     private</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line709">709</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line710">710</a>     def parse_sequence</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line711">711</a>       sequence = Sequence.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line712">712</a>       parse_attributes(sequence, [&quot;type&quot;, &quot;id_source&quot;, &quot;id_ref&quot;])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line713">713</a>       </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line714">714</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line715">715</a>       while not(is_end_element?('sequence'))</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line716">716</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line717">717</a>         if @reader.node_type == XML::Reader::TYPE_ELEMENT</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line718">718</a>           case @reader.name</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line719">719</a>           when 'symbol'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line720">720</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line721">721</a>             sequence.symbol = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line722">722</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line723">723</a>           when 'name'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line724">724</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line725">725</a>             sequence.name = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line726">726</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line727">727</a>           when 'location'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line728">728</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line729">729</a>             sequence.location = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line730">730</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line731">731</a>           when 'mol_seq'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line732">732</a>             sequence.is_aligned = @reader[&quot;is_aligned&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line733">733</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line734">734</a>             sequence.mol_seq = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line735">735</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line736">736</a>             has_reached_end_element?('mol_seq')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line737">737</a>           when 'accession'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line738">738</a>             sequence.accession = Accession.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line739">739</a>             sequence.accession.source = @reader[&quot;source&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line740">740</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line741">741</a>             sequence.accession.value = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line742">742</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line743">743</a>             has_reached_end_element?('accession')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line744">744</a>           when 'uri'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line745">745</a>             sequence.uri = parse_uri</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line746">746</a>           when 'annotation'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line747">747</a>             sequence.annotations &lt;&lt; parse_annotation</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line748">748</a>           when 'domain_architecture'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line749">749</a>             sequence.domain_architecture = DomainArchitecture.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line750">750</a>             sequence.domain_architecture.length = @reader[&quot;length&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line751">751</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line752">752</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line753">753</a>             while not(is_end_element?('domain_architecture'))</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line754">754</a>               sequence.domain_architecture.domains &lt;&lt; parse_domain</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line755">755</a>               @reader.read #go to next domain element</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line756">756</a>             end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line757">757</a>           else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line758">758</a>             sequence.other &lt;&lt; parse_other</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line759">759</a>             #@todo add unit test            </pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line760">760</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line761">761</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line762">762</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line763">763</a>         @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line764">764</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line765">765</a>       return sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line766">766</a>     end #parse_sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line767">767</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line768">768</a>     def parse_uri</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line769">769</a>       uri = Uri.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line770">770</a>       parse_attributes(uri, [&quot;desc&quot;, &quot;type&quot;])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line771">771</a>       parse_simple_element(uri, 'uri')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line772">772</a>       return uri</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line773">773</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line774">774</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line775">775</a>     def parse_annotation</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line776">776</a>       annotation = Annotation.new</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line777">777</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line778">778</a>       parse_attributes(annotation, ['ref', 'source', 'evidence', 'type'])</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line779">779</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line780">780</a>       if not @reader.empty_element?</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line781">781</a>         while not(is_end_element?('annotation'))</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line782">782</a>           parse_simple_element(annotation, 'desc') if is_element?('desc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line783">783</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line784">784</a>           annotation.confidence  = parse_confidence if is_element?('confidence')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line785">785</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line786">786</a>           annotation.properties &lt;&lt; parse_property if is_element?('property')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line787">787</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line788">788</a>           if is_element?('uri')</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line789">789</a>             annotation.uri = parse_uri        </pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line790">790</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line791">791</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line792">792</a>           @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line793">793</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line794">794</a>         </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line795">795</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line796">796</a>       return annotation</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line797">797</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line798">798</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line799">799</a>     def parse_property</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line800">800</a>       property = Property.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line801">801</a>       parse_attributes(property, [&quot;ref&quot;, &quot;unit&quot;, &quot;datatype&quot;, &quot;applies_to&quot;, &quot;id_ref&quot;])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line802">802</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line803">803</a>       property.value = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line804">804</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line805">805</a>       has_reached_end_element?('property')     </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line806">806</a>       return property</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line807">807</a>     end #parse_property</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line808">808</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line809">809</a>     def parse_confidence</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line810">810</a>       type = @reader[&quot;type&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line811">811</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line812">812</a>       value = @reader.value.to_f</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line813">813</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line814">814</a>       has_reached_end_element?('confidence')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line815">815</a>       return Confidence.new(type, value)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line816">816</a>     end #parse_confidence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line817">817</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line818">818</a>     def parse_distribution</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line819">819</a>       distribution = Distribution.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line820">820</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line821">821</a>       while not(is_end_element?('distribution')) do</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line822">822</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line823">823</a>         parse_simple_element(distribution, 'desc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line824">824</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line825">825</a>         distribution.points &lt;&lt; parse_point if is_element?('point')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line826">826</a>         distribution.polygons &lt;&lt; parse_polygon if is_element?('polygon')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line827">827</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line828">828</a>         @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line829">829</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line830">830</a>       return distribution</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line831">831</a>     end #parse_distribution</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line832">832</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line833">833</a>     def parse_point</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line834">834</a>       point = Point.new</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line835">835</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line836">836</a>       point.geodetic_datum = @reader[&quot;geodetic_datum&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line837">837</a>       point.alt_unit = @reader[&quot;alt_unit&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line838">838</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line839">839</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line840">840</a>       while not(is_end_element?('point')) do</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line841">841</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line842">842</a>         parse_simple_elements(point, ['lat', 'long'] )</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line843">843</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line844">844</a>         if is_element?('alt')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line845">845</a>           @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line846">846</a>           point.alt = @reader.value.to_f</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line847">847</a>           @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line848">848</a>           has_reached_end_element?('alt')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line849">849</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line850">850</a>         #advance reader</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line851">851</a>         @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line852">852</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line853">853</a>       return point</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line854">854</a>     end #parse_point</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line855">855</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line856">856</a>     def parse_polygon</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line857">857</a>       polygon = Polygon.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line858">858</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line859">859</a>       while not(is_end_element?('polygon')) do</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line860">860</a>         polygon.points &lt;&lt; parse_point if is_element?('point')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line861">861</a>         @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line862">862</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line863">863</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line864">864</a>       #@todo should check for it at all? Probably not if xml is valid.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line865">865</a>       if polygon.points.length &lt;3</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line866">866</a>         puts &quot;Warning: &lt;polygon&gt; should have at least 3 points&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line867">867</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line868">868</a>       return polygon</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line869">869</a>     end #parse_polygon</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line870">870</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line871">871</a>     def parse_id(tag_name)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line872">872</a>       id = Id.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line873">873</a>       id.provider = @reader[&quot;provider&quot;]</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line874">874</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line875">875</a>       id.value = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line876">876</a>       @reader.read #@todo shouldn't there be another read?</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line877">877</a>       has_reached_end_element?(tag_name)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line878">878</a>       return id</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line879">879</a>     end #parse_id</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line880">880</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line881">881</a>     def parse_domain</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line882">882</a>       domain = ProteinDomain.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line883">883</a>       parse_attributes(domain, [&quot;from&quot;, &quot;to&quot;, &quot;confidence&quot;, &quot;id&quot;])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line884">884</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line885">885</a>       domain.value = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line886">886</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line887">887</a>       has_reached_end_element?('domain')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line888">888</a>       @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line889">889</a>       return domain</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line890">890</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line891">891</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line892">892</a>     def parse_binary_characters</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line893">893</a>       b = PhyloXML::BinaryCharacters.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line894">894</a>       b.bc_type = @reader['type']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line895">895</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line896">896</a>       parse_attributes(b, ['gained_count', 'absent_count', 'lost_count', 'present_count'])</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line897">897</a>       if not @reader.empty_element?</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line898">898</a>         @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line899">899</a>         while not is_end_element?('binary_characters')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line900">900</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line901">901</a>           parse_bc(b, 'lost')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line902">902</a>           parse_bc(b, 'gained')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line903">903</a>           parse_bc(b, 'absent')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line904">904</a>           parse_bc(b, 'present')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line905">905</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line906">906</a>           @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line907">907</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line908">908</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line909">909</a>       return b</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line910">910</a>     end #parse_binary_characters</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line911">911</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line912">912</a>     def parse_bc(object, element)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line913">913</a>       if is_element?(element)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line914">914</a>         @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line915">915</a>         while not is_end_element?(element)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line916">916</a>           if is_element?('bc')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line917">917</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line918">918</a>             object.send(element) &lt;&lt; @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line919">919</a>             @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line920">920</a>             has_reached_end_element?('bc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line921">921</a>           end</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line922">922</a>         @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line923">923</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line924">924</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line925">925</a>     end #parse_bc</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line926">926</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line927">927</a>     def parse_other</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line928">928</a>       other_obj = PhyloXML::Other.new</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line929">929</a>       other_obj.element_name = @reader.name</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line930">930</a>       #parse attributes</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line931">931</a>       code = @reader.move_to_first_attribute</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line932">932</a>       while code ==1</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line933">933</a>         other_obj.attributes[@reader.name] = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line934">934</a>         code = @reader.move_to_next_attribute        </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line935">935</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line936">936</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line937">937</a>       while not is_end_element?(other_obj.element_name) do</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line938">938</a>         @reader.read</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line939">939</a>         if @reader.node_type == XML::Reader::TYPE_ELEMENT</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line940">940</a>            other_obj.children &lt;&lt; parse_other #recursice call to parse children</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line941">941</a>         elsif @reader.node_type == XML::Reader::TYPE_TEXT</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line942">942</a>           other_obj.value = @reader.value</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line943">943</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line944">944</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line945">945</a>       #just a check</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line946">946</a>       has_reached_end_element?(other_obj.element_name)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line947">947</a>       return other_obj</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line948">948</a>     end #parse_other</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line949">949</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line950">950</a>   end #class phyloxmlParser</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line951">951</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line952">952</a> end #module PhyloXML</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line953">953</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line954">954</a> end #module Bio</pre></td>
          </tr>
        
      </tbody>
    </table>

    <p>Generated on Fri Jun 25 17:37:23 +0900 2010 with <a href="http://github.com/relevance/rcov">rcov 0.9.8</a></p>

  </body>
</html>
