<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang='en' xml:lang='en' xmlns='http://www.w3.org/1999/xhtml'>
  <head>
    <title>lib/bio/db/lasergene.rb</title>
    <link href="screen.css" media="all" rel="stylesheet" type="text/css" />
    <link href="print.css" media="print" rel="stylesheet" type="text/css" />
    
    <meta http-equiv="Content-type" content="text/html;charset=UTF-8" />
    <script type="text/javascript" src="rcov.js"></script>
  </head>
  <body>
    <h1>Bioruby C0 Coverage Information - RCov</h1>
    <h2>lib/bio/db/lasergene.rb</h2>

    

    <div class="report_table_wrapper">
      <table class='report' id='report_table'>
        <thead>
          <tr>
            <th class="left_align">Name</th>
            <th class="right_align">Total Lines</th>
            <th class="right_align">Lines of Code</th>
            <th class="left_align">Total Coverage</th>
            <th class="left_align">Code Coverage</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td class="left_align"><a href="lib-bio-db-lasergene_rb.html">lib/bio/db/lasergene.rb</a></td>
            <td class='right_align'><tt>209</tt></td>
            <td class='right_align'><tt>72</tt></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>100.00%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:100px"></div>
            <div class="uncovered" style="width:0px"></div>
          </div></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>100.00%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:100px"></div>
            <div class="uncovered" style="width:0px"></div>
          </div></td>
          </tr>
        </tbody>
      </table>
    </div>
    
    <h3>Key</h3>
    
    <div class="key"><pre><span class='marked'>Code reported as executed by Ruby looks like this...</span><span class='marked1'>and this: this line is also marked as covered.</span><span class='inferred'>Lines considered as run by rcov, but not reported by Ruby, look like this,</span><span class='inferred1'>and this: these lines were inferred by rcov (using simple heuristics).</span><span class='uncovered'>Finally, here's a line marked as not executed.</span></pre></div>

    <h3>Coverage Details</h3>

    <table class="details">
      <tbody>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1">1</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line2">2</a> # bio/db/lasergene.rb - Interface for DNAStar Lasergene sequence file format</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line3">3</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line4">4</a> # Author::    Trevor Wennblom  &lt;mailto:trevor@corevx.com&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line5">5</a> # Copyright:: Copyright (c) 2007 Center for Biomedical Research Informatics, University of Minnesota (http://cbri.umn.edu)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line6">6</a> # License::   The Ruby License</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line7">7</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line8">8</a> #  $Id:$</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line9">9</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line10">10</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line11">11</a> module Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line12">12</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line13">13</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line14">14</a> # bio/db/lasergene.rb - Interface for DNAStar Lasergene sequence file format</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line15">15</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line16">16</a> # Author::    Trevor Wennblom  &lt;mailto:trevor@corevx.com&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line17">17</a> # Copyright:: Copyright (c) 2007 Center for Biomedical Research Informatics, University of Minnesota (http://cbri.umn.edu)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line18">18</a> # License::   The Ruby License</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line19">19</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line20">20</a> # = Description</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line21">21</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line22">22</a> # Bio::Lasergene reads DNAStar Lasergene formatted sequence files, or +.seq+ </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line23">23</a> # files.  It only expects to find one sequence per file.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line24">24</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line25">25</a> # = Usage</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line26">26</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line27">27</a> #   require 'bio'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line28">28</a> #   filename = 'MyFile.seq'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line29">29</a> #   lseq = Bio::Lasergene.new( IO.readlines(filename) )</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line30">30</a> #   lseq.entry_id  # =&gt; &quot;Contig 1&quot;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line31">31</a> #   lseq.seq  # =&gt; ATGACGTATCCAAAGAGGCGTTACC</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line32">32</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line33">33</a> # = Comments</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line34">34</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line35">35</a> # I'm only aware of the following three kinds of Lasergene file formats.  Feel</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line36">36</a> # free to send me other examples that may not currently be accounted for.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line37">37</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line38">38</a> # File format 1:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line39">39</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line40">40</a> #   ## begin ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line41">41</a> #   &quot;Contig 1&quot; (1,934)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line42">42</a> #     Contig Length:                  934 bases</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line43">43</a> #     Average Length/Sequence:        467 bases</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line44">44</a> #     Total Sequence Length:         1869 bases</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line45">45</a> #     Top Strand:                       2 sequences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line46">46</a> #     Bottom Strand:                    2 sequences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line47">47</a> #     Total:                            4 sequences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line48">48</a> #   ^^</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line49">49</a> #   ATGACGTATCCAAAGAGGCGTTACCGGAGAAGAAGACACCGCCCCCGCAGTCCTCTTGGCCAGATCCTCCGCCGCCGCCCCTGGCTCGTCCACCCCCGCCACAGTTACCGCTGGAGAAGGAAAAATGGCATCTTCAWCACCCGCCTATCCCGCAYCTTCGGAWRTACTATCAAGCGAACCACAGTCAGAACGCCCTCCTGGGCGGTGGACATGATGAGATTCAATATTAATGACTTTCTTCCCCCAGGAGGGGGCTCAAACCCCCGCTCTGTGCCCTTTGAATACTACAGAATAAGAAAGGTTAAGGTTGAATTCTGGCCCTGCTCCCCGATCACCCAGGGTGACAGGGGAATGGGCTCCAGTGCTGWTATTCTAGMTGATRRCTTKGTAACAAAGRCCACAGCCCTCACCTATGACCCCTATGTAAACTTCTCCTCCCGCCATACCATAACCCAGCCCTTCTCCTACCRCTCCCGYTACTTTACCCCCAAACCTGTCCTWGATKCCACTATKGATKACTKCCAACCAAACAACAAAAGAAACCAGCTGTGGSTGAGACTACAWACTGCTGGAAATGTAGACCWCGTAGGCCTSGGCACTGCGTKCGAAAACAGTATATACGACCAGGAATACAATATCCGTGTMACCATGTATGTACAATTCAGAGAATTTAATCTTAAAGACCCCCCRCTTMACCCKTAATGAATAATAAMAACCATTACGAAGTGATAAAAWAGWCTCAGTAATTTATTYCATATGGAAATTCWSGGCATGGGGGGGAAAGGGTGACGAACKKGCCCCCTTCCTCCSTSGMYTKTTCYGTAGCATTCYTCCAMAAYACCWAGGCAGYAMTCCTCCSATCAAGAGcYTSYACAGCTGGGACAGCAGTTGAGGAGGACCATTCAAAGGGGGTCGGATTGCTGGTAATCAGA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line50">50</a> #   ## end ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line51">51</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line52">52</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line53">53</a> # File format 2:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line54">54</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line55">55</a> #   ## begin ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line56">56</a> #   ^^:                                  350,935</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line57">57</a> #   Contig 1 (1,935)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line58">58</a> #     Contig Length:                  935 bases</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line59">59</a> #     Average Length/Sequence:        580 bases</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line60">60</a> #     Total Sequence Length:         2323 bases</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line61">61</a> #     Top Strand:                       2 sequences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line62">62</a> #     Bottom Strand:                    2 sequences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line63">63</a> #     Total:                            4 sequences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line64">64</a> #   ^^</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line65">65</a> #   ATGTCGGGGAAATGCTTGACCGCGGGCTACTGCTCATCATTGCTTTCTTTGTGGTATATCGTGCCGTTCTGTTTTGCTGTGCTCGTCAACGCCAGCGGCGACAGCAGCTCTCATTTTCAGTCGATTTATAACTTGACGTTATGTGAGCTGAATGGCACGAACTGGCTGGCAGACAACTTTAACTGGGCTGTGGAGACTTTTGTCATCTTCCCCGTGTTGACTCACATTGTTTCCTATGGTGCACTCACTACCAGTCATTTTCTTGACACAGTTGGTCTAGTTACTGTGTCTACCGCCGGGTTTTATCACGGGCGGTACGTCTTGAGTAGCATCTACGCGGTCTGTGCTCTGGCTGCGTTGATTTGCTTCGCCATCAGGTTTGCGAAGAACTGCATGTCCTGGCGCTACTCTTGCACTAGATACACCAACTTCCTCCTGGACACCAAGGGCAGACTCTATCGTTGGCGGTCGCCTGTCATCATAGAGAAAGGGGGTAAGGTTGAGGTCGAAGGTCATCTGATCGATCTCAAAAGAGTTGTGCTTGATGGCTCTGTGGCGACACCTTTAACCAGAGTTTCAGCGGAACAATGGGGTCGTCCCTAGACGACTTTTGCCATGATAGTACAGCCCCACAGAAGGTGCTCTTGGCGTTTTCCATCACCTACACGCCAGTGATGATATATGCCCTAAAGGTAAGCCGCGGCCGACTTTTGGGGCTTCTGCACCTTTTGATTTTTTTGAACTGTGCCTTTACTTTCGGGTACATGACATTCGTGCACTTTCGGAGCACGAACAAGGTCGCGCTCACTATGGGAGCAGTAGTCGCACTCCTTTGGGGGGTGTACTCAGCCATAGAAACCTGGAAATTCATCACCTCCAGATGCCGTTGTGCTTGCTAGGCCGCAAGTACATTCTGGCCCCTGCCCACCACGTTG</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line66">66</a> #   ## end ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line67">67</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line68">68</a> # File format 3 (non-standard Lasergene header):</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line69">69</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line70">70</a> #   ## begin ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line71">71</a> #   LOCUS       PRU87392               15411 bp    RNA     linear   VRL 17-NOV-2000</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line72">72</a> #   DEFINITION  Porcine reproductive and respiratory syndrome virus strain VR-2332,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line73">73</a> #               complete genome.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line74">74</a> #   ACCESSION   U87392 AF030244 U00153</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line75">75</a> #   VERSION     U87392.3  GI:11192298</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line76">76</a> #   [...cut...]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line77">77</a> #        3'UTR           15261..15411</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line78">78</a> #        polyA_site      15409</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line79">79</a> #   ORIGIN      </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line80">80</a> #   ^^</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line81">81</a> #   atgacgtataggtgttggctctatgccttggcatttgtattgtcaggagctgtgaccattggcacagcccaaaacttgctgcacagaaacacccttctgtgatagcctccttcaggggagcttagggtttgtccctagcaccttgcttccggagttgcactgctttacggtctctccacccctttaaccatgtctgggatacttgatcggtgcacgtgtacccccaatgccagggtgtttatggcggagggccaagtctactgcacacgatgcctcagtgcacggtctctccttcccctgaacctccaagtttctgagctcggggtgctaggcctattctacaggcccgaagagccactccggtggacgttgccacgtgcattccccactgttgagtgctcccccgccggggcctgctggctttctgcaatctttccaatcgcacgaatgaccagtggaaacctgaacttccaacaaagaatggtacgggtcgcagctgagctttacagagccggccagctcacccctgcagtcttgaaggctctacaagtttatgaacggggttgccgctggtaccccattgttggacctgtccctggagtggccgttttcgccaattccctacatgtgagtgataaacctttcccgggagcaactcacgtgttgaccaacctgccgctcccgcagagacccaagcctgaagacttttgcccctttgagtgtgctatggctactgtctatgacattggtcatgacgccgtcatgtatgtggccgaaaggaaagtctcctgggcccctcgtggcggggatgaagtgaaatttgaagctgtccccggggagttgaagttgattgcgaaccggctccgcacctccttcccgccccaccacacagtggacatgtctaagttcgccttcacagcccctgggtgtggtgtttctatgcgggtcgaacgccaacacggctgccttcccgctgacactgtccctgaaggcaactgctggtggagcttgtttgacttgcttccactggaagttcagaacaaagaaattcgccatgctaaccaatttggctaccagaccaagcatggtgtctctggcaagtacctacagcggaggctgca[...cut...]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line82">82</a> #   ## end ##</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line83">83</a> #</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line84">84</a> class Lasergene</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line85">85</a>   # Entire header before the sequence</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line86">86</a>   attr_reader :comments</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line87">87</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line88">88</a>   # Sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line89">89</a>   # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line90">90</a>   # Bio::Sequence::NA or Bio::Sequence::AA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line91">91</a>   attr_reader :sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line92">92</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line93">93</a>   # Name of sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line94">94</a>   # * Parsed from standard Lasergene header</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line95">95</a>   attr_reader :name</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line96">96</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line97">97</a>   # Contig length, length of present sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line98">98</a>   # * Parsed from standard Lasergene header</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line99">99</a>   attr_reader :contig_length</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line100">100</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line101">101</a>   # Average length per sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line102">102</a>   # * Parsed from standard Lasergene header</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line103">103</a>   attr_reader :average_length</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line104">104</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line105">105</a>   # Length of parent sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line106">106</a>   # * Parsed from standard Lasergene header</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line107">107</a>   attr_reader :total_length</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line108">108</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line109">109</a>   # Number of top strand sequences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line110">110</a>   # * Parsed from standard Lasergene header</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line111">111</a>   attr_reader :top_strand_sequences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line112">112</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line113">113</a>   # Number of bottom strand sequences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line114">114</a>   # * Parsed from standard Lasergene header</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line115">115</a>   attr_reader :bottom_strand_sequences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line116">116</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line117">117</a>   # Number of sequences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line118">118</a>   # * Parsed from standard Lasergene header</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line119">119</a>   attr_reader :total_sequences</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line120">120</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line121">121</a>   DELIMITER_1 = '^\^\^:' # Match '^^:' at the beginning of a line</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line122">122</a>   DELIMITER_2 = '^\^\^'  # Match '^^' at the beginning of a line</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line123">123</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line124">124</a>   def initialize(lines)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line125">125</a>     process(lines)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line126">126</a>   end  </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line127">127</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line128">128</a>   # Is the comment header recognized as standard Lasergene format?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line129">129</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line130">130</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line131">131</a>   # *Arguments*</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line132">132</a>   # * _none_</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line133">133</a>   # *Returns*:: +true+ _or_ +false+</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line134">134</a>   def standard_comment?</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line135">135</a>     @standard_comment</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line136">136</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line137">137</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line138">138</a>   # Sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line139">139</a>   # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line140">140</a>   # Bio::Sequence::NA or Bio::Sequence::AA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line141">141</a>   def seq</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line142">142</a>     @sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line143">143</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line144">144</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line145">145</a>   # Name of sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line146">146</a>   # * Parsed from standard Lasergene header</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line147">147</a>   def entry_id</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line148">148</a>     @name</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line149">149</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line150">150</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line151">151</a>   #########</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line152">152</a>   protected</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line153">153</a>   #########</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line154">154</a>   </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line155">155</a>   def process(lines)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line156">156</a>     delimiter_1_indices = []</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line157">157</a>     delimiter_2_indices = []</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line158">158</a>     </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line159">159</a>     # If the data from the file is passed as one big String instead of</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line160">160</a>     # broken into an Array, convert lines to an Array</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line161">161</a>     if lines.kind_of? String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line162">162</a>       lines = lines.tr(&quot;\r&quot;, '').split(&quot;\n&quot;)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line163">163</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line164">164</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line165">165</a>     lines.each_with_index do |line, index|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line166">166</a>       if line.match DELIMITER_1</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line167">167</a>         delimiter_1_indices &lt;&lt; index</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line168">168</a>       elsif line.match DELIMITER_2</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line169">169</a>         delimiter_2_indices &lt;&lt; index</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line170">170</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line171">171</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line172">172</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line173">173</a>     raise InputError, &quot;More than one delimiter of type '#{DELIMITER_1}'&quot; if delimiter_1_indices.size &gt; 1</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line174">174</a>     raise InputError, &quot;More than one delimiter of type '#{DELIMITER_2}'&quot; if delimiter_2_indices.size &gt; 1</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line175">175</a>     raise InputError, &quot;No comment to data separator of type '#{DELIMITER_2}'&quot; if delimiter_2_indices.size &lt; 1</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line176">176</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line177">177</a>     if !delimiter_1_indices.empty?</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line178">178</a>       # toss out DELIMETER_1 and anything preceding it</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line179">179</a>       @comments = lines[ (delimiter_1_indices[0] + 1) .. (delimiter_2_indices[0] - 1) ]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line180">180</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line181">181</a>       @comments = lines[ 0 .. (delimiter_2_indices[0] - 1) ]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line182">182</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line183">183</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line184">184</a>     @standard_comment = false</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line185">185</a>     if @comments[0] =~ %r{(.+)\s+\(\d+,\d+\)} # if we have a standard Lasergene comment</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line186">186</a>       @standard_comment = true</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line187">187</a>       @name = $1</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line188">188</a>       comments.each do |comment|</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line189">189</a>         if comment.match('Contig Length:\s+(\d+)')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line190">190</a>           @contig_length = $1.to_i</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line191">191</a>         elsif comment.match('Average Length/Sequence:\s+(\d+)')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line192">192</a>           @average_length = $1.to_i</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line193">193</a>         elsif comment.match('Total Sequence Length:\s+(\d+)')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line194">194</a>           @total_length = $1.to_i</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line195">195</a>         elsif comment.match('Top Strand:\s+(\d+)')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line196">196</a>           @top_strand_sequences = $1.to_i</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line197">197</a>         elsif comment.match('Bottom Strand:\s+(\d+)')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line198">198</a>           @bottom_strand_sequences = $1.to_i</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line199">199</a>         elsif comment.match('Total:\s+(\d+)')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line200">200</a>           @total_sequences = $1.to_i</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line201">201</a>         end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line202">202</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line203">203</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line204">204</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line205">205</a>     @comments = @comments.join('')</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line206">206</a>     @sequence = Bio::Sequence.auto( lines[ (delimiter_2_indices[0] + 1) .. -1 ].join('') )</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line207">207</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line208">208</a> end # Lasergene</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line209">209</a> end # Bio</pre></td>
          </tr>
        
      </tbody>
    </table>

    <p>Generated on Fri Jun 25 17:37:23 +0900 2010 with <a href="http://github.com/relevance/rcov">rcov 0.9.8</a></p>

  </body>
</html>
