<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang='en' xml:lang='en' xmlns='http://www.w3.org/1999/xhtml'>
  <head>
    <title>lib/bio/sequence.rb</title>
    <link href="screen.css" media="all" rel="stylesheet" type="text/css" />
    <link href="print.css" media="print" rel="stylesheet" type="text/css" />
    
    <meta http-equiv="Content-type" content="text/html;charset=UTF-8" />
    <script type="text/javascript" src="rcov.js"></script>
  </head>
  <body>
    <h1>Bioruby C0 Coverage Information - RCov</h1>
    <h2>lib/bio/sequence.rb</h2>

    

    <div class="report_table_wrapper">
      <table class='report' id='report_table'>
        <thead>
          <tr>
            <th class="left_align">Name</th>
            <th class="right_align">Total Lines</th>
            <th class="right_align">Lines of Code</th>
            <th class="left_align">Total Coverage</th>
            <th class="left_align">Code Coverage</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td class="left_align"><a href="lib-bio-sequence_rb.html">lib/bio/sequence.rb</a></td>
            <td class='right_align'><tt>473</tt></td>
            <td class='right_align'><tt>124</tt></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>92.39%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:92px"></div>
            <div class="uncovered" style="width:8px"></div>
          </div></td>
            <td class="left_align"><div class="percent_graph_legend"><tt class=''>74.19%</tt></div>
          <div class="percent_graph">
            <div class="covered" style="width:74px"></div>
            <div class="uncovered" style="width:26px"></div>
          </div></td>
          </tr>
        </tbody>
      </table>
    </div>
    
    <h3>Key</h3>
    
    <div class="key"><pre><span class='marked'>Code reported as executed by Ruby looks like this...</span><span class='marked1'>and this: this line is also marked as covered.</span><span class='inferred'>Lines considered as run by rcov, but not reported by Ruby, look like this,</span><span class='inferred1'>and this: these lines were inferred by rcov (using simple heuristics).</span><span class='uncovered'>Finally, here's a line marked as not executed.</span></pre></div>

    <h3>Coverage Details</h3>

    <table class="details">
      <tbody>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line1">1</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line2">2</a> # = bio/sequence.rb - biological sequence class</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line3">3</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line4">4</a> # Copyright::   Copyright (C) 2000-2006</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line5">5</a> #               Toshiaki Katayama &lt;k@bioruby.org&gt;,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line6">6</a> #               Yoshinori K. Okuji &lt;okuji@enbug.org&gt;,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line7">7</a> #               Naohisa Goto &lt;ng@bioruby.org&gt;,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line8">8</a> #               Ryan Raaum &lt;ryan@raaum.org&gt;,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line9">9</a> #               Jan Aerts &lt;jan.aerts@bbsrc.ac.uk&gt;</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line10">10</a> # License::     The Ruby License</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line11">11</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line12">12</a> # $Id:$</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line13">13</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line14">14</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line15">15</a> require 'bio/sequence/compat'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line16">16</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line17">17</a> module Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line18">18</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line19">19</a> # = DESCRIPTION</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line20">20</a> # Bio::Sequence objects represent annotated sequences in bioruby.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line21">21</a> # A Bio::Sequence object is a wrapper around the actual sequence, </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line22">22</a> # represented as either a Bio::Sequence::NA or a Bio::Sequence::AA object.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line23">23</a> # For most users, this encapsulation will be completely transparent.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line24">24</a> # Bio::Sequence responds to all methods defined for Bio::Sequence::NA/AA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line25">25</a> # objects using the same arguments and returning the same values (even though </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line26">26</a> # these methods are not documented specifically for Bio::Sequence).</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line27">27</a> #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line28">28</a> # = USAGE</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line29">29</a> #   # Create a nucleic or amino acid sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line30">30</a> #   dna = Bio::Sequence.auto('atgcatgcATGCATGCAAAA')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line31">31</a> #   rna = Bio::Sequence.auto('augcaugcaugcaugcaaaa')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line32">32</a> #   aa = Bio::Sequence.auto('ACDEFGHIKLMNPQRSTVWYU')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line33">33</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line34">34</a> #   # Print it out</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line35">35</a> #   puts dna.to_s</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line36">36</a> #   puts aa.to_s</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line37">37</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line38">38</a> #   # Get a subsequence, bioinformatics style (first nucleotide is '1')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line39">39</a> #   puts dna.subseq(2,6)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line40">40</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line41">41</a> #   # Get a subsequence, informatics style (first nucleotide is '0')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line42">42</a> #   puts dna[2,6]</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line43">43</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line44">44</a> #   # Print in FASTA format</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line45">45</a> #   puts dna.output(:fasta)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line46">46</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line47">47</a> #   # Print all codons</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line48">48</a> #   dna.window_search(3,3) do |codon|</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line49">49</a> #     puts codon</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line50">50</a> #   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line51">51</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line52">52</a> #   # Splice or otherwise mangle your sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line53">53</a> #   puts dna.splicing(&quot;complement(join(1..5,16..20))&quot;)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line54">54</a> #   puts rna.splicing(&quot;complement(join(1..5,16..20))&quot;)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line55">55</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line56">56</a> #   # Convert a sequence containing ambiguity codes into a </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line57">57</a> #   # regular expression you can use for subsequent searching</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line58">58</a> #   puts aa.to_re</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line59">59</a> # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line60">60</a> #   # These should speak for themselves</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line61">61</a> #   puts dna.complement</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line62">62</a> #   puts dna.composition</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line63">63</a> #   puts dna.molecular_weight</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line64">64</a> #   puts dna.translate</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line65">65</a> #   puts dna.gc_percent</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line66">66</a> class Sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line67">67</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line68">68</a>   autoload :Common,  'bio/sequence/common'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line69">69</a>   autoload :NA,      'bio/sequence/na'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line70">70</a>   autoload :AA,      'bio/sequence/aa'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line71">71</a>   autoload :Generic, 'bio/sequence/generic'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line72">72</a>   autoload :Format,  'bio/sequence/format'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line73">73</a>   autoload :Adapter, 'bio/sequence/adapter'</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line74">74</a>   autoload :QualityScore, 'bio/sequence/quality_score'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line75">75</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line76">76</a>   include Format</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line77">77</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line78">78</a>   # Create a new Bio::Sequence object</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line79">79</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line80">80</a>   #   s = Bio::Sequence.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line81">81</a>   #   puts s                                  #=&gt; 'atgc'</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line82">82</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line83">83</a>   # Note that this method does not intialize the contained sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line84">84</a>   # as any kind of bioruby object, only as a simple string</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line85">85</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line86">86</a>   #   puts s.seq.class                        #=&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line87">87</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line88">88</a>   # See Bio::Sequence#na, Bio::Sequence#aa, and Bio::Sequence#auto </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line89">89</a>   # for methods to transform the basic String of a just created </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line90">90</a>   # Bio::Sequence object to a proper bioruby object</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line91">91</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line92">92</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line93">93</a>   # * (required) _str_: String or Bio::Sequence::NA/AA object</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line94">94</a>   # *Returns*:: Bio::Sequence object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line95">95</a>   def initialize(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line96">96</a>     @seq = str</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line97">97</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line98">98</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line99">99</a>   # Pass any unknown method calls to the wrapped sequence object.  see</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line100">100</a>   # http://www.rubycentral.com/book/ref_c_object.html#Object.method_missing</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line101">101</a>   def method_missing(sym, *args, &amp;block) #:nodoc:</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line102">102</a>     begin</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line103">103</a>       seq.__send__(sym, *args, &amp;block)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line104">104</a>     rescue NoMethodError =&gt; evar</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line105">105</a>       lineno = __LINE__ - 2</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line106">106</a>       file = __FILE__</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line107">107</a>       bt_here = [ &quot;#{file}:#{lineno}:in \`__send__\'&quot;,</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line108">108</a>                   &quot;#{file}:#{lineno}:in \`method_missing\'&quot;</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line109">109</a>                 ]</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line110">110</a>       if bt_here == evar.backtrace[0, 2] then</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line111">111</a>         bt = evar.backtrace[2..-1]</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line112">112</a>         evar = evar.class.new(&quot;undefined method \`#{sym.to_s}\' for #{self.inspect}&quot;)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line113">113</a>         evar.set_backtrace(bt)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line114">114</a>       end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line115">115</a>       #p lineno</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line116">116</a>       #p file</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line117">117</a>       #p bt_here</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line118">118</a>       #p evar.backtrace</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line119">119</a>       raise(evar)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line120">120</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line121">121</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line122">122</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line123">123</a>   # The sequence identifier (String).  For example, for a sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line124">124</a>   # of Genbank origin, this is the locus name.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line125">125</a>   # For a sequence of EMBL origin, this is the primary accession number.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line126">126</a>   attr_accessor :entry_id</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line127">127</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line128">128</a>   # A String with a description of the sequence (String)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line129">129</a>   attr_accessor :definition</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line130">130</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line131">131</a>   # Features (An Array of Bio::Feature objects)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line132">132</a>   attr_accessor :features</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line133">133</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line134">134</a>   # References (An Array of Bio::Reference objects)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line135">135</a>   attr_accessor :references</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line136">136</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line137">137</a>   # Comments (String or an Array of String)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line138">138</a>   attr_accessor :comments</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line139">139</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line140">140</a>   # Keywords (An Array of String)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line141">141</a>   attr_accessor :keywords</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line142">142</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line143">143</a>   # Links to other database entries.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line144">144</a>   # (An Array of Bio::Sequence::DBLink objects)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line145">145</a>   attr_accessor :dblinks</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line146">146</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line147">147</a>   # Bio::Sequence::NA/AA</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line148">148</a>   attr_accessor :moltype</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line149">149</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line150">150</a>   # The sequence object, usually Bio::Sequence::NA/AA, </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line151">151</a>   # but could be a simple String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line152">152</a>   attr_accessor :seq</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line153">153</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line154">154</a>   # Quality scores of the bases/residues in the sequence.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line155">155</a>   # (Array containing Integer, or nil)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line156">156</a>   attr_accessor :quality_scores</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line157">157</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line158">158</a>   # The meaning (calculation method) of the quality scores stored in</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line159">159</a>   # the &lt;tt&gt;quality_scores&lt;/tt&gt; attribute.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line160">160</a>   # Maybe one of :phred, :solexa, or nil.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line161">161</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line162">162</a>   # Note that if it is nil, and &lt;tt&gt;error_probabilities&lt;/tt&gt; is empty,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line163">163</a>   # some methods implicitly assumes that it is :phred (PHRED score).</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line164">164</a>   attr_accessor :quality_score_type</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line165">165</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line166">166</a>   # Error probabilities of the bases/residues in the sequence.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line167">167</a>   # (Array containing Float, or nil)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line168">168</a>   attr_accessor :error_probabilities</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line169">169</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line170">170</a>   #---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line171">171</a>   # Attributes below have been added during BioHackathon2008</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line172">172</a>   #+++</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line173">173</a>   </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line174">174</a>   # Version number of the sequence (String or Integer).</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line175">175</a>   # Unlike &lt;tt&gt;entry_version&lt;/tt&gt;, &lt;tt&gt;sequence_version&lt;/tt&gt; will be changed</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line176">176</a>   # when the submitter of the sequence updates the entry.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line177">177</a>   # Normally, the same entry taken from different databases (EMBL, GenBank,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line178">178</a>   # and DDBJ) may have the same sequence_version.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line179">179</a>   attr_accessor :sequence_version</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line180">180</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line181">181</a>   # Topology (String). &quot;circular&quot;, &quot;linear&quot;, or nil.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line182">182</a>   attr_accessor :topology</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line183">183</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line184">184</a>   # Strandedness (String). &quot;single&quot; (single-stranded),</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line185">185</a>   # &quot;double&quot; (double-stranded), &quot;mixed&quot; (mixed-stranded), or nil.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line186">186</a>   attr_accessor :strandedness</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line187">187</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line188">188</a>   # molecular type (String). &quot;DNA&quot; or &quot;RNA&quot; for nucleotide sequence.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line189">189</a>   attr_accessor :molecule_type</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line190">190</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line191">191</a>   # Data Class defined by EMBL (String)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line192">192</a>   # See http://www.ebi.ac.uk/embl/Documentation/User_manual/usrman.html#3_1</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line193">193</a>   attr_accessor :data_class</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line194">194</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line195">195</a>   # Taxonomic Division defined by EMBL/GenBank/DDBJ (String)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line196">196</a>   # See http://www.ebi.ac.uk/embl/Documentation/User_manual/usrman.html#3_2</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line197">197</a>   attr_accessor :division</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line198">198</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line199">199</a>   # Primary accession number (String)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line200">200</a>   attr_accessor :primary_accession</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line201">201</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line202">202</a>   # Secondary accession numbers (Array of String)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line203">203</a>   attr_accessor :secondary_accessions</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line204">204</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line205">205</a>   # Created date of the sequence entry (Date, DateTime, Time, or String)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line206">206</a>   attr_accessor :date_created</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line207">207</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line208">208</a>   # Last modified date of the sequence entry (Date, DateTime, Time, or String)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line209">209</a>   attr_accessor :date_modified</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line210">210</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line211">211</a>   # Release information when created (String)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line212">212</a>   attr_accessor :release_created</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line213">213</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line214">214</a>   # Release information when last-modified (String)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line215">215</a>   attr_accessor :release_modified</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line216">216</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line217">217</a>   # Version of the entry (String or Integer).</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line218">218</a>   # Unlike &lt;tt&gt;sequence_version&lt;/tt&gt;, &lt;tt&gt;entry_version&lt;/tt&gt; is a database</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line219">219</a>   # maintainer's internal version number.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line220">220</a>   # The version number will be changed when the database maintainer</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line221">221</a>   # modifies the entry.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line222">222</a>   # The same enrty in EMBL, GenBank, and DDBJ may have different</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line223">223</a>   # entry_version.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line224">224</a>   attr_accessor :entry_version</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line225">225</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line226">226</a>   # Organism species (String). For example, &quot;Escherichia coli&quot;.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line227">227</a>   attr_accessor :species</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line228">228</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line229">229</a>   # Organism classification, taxonomic classification of the source organism.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line230">230</a>   # (Array of String)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line231">231</a>   attr_accessor :classification</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line232">232</a>   alias taxonomy classification</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line233">233</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line234">234</a>   # (not well supported) Organelle information (String).</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line235">235</a>   attr_accessor :organelle</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line236">236</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line237">237</a>   # Namespace of the sequence IDs described in entry_id, primary_accession,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line238">238</a>   # and secondary_accessions methods (String).</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line239">239</a>   # For example, 'EMBL', 'GenBank', 'DDBJ', 'RefSeq'.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line240">240</a>   attr_accessor :id_namespace</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line241">241</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line242">242</a>   # Sequence identifiers which are not described in entry_id,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line243">243</a>   # primary_accession,and secondary_accessions methods</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line244">244</a>   # (Array of Bio::Sequence::DBLink objects).</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line245">245</a>   # For example, NCBI GI number can be stored.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line246">246</a>   # Note that only identifiers of the entry itself should be stored.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line247">247</a>   # For database cross references, &lt;tt&gt;dblinks&lt;/tt&gt; should be used.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line248">248</a>   attr_accessor :other_seqids</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line249">249</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line250">250</a>   # Guess the type of sequence, Amino Acid or Nucleic Acid, and create a </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line251">251</a>   # new sequence object (Bio::Sequence::AA or Bio::Sequence::NA) on the basis</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line252">252</a>   # of this guess.  This method will change the current Bio::Sequence object.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line253">253</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line254">254</a>   #   s = Bio::Sequence.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line255">255</a>   #   puts s.seq.class                        #=&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line256">256</a>   #   s.auto</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line257">257</a>   #   puts s.seq.class                        #=&gt; Bio::Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line258">258</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line259">259</a>   # *Returns*:: Bio::Sequence::NA/AA object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line260">260</a>   def auto</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line261">261</a>     @moltype = guess</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line262">262</a>     if @moltype == NA</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line263">263</a>       @seq = NA.new(seq)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line264">264</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line265">265</a>       @seq = AA.new(seq)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line266">266</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line267">267</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line268">268</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line269">269</a>   # Given a sequence String, guess its type, Amino Acid or Nucleic Acid, and</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line270">270</a>   # return a new Bio::Sequence object wrapping a sequence of the guessed type</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line271">271</a>   # (either Bio::Sequence::AA or Bio::Sequence::NA)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line272">272</a>   # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line273">273</a>   #   s = Bio::Sequence.auto('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line274">274</a>   #   puts s.seq.class                        #=&gt; Bio::Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line275">275</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line276">276</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line277">277</a>   # * (required) _str_: String *or* Bio::Sequence::NA/AA object</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line278">278</a>   # *Returns*:: Bio::Sequence object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line279">279</a>   def self.auto(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line280">280</a>     seq = self.new(str)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line281">281</a>     seq.auto</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line282">282</a>     return seq</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line283">283</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line284">284</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line285">285</a>   # Guess the class of the current sequence.  Returns the class</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line286">286</a>   # (Bio::Sequence::AA or Bio::Sequence::NA) guessed.  In general, used by</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line287">287</a>   # developers only, but if you know what you are doing, feel free.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line288">288</a>   # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line289">289</a>   #   s = Bio::Sequence.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line290">290</a>   #   puts s.guess                            #=&gt; Bio::Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line291">291</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line292">292</a>   # There are three parameters: `threshold`, `length`, and `index`.  </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line293">293</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line294">294</a>   # The `threshold` value (defaults to 0.9) is the frequency of </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line295">295</a>   # nucleic acid bases [AGCTUagctu] required in the sequence for this method</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line296">296</a>   # to produce a Bio::Sequence::NA &quot;guess&quot;.  In the default case, if less</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line297">297</a>   # than 90% of the bases (after excluding [Nn]) are in the set [AGCTUagctu],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line298">298</a>   # then the guess is Bio::Sequence::AA.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line299">299</a>   # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line300">300</a>   #   s = Bio::Sequence.new('atgcatgcqq')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line301">301</a>   #   puts s.guess                            #=&gt; Bio::Sequence::AA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line302">302</a>   #   puts s.guess(0.8)                       #=&gt; Bio::Sequence::AA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line303">303</a>   #   puts s.guess(0.7)                       #=&gt; Bio::Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line304">304</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line305">305</a>   # The `length` value is how much of the total sequence to use in the</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line306">306</a>   # guess (default 10000).  If your sequence is very long, you may </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line307">307</a>   # want to use a smaller amount to reduce the computational burden.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line308">308</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line309">309</a>   #   s = Bio::Sequence.new(A VERY LONG SEQUENCE)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line310">310</a>   #   puts s.guess(0.9, 1000)  # limit the guess to the first 1000 positions</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line311">311</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line312">312</a>   # The `index` value is where to start the guess.  Perhaps you know there</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line313">313</a>   # are a lot of gaps at the start...</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line314">314</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line315">315</a>   #   s = Bio::Sequence.new('-----atgcc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line316">316</a>   #   puts s.guess                            #=&gt; Bio::Sequence::AA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line317">317</a>   #   puts s.guess(0.9,10000,5)               #=&gt; Bio::Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line318">318</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line319">319</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line320">320</a>   # * (optional) _threshold_: Float in range 0,1 (default 0.9)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line321">321</a>   # * (optional) _length_: Fixnum (default 10000)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line322">322</a>   # * (optional) _index_: Fixnum (default 1)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line323">323</a>   # *Returns*:: Bio::Sequence::NA/AA</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line324">324</a>   def guess(threshold = 0.9, length = 10000, index = 0)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line325">325</a>     str = seq.to_s[index,length].to_s.extend Bio::Sequence::Common</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line326">326</a>     cmp = str.composition</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line327">327</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line328">328</a>     bases = cmp['A'] + cmp['T'] + cmp['G'] + cmp['C'] + cmp['U'] +</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line329">329</a>             cmp['a'] + cmp['t'] + cmp['g'] + cmp['c'] + cmp['u']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line330">330</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line331">331</a>     total = str.length - cmp['N'] - cmp['n']</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line332">332</a> </pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line333">333</a>     if bases.to_f / total &gt; threshold</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line334">334</a>       return NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line335">335</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line336">336</a>       return AA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line337">337</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line338">338</a>   end </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line339">339</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line340">340</a>   # Guess the class of a given sequence.  Returns the class</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line341">341</a>   # (Bio::Sequence::AA or Bio::Sequence::NA) guessed.  In general, used by</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line342">342</a>   # developers only, but if you know what you are doing, feel free.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line343">343</a>   # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line344">344</a>   #   puts .guess('atgc')        #=&gt; Bio::Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line345">345</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line346">346</a>   # There are three optional parameters: `threshold`, `length`, and `index`.  </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line347">347</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line348">348</a>   # The `threshold` value (defaults to 0.9) is the frequency of </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line349">349</a>   # nucleic acid bases [AGCTUagctu] required in the sequence for this method</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line350">350</a>   # to produce a Bio::Sequence::NA &quot;guess&quot;.  In the default case, if less</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line351">351</a>   # than 90% of the bases (after excluding [Nn]) are in the set [AGCTUagctu],</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line352">352</a>   # then the guess is Bio::Sequence::AA.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line353">353</a>   # </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line354">354</a>   #   puts Bio::Sequence.guess('atgcatgcqq')      #=&gt; Bio::Sequence::AA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line355">355</a>   #   puts Bio::Sequence.guess('atgcatgcqq', 0.8) #=&gt; Bio::Sequence::AA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line356">356</a>   #   puts Bio::Sequence.guess('atgcatgcqq', 0.7) #=&gt; Bio::Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line357">357</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line358">358</a>   # The `length` value is how much of the total sequence to use in the</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line359">359</a>   # guess (default 10000).  If your sequence is very long, you may </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line360">360</a>   # want to use a smaller amount to reduce the computational burden.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line361">361</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line362">362</a>   #   # limit the guess to the first 1000 positions</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line363">363</a>   #   puts Bio::Sequence.guess('A VERY LONG SEQUENCE', 0.9, 1000)  </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line364">364</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line365">365</a>   # The `index` value is where to start the guess.  Perhaps you know there</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line366">366</a>   # are a lot of gaps at the start...</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line367">367</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line368">368</a>   #   puts Bio::Sequence.guess('-----atgcc')             #=&gt; Bio::Sequence::AA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line369">369</a>   #   puts Bio::Sequence.guess('-----atgcc',0.9,10000,5) #=&gt; Bio::Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line370">370</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line371">371</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line372">372</a>   # * (required) _str_: String *or* Bio::Sequence::NA/AA object</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line373">373</a>   # * (optional) _threshold_: Float in range 0,1 (default 0.9)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line374">374</a>   # * (optional) _length_: Fixnum (default 10000)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line375">375</a>   # * (optional) _index_: Fixnum (default 1)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line376">376</a>   # *Returns*:: Bio::Sequence::NA/AA</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line377">377</a>   def self.guess(str, *args)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line378">378</a>     self.new(str).guess(*args)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line379">379</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line380">380</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line381">381</a>   # Transform the sequence wrapped in the current Bio::Sequence object</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line382">382</a>   # into a Bio::Sequence::NA object.  This method will change the current</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line383">383</a>   # object.  This method does not validate your choice, so be careful!</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line384">384</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line385">385</a>   #   s = Bio::Sequence.new('RRLE')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line386">386</a>   #   puts s.seq.class                        #=&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line387">387</a>   #   s.na</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line388">388</a>   #   puts s.seq.class                        #=&gt; Bio::Sequence::NA !!!</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line389">389</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line390">390</a>   # However, if you know your sequence type, this method may be </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line391">391</a>   # constructively used after initialization,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line392">392</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line393">393</a>   #   s = Bio::Sequence.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line394">394</a>   #   s.na</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line395">395</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line396">396</a>   # *Returns*:: Bio::Sequence::NA</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line397">397</a>   def na</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line398">398</a>     @seq = NA.new(seq)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line399">399</a>     @moltype = NA</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line400">400</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line401">401</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line402">402</a>   # Transform the sequence wrapped in the current Bio::Sequence object</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line403">403</a>   # into a Bio::Sequence::NA object.  This method will change the current</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line404">404</a>   # object.  This method does not validate your choice, so be careful!</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line405">405</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line406">406</a>   #   s = Bio::Sequence.new('atgc')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line407">407</a>   #   puts s.seq.class                        #=&gt; String</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line408">408</a>   #   s.aa</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line409">409</a>   #   puts s.seq.class                        #=&gt; Bio::Sequence::AA !!!</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line410">410</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line411">411</a>   # However, if you know your sequence type, this method may be </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line412">412</a>   # constructively used after initialization,</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line413">413</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line414">414</a>   #   s = Bio::Sequence.new('RRLE')</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line415">415</a>   #   s.aa</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line416">416</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line417">417</a>   # *Returns*:: Bio::Sequence::AA</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line418">418</a>   def aa</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line419">419</a>     @seq = AA.new(seq)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line420">420</a>     @moltype = AA</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line421">421</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line422">422</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line423">423</a>   # Create a new Bio::Sequence object from a formatted string</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line424">424</a>   # (GenBank, EMBL, fasta format, etc.)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line425">425</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line426">426</a>   #   s = Bio::Sequence.input(str)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line427">427</a>   # ---</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line428">428</a>   # *Arguments*:</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line429">429</a>   # * (required) _str_: string</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line430">430</a>   # * (optional) _format_: format specification (class or nil)</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line431">431</a>   # *Returns*:: Bio::Sequence object</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line432">432</a>   def self.input(str, format = nil)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line433">433</a>     if format then</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line434">434</a>       klass = format</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line435">435</a>     else</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line436">436</a>       klass = Bio::FlatFile::AutoDetect.default.autodetect(str)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line437">437</a>     end</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line438">438</a>     obj = klass.new(str)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line439">439</a>     obj.to_biosequence</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line440">440</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line441">441</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line442">442</a>   # alias of Bio::Sequence.input</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line443">443</a>   def self.read(str, format = nil)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line444">444</a>     input(str, format)</pre></td>
          </tr>
        
          
          
          <tr class="uncovered">
            <td><pre><a name="line445">445</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line446">446</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line447">447</a>   # accession numbers of the sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line448">448</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line449">449</a>   # *Returns*:: Array of String</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line450">450</a>   def accessions</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line451">451</a>     [ primary_accession, secondary_accessions ].flatten.compact</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line452">452</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line453">453</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line454">454</a>   # Normally, users should not call this method directly.</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line455">455</a>   # Use Bio::*#to_biosequence (e.g. Bio::GenBank#to_biosequence).</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line456">456</a>   #</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line457">457</a>   # Creates a new Bio::Sequence object from database data with an</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line458">458</a>   # adapter module.</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line459">459</a>   def self.adapter(source_data, adapter_module)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line460">460</a>     biosequence = self.new(nil)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line461">461</a>     biosequence.instance_eval {</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line462">462</a>       remove_instance_variable(:@seq)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line463">463</a>       @source_data = source_data</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line464">464</a>     }</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line465">465</a>     biosequence.extend(adapter_module)</pre></td>
          </tr>
        
          
          
          <tr class="marked">
            <td><pre><a name="line466">466</a>     biosequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line467">467</a>   end</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line468">468</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line469">469</a> end # Sequence</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line470">470</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line471">471</a> </pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line472">472</a> end # Bio</pre></td>
          </tr>
        
          
          
          <tr class="inferred">
            <td><pre><a name="line473">473</a> </pre></td>
          </tr>
        
      </tbody>
    </table>

    <p>Generated on Fri Jun 25 17:37:23 +0900 2010 with <a href="http://github.com/relevance/rcov">rcov 0.9.8</a></p>

  </body>
</html>
